<!DOCTYPE html>
<!-- START: inst/pkgdown/templates/layout.html --><!-- Generated by pkgdown: do not edit by hand --><html lang="en" data-bs-theme="auto">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<title>Machine Learning for Biomedical Science: All in One View</title>
<meta name="viewport" content="width=device-width, initial-scale=1">
<script src="../assets/themetoggle.js"></script><link rel="stylesheet" type="text/css" href="../assets/styles.css">
<script src="../assets/scripts.js" type="text/javascript"></script><!-- mathjax --><script type="text/x-mathjax-config">
    MathJax.Hub.Config({
      config: ["MMLorHTML.js"],
      jax: ["input/TeX","input/MathML","output/HTML-CSS","output/NativeMML", "output/PreviewHTML"],
      extensions: ["tex2jax.js","mml2jax.js","MathMenu.js","MathZoom.js", "fast-preview.js", "AssistiveMML.js", "a11y/accessibility-menu.js"],
      TeX: {
        extensions: ["AMSmath.js","AMSsymbols.js","noErrors.js","noUndefined.js"]
      },
      tex2jax: {
        inlineMath: [['\\(', '\\)']],
        displayMath: [ ['$$','$$'], ['\\[', '\\]'] ],
        processEscapes: true
      }
    });
    </script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><!-- Responsive Favicon for The Carpentries --><link rel="apple-touch-icon" sizes="180x180" href="../favicons/incubator/apple-touch-icon.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicons/incubator/favicon-32x32.png">
<link rel="icon" type="image/png" sizes="16x16" href="../favicons/incubator/favicon-16x16.png">
<link rel="manifest" href="../favicons/incubator/site.webmanifest">
<link rel="mask-icon" href="../favicons/incubator/safari-pinned-tab.svg" color="#5bbad5">
<meta name="msapplication-TileColor" content="#da532c">
<meta name="theme-color" media="(prefers-color-scheme: light)" content="white">
<meta name="theme-color" media="(prefers-color-scheme: dark)" content="black">
</head>
<body>
    <header id="top" class="navbar navbar-expand-md top-nav incubator"><svg xmlns="http://www.w3.org/2000/svg" class="d-none"><symbol id="check2" viewbox="0 0 16 16"><path d="M13.854 3.646a.5.5 0 0 1 0 .708l-7 7a.5.5 0 0 1-.708 0l-3.5-3.5a.5.5 0 1 1 .708-.708L6.5 10.293l6.646-6.647a.5.5 0 0 1 .708 0z"></path></symbol><symbol id="circle-half" viewbox="0 0 16 16"><path d="M8 15A7 7 0 1 0 8 1v14zm0 1A8 8 0 1 1 8 0a8 8 0 0 1 0 16z"></path></symbol><symbol id="moon-stars-fill" viewbox="0 0 16 16"><path d="M6 .278a.768.768 0 0 1 .08.858 7.208 7.208 0 0 0-.878 3.46c0 4.021 3.278 7.277 7.318 7.277.527 0 1.04-.055 1.533-.16a.787.787 0 0 1 .81.316.733.733 0 0 1-.031.893A8.349 8.349 0 0 1 8.344 16C3.734 16 0 12.286 0 7.71 0 4.266 2.114 1.312 5.124.06A.752.752 0 0 1 6 .278z"></path><path d="M10.794 3.148a.217.217 0 0 1 .412 0l.387 1.162c.173.518.579.924 1.097 1.097l1.162.387a.217.217 0 0 1 0 .412l-1.162.387a1.734 1.734 0 0 0-1.097 1.097l-.387 1.162a.217.217 0 0 1-.412 0l-.387-1.162A1.734 1.734 0 0 0 9.31 6.593l-1.162-.387a.217.217 0 0 1 0-.412l1.162-.387a1.734 1.734 0 0 0 1.097-1.097l.387-1.162zM13.863.099a.145.145 0 0 1 .274 0l.258.774c.115.346.386.617.732.732l.774.258a.145.145 0 0 1 0 .274l-.774.258a1.156 1.156 0 0 0-.732.732l-.258.774a.145.145 0 0 1-.274 0l-.258-.774a1.156 1.156 0 0 0-.732-.732l-.774-.258a.145.145 0 0 1 0-.274l.774-.258c.346-.115.617-.386.732-.732L13.863.1z"></path></symbol><symbol id="sun-fill" viewbox="0 0 16 16"><path d="M8 12a4 4 0 1 0 0-8 4 4 0 0 0 0 8zM8 0a.5.5 0 0 1 .5.5v2a.5.5 0 0 1-1 0v-2A.5.5 0 0 1 8 0zm0 13a.5.5 0 0 1 .5.5v2a.5.5 0 0 1-1 0v-2A.5.5 0 0 1 8 13zm8-5a.5.5 0 0 1-.5.5h-2a.5.5 0 0 1 0-1h2a.5.5 0 0 1 .5.5zM3 8a.5.5 0 0 1-.5.5h-2a.5.5 0 0 1 0-1h2A.5.5 0 0 1 3 8zm10.657-5.657a.5.5 0 0 1 0 .707l-1.414 1.415a.5.5 0 1 1-.707-.708l1.414-1.414a.5.5 0 0 1 .707 0zm-9.193 9.193a.5.5 0 0 1 0 .707L3.05 13.657a.5.5 0 0 1-.707-.707l1.414-1.414a.5.5 0 0 1 .707 0zm9.193 2.121a.5.5 0 0 1-.707 0l-1.414-1.414a.5.5 0 0 1 .707-.707l1.414 1.414a.5.5 0 0 1 0 .707zM4.464 4.465a.5.5 0 0 1-.707 0L2.343 3.05a.5.5 0 1 1 .707-.707l1.414 1.414a.5.5 0 0 1 0 .708z"></path></symbol></svg><a class="visually-hidden-focusable skip-link" href="#main-content">Skip to main content</a>
  <div class="container-fluid top-nav-container">
    <div class="col-md-8">
      <div class="large-logo">
        <img id="incubator-logo" alt="Lesson Description" src="../assets/images/incubator-logo.svg"><span class="badge text-bg-warning">
          <abbr title="This lesson is in the alpha phase, which means that it has been taught once and lesson authors are iterating on feedback.">
            <a href="https://docs.carpentries.org/resources/curriculum/lesson-life-cycle.html" class="external-link alert-link">
              <i aria-hidden="true" class="icon" data-feather="alert-triangle" style="border-radius: 5px"></i>
              Alpha
            </a>
            <span class="visually-hidden">This lesson is in the alpha phase, which means that it has been taught once and lesson authors are iterating on feedback.</span>
          </abbr>
        </span>

      </div>
    </div>
    <div class="selector-container">
      <div id="theme-selector">
        <li class="nav-item dropdown" id="theme-button-list">
          <button class="btn btn-link nav-link px-0 px-lg-2 dropdown-toggle d-flex align-items-center" id="bd-theme" type="button" aria-expanded="false" data-bs-toggle="dropdown" data-bs-display="static" aria-label="Toggle theme (auto)">
            <svg class="bi my-1 theme-icon-active"><use href="#circle-half"></use></svg><i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="bd-theme-text">
<li>
              <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="light" aria-pressed="false">
                <svg class="bi me-2 theme-icon"><use href="#sun-fill"></use></svg>
                Light
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
            <li>
              <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="dark" aria-pressed="false">
                <svg class="bi me-2 theme-icon"><use href="#moon-stars-fill"></use></svg>
                Dark
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
            <li>
              <button type="button" class="btn dropdown-item d-flex align-items-center active" data-bs-theme-value="auto" aria-pressed="true">
                <svg class="bi me-2 theme-icon"><use href="#circle-half"></use></svg>
                Auto
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
          </ul>
</li>
      </div>

      <div class="dropdown" id="instructor-dropdown">
        <button class="btn btn-secondary dropdown-toggle bordered-button" type="button" id="dropdownMenu1" data-bs-toggle="dropdown" aria-expanded="false">
          <i aria-hidden="true" class="icon" data-feather="eye"></i> Instructor View <i data-feather="chevron-down"></i>
        </button>
        <ul class="dropdown-menu" aria-labelledby="dropdownMenu1">
<li><button class="dropdown-item" type="button" onclick="window.location.href='../aio.html';">Learner View</button></li>
        </ul>
</div>
    </div>
  </div>
  <hr></header><nav class="navbar navbar-expand-xl bottom-nav incubator" aria-label="Main Navigation"><div class="container-fluid nav-container">
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarSupportedContent" aria-controls="navbarSupportedContent" aria-expanded="false" aria-label="Toggle Navigation">
      <span class="navbar-toggler-icon"></span>
      <span class="menu-title">Menu</span>
    </button>
    <div class="nav-logo">
      <img class="small-logo" alt="Lesson Description" src="../assets/images/incubator-logo-sm.svg">
</div>
    <div class="lesson-title-md">
      Machine Learning for Biomedical Science
    </div>
    <div class="search-icon-sm">
      <!-- TODO: do not show until we have search
        <i role="img" aria-label="Search the All In One page" data-feather="search"></i>
      -->
    </div>
    <div class="desktop-nav">
      <ul class="navbar-nav me-auto mb-2 mb-lg-0">
<li class="nav-item">
          <span class="lesson-title">
            Machine Learning for Biomedical Science
          </span>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="../instructor/key-points.html">Key Points</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="../instructor/instructor-notes.html">Instructor Notes</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="../instructor/images.html">Extract All Images</a>
        </li>
        <li class="nav-item dropdown">
          <button class="nav-link dropdown-toggle" id="navbarDropdown" data-bs-toggle="dropdown" aria-expanded="false">
            More <i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu" aria-labelledby="navbarDropdown">
<hr>
<li><a class="dropdown-item" href="reference.html">Reference</a></li>
          </ul>
</li>
      </ul>
</div>
    <!--
    <form class="d-flex col-md-2 search-form">
      <fieldset disabled>
      <input class="form-control me-2 searchbox" type="search" placeholder="" aria-label="">
        <button class="btn btn-outline-success tablet-search-button"  type="submit">
          <i class="search-icon" data-feather="search" role="img" aria-label="Search the All In One page"></i>
        </button>
      </fieldset>
    </form>
    -->
    <a id="search-button" class="btn btn-primary" href="../instructor/aio.html" role="button" aria-label="Search the All In One page">Search the All In One page</a>
  </div>
<!--/div.container-fluid -->
</nav><div class="col-md-12 mobile-title">
  Machine Learning for Biomedical Science
</div>

<aside class="col-md-12 lesson-progress"><div style="width: %" class="percentage">
    %
  </div>
  <div class="progress incubator">
    <div class="progress-bar incubator" role="progressbar" style="width: %" aria-valuenow="" aria-label="Lesson Progress" aria-valuemin="0" aria-valuemax="100">
    </div>
  </div>
</aside><div class="container">
      <div class="row">
        <!-- START: inst/pkgdown/templates/navbar.html -->
<div id="sidebar-col" class="col-lg-4">
  <div id="sidebar" class="sidebar">
      <nav aria-labelledby="flush-headingEleven"><button role="button" aria-label="close menu" alt="close menu" aria-expanded="true" aria-controls="sidebar" class="collapse-toggle" data-collapse="Collapse " data-episodes="Episodes ">
          <i class="search-icon" data-feather="x" role="img"></i>
        </button>
        <div class="sidebar-inner">
          <div class="row mobile-row" id="theme-row-mobile">
            <div class="col" id="theme-selector">
              <li class="nav-item dropdown" id="theme-button-list">
                <button class="btn btn-link nav-link px-0 px-lg-2 dropdown-toggle d-flex align-items-center" id="bd-theme" type="button" aria-expanded="false" data-bs-toggle="dropdown" data-bs-display="static" aria-label="Toggle theme (auto)">
                  <svg class="bi my-1 theme-icon-active"><use href="#circle-half"></use></svg><span class="d-lg-none ms-1" id="bd-theme-text">Toggle Theme</span>
                </button>
                <ul class="dropdown-menu dropdown-menu-right" aria-labelledby="bd-theme-text">
<li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="light" aria-pressed="false">
                      <svg class="bi me-2 theme-icon"><use href="#sun-fill"></use></svg>
                      Light
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                  <li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="dark" aria-pressed="false">
                      <svg class="bi me-2 theme-icon"><use href="#moon-stars-fill"></use></svg>
                      Dark
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                  <li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center active" data-bs-theme-value="auto" aria-pressed="true">
                      <svg class="bi me-2 theme-icon"><use href="#circle-half"></use></svg>
                      Auto
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                </ul>
</li>
            </div>
          </div>
          <div class="row mobile-row">
            <div class="col">
              <div class="sidenav-view-selector">
                <div class="accordion accordion-flush" id="accordionFlush9">
                  <div class="accordion-item">
                    <h2 class="accordion-header" id="flush-headingNine">
                      <button class="accordion-button collapsed" id="instructor" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseNine" aria-expanded="false" aria-controls="flush-collapseNine">
                        <i id="eye" aria-hidden="true" class="icon" data-feather="eye"></i> Instructor View
                      </button>
                    </h2>
                    <div id="flush-collapseNine" class="accordion-collapse collapse" aria-labelledby="flush-headingNine" data-bs-parent="#accordionFlush2">
                      <div class="accordion-body">
                        <a href="../aio.html">Learner View</a>
                      </div>
                    </div>
                  </div>
<!--/div.accordion-item-->
                </div>
<!--/div.accordion-flush-->
              </div>
<!--div.sidenav-view-selector -->
            </div>
<!--/div.col -->

            <hr>
</div>
<!--/div.mobile-row -->

          <div class="accordion accordion-flush" id="accordionFlush11">
            <div class="accordion-item">

              <button id="chapters" class="accordion-button show" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseEleven" aria-expanded="false" aria-controls="flush-collapseEleven">
                <h2 class="accordion-header chapters" id="flush-headingEleven">
                  EPISODES
                </h2>
              </button>
              <div id="flush-collapseEleven" class="accordion-collapse show collapse" aria-labelledby="flush-headingEleven" data-bs-parent="#accordionFlush11">

                <div class="accordion-body">
                  <div class="accordion accordion-flush" id="accordionFlush1">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading1">
        <a href="index.html">Summary and Schedule</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush2">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading2">
        <a href="introduction.html">1. Basic Machine Learning</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush3">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading3">
        <a href="clustering.html">2. Clustering</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush4">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading4">
        <a href="conditional-probabilities-expectations.html">3. Conditional Probabilities and Expectations</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush5">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading5">
        <a href="smoothing.html">4. Smoothing</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush6">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading6">
        <a href="class-prediction.html">5. Class Prediction</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush7">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading7">
        <a href="cross-validation.html">6. Cross-validation</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

                </div>
              </div>
            </div>

            <hr class="half-width">
<div class="accordion accordion-flush lesson-resources" id="accordionFlush12">
              <div class="accordion-item">
                <h2 class="accordion-header" id="flush-headingTwelve">
                  <button class="accordion-button collapsed" id="lesson-resources" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseTwelve" aria-expanded="false" aria-controls="flush-collapseTwelve">
                    RESOURCES
                  </button>
                </h2>
                <div id="flush-collapseTwelve" class="accordion-collapse collapse" aria-labelledby="flush-headingTwelve" data-bs-parent="#accordionFlush12">
                  <div class="accordion-body">
                    <ul>
<li>
                        <a href="../instructor/key-points.html">Key Points</a>
                      </li>
                      <li>
                        <a href="../instructor/instructor-notes.html">Instructor Notes</a>
                      </li>
                      <li>
                        <a href="../instructor/images.html">Extract All Images</a>
                      </li>
                      <hr>
<li><a class="dropdown-item" href="reference.html">Reference</a></li>
                    </ul>
</div>
                </div>
              </div>
            </div>
            <hr class="half-width lesson-resources">
<a href="../instructor/aio.html">See all in one page</a>


            <hr class="d-none d-sm-block d-md-none">
<div class="d-grid gap-1">

            </div>
          </div>
<!-- /div.accordion -->
        </div>
<!-- /div.sidebar-inner -->
      </nav>
</div>
<!-- /div.sidebar -->
  </div>
<!-- /div.sidebar-col -->
<!-- END:   inst/pkgdown/templates/navbar.html-->

        <!-- START: inst/pkgdown/templates/content-extra.html -->
  <div class="col-xl-8 col-lg-12 primary-content">
    <main id="main-content" class="main-content"><div class="container lesson-content">
        
        
<section id="aio-introduction"><p>Content from <a href="introduction.html">Basic Machine Learning</a></p>
<hr>
<p>Last updated on 2025-07-18 |

        <a href="https://github.com/smcclatchy/machine-learning-biomedical-science/edit/main/episodes/introduction.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<p>Estimated time: <i aria-hidden="true" data-feather="clock"></i> 10 minutes</p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>What is machine learning?</li>
<li>How is machine learning used in biomedical studies?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Describe how machine learning is used in biomedical science.</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<section><h2 class="section-heading" id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<hr class="half-width">
<p>Machine learning is a very broad topic and a highly active research
area. In the life sciences, much of what is described as “precision
medicine” is an application of machine learning to biomedical data. The
general idea is to predict or discover outcomes from measured
predictors. Can we discover new types of cancer from gene expression
profiles? Can we predict drug response from a series of genotypes? Here
we give a very brief introduction to two major machine learning
components: clustering and class prediction. There are many good
resources to learn more about machine learning, for example the
excellent textbook <em>The Elements of Statistical Learning: Data
Mining, Inference, and Prediction</em>, by Trevor Hastie, Robert
Tibshirani and Jerome Friedman. A free PDF of this book can be found <a href="https://www.sas.upenn.edu/~fdiebold/NoHesitations/BookAdvanced.pdf" class="external-link">here</a>.</p>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Key Points</h3>
<div class="callout-content">
<ul>
<li>Machine learning predicts outcomes from data.</li>
<li>As examples, machine learning can be used to discover new kinds of
cancers or predict drug response in biomedical studies.</li>
</ul>
</div>
</div>
</div>
<!--
Place links that you need to refer to multiple times across pages here. Delete
any links that you are not going to use.
 -->
</section></section><section id="aio-clustering"><p>Content from <a href="clustering.html">Clustering</a></p>
<hr>
<p>Last updated on 2025-07-18 |

        <a href="https://github.com/smcclatchy/machine-learning-biomedical-science/edit/main/episodes/clustering.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<p>Estimated time: <i aria-hidden="true" data-feather="clock"></i> 60 minutes</p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>How can clusters within high dimensional data can be
discovered?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Perform clustering analysis.</li>
<li>Interpret a heatmap.</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<section><h2 class="section-heading" id="clustering">Clustering<a class="anchor" aria-label="anchor" href="#clustering"></a>
</h2>
<hr class="half-width">
<p>We will demonstrate the concepts and code needed to perform
clustering analysis with the tissue gene expression data:</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">load</span><span class="op">(</span><span class="st">"data/tissuesGeneExpression.rda"</span><span class="op">)</span></span></code></pre>
</div>
<p>To illustrate the main application of clustering in the life
sciences, let’s pretend that we don’t know these are different tissues
and are interested in clustering. The first step is to compute the
distance between each sample:</p>
<div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">d</span> <span class="op">&lt;-</span> <span class="fu">dist</span><span class="op">(</span> <span class="fu">t</span><span class="op">(</span><span class="va">e</span><span class="op">)</span> <span class="op">)</span></span></code></pre>
</div>
<div class="section level4">
<h4 id="hierarchical-clustering">Hierarchical clustering<a class="anchor" aria-label="anchor" href="#hierarchical-clustering"></a>
</h4>
<p>With the distance between each pair of samples computed, we need
clustering algorithms to join them into groups. Hierarchical clustering
is one of the many clustering algorithms available to do this. Each
sample is assigned to its own group and then the algorithm continues
iteratively, joining the two most similar clusters at each step, and
continuing until there is just one group. While we have defined
distances between samples, we have not yet defined distances between
groups. There are various ways this can be done and they all rely on the
individual pairwise distances. The helpfile for <code>hclust</code>
includes detailed information.</p>
<p>We can perform hierarchical clustering based on the distances defined
above using the <code>hclust</code> function. This function returns an
<code>hclust</code> object that describes the groupings that were
created using the algorithm described above. The <code>plot</code>
method represents these relationships with a tree or dendrogram:</p>
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">rafalib</span><span class="op">)</span></span>
<span><span class="fu">mypar</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">hc</span> <span class="op">&lt;-</span> <span class="fu">hclust</span><span class="op">(</span><span class="va">d</span><span class="op">)</span></span>
<span><span class="va">hc</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>
Call:
hclust(d = d)

Cluster method   : complete
Distance         : euclidean
Number of objects: 189 </code></pre>
</div>
<div class="codewrapper sourceCode" id="cb5">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">plot</span><span class="op">(</span><span class="va">hc</span>,labels<span class="op">=</span><span class="va">tissue</span>,cex<span class="op">=</span><span class="fl">0.5</span><span class="op">)</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="../fig/clustering-rendered-dendrogram-1.png" alt="Dendrogram showing hierarchical clustering of tissue gene expression data." class="figure mx-auto d-block"><figcaption>
Dendrogram showing hierarchical clustering of tissue gene expression
data.
</figcaption></figure><p>Does this technique “discover” the clusters defined by the different
tissues? In this plot, it is not easy to see the different tissues so we
add colors by using the <code>myplclust</code> function from the
<code>rafalib</code> package.</p>
<div class="codewrapper sourceCode" id="cb6">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">myplclust</span><span class="op">(</span><span class="va">hc</span>, labels<span class="op">=</span><span class="va">tissue</span>, lab.col<span class="op">=</span><span class="fu">as.fumeric</span><span class="op">(</span><span class="va">tissue</span><span class="op">)</span>, cex<span class="op">=</span><span class="fl">0.5</span><span class="op">)</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="../fig/clustering-rendered-color_dendrogram-1.png" alt="Dendrogram showing hierarchical clustering of tissue gene expression data with colors denoting tissues." class="figure mx-auto d-block"><figcaption>
Dendrogram showing hierarchical clustering of tissue gene expression
data with colors denoting tissues.
</figcaption></figure><p>Visually, it does seem as if the clustering technique has discovered
the tissues. However, hierarchical clustering does not define specific
clusters, but rather defines the dendrogram above. From the dendrogram
we can decipher the distance between any two groups by looking at the
height at which the two groups split into two. To define clusters, we
need to “cut the tree” at some distance and group all samples that are
within that distance into groups below. To visualize this, we draw a
horizontal line at the height we wish to cut and this defines that line.
We use 120 as an example:</p>
<div class="codewrapper sourceCode" id="cb7">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">myplclust</span><span class="op">(</span><span class="va">hc</span>, labels<span class="op">=</span><span class="va">tissue</span>, lab.col<span class="op">=</span><span class="fu">as.fumeric</span><span class="op">(</span><span class="va">tissue</span><span class="op">)</span>,cex<span class="op">=</span><span class="fl">0.5</span><span class="op">)</span></span>
<span><span class="fu">abline</span><span class="op">(</span>h<span class="op">=</span><span class="fl">120</span><span class="op">)</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="../fig/clustering-rendered-color_dendrogram2-1.png" alt="Dendrogram showing hierarchical clustering of tissue gene expression data with colors denoting tissues. Horizontal line defines actual clusters." class="figure mx-auto d-block"><figcaption>
Dendrogram showing hierarchical clustering of tissue gene expression
data with colors denoting tissues. Horizontal line defines actual
clusters.
</figcaption></figure><p>If we use the line above to cut the tree into clusters, we can
examine how the clusters overlap with the actual tissues:</p>
<div class="codewrapper sourceCode" id="cb8">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">hclusters</span> <span class="op">&lt;-</span> <span class="fu">cutree</span><span class="op">(</span><span class="va">hc</span>, h<span class="op">=</span><span class="fl">120</span><span class="op">)</span></span>
<span><span class="fu">table</span><span class="op">(</span>true<span class="op">=</span><span class="va">tissue</span>, cluster<span class="op">=</span><span class="va">hclusters</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>             cluster
true           1  2  3  4  5  6  7  8  9 10 11 12 13 14
  cerebellum   0  0  0  0 31  0  0  0  2  0  0  5  0  0
  colon        0  0  0  0  0  0 34  0  0  0  0  0  0  0
  endometrium  0  0  0  0  0  0  0  0  0  0 15  0  0  0
  hippocampus  0  0 12 19  0  0  0  0  0  0  0  0  0  0
  kidney       9 18  0  0  0 10  0  0  2  0  0  0  0  0
  liver        0  0  0  0  0  0  0 24  0  2  0  0  0  0
  placenta     0  0  0  0  0  0  0  0  0  0  0  0  2  4</code></pre>
</div>
<p>We can also ask <code>cutree</code> to give us back a given number of
clusters. The function then automatically finds the height that results
in the requested number of clusters:</p>
<div class="codewrapper sourceCode" id="cb10">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">hclusters</span> <span class="op">&lt;-</span> <span class="fu">cutree</span><span class="op">(</span><span class="va">hc</span>, k<span class="op">=</span><span class="fl">8</span><span class="op">)</span></span>
<span><span class="fu">table</span><span class="op">(</span>true<span class="op">=</span><span class="va">tissue</span>, cluster<span class="op">=</span><span class="va">hclusters</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>             cluster
true           1  2  3  4  5  6  7  8
  cerebellum   0  0 31  0  0  2  5  0
  colon        0  0  0 34  0  0  0  0
  endometrium 15  0  0  0  0  0  0  0
  hippocampus  0 12 19  0  0  0  0  0
  kidney      37  0  0  0  0  2  0  0
  liver        0  0  0  0 24  2  0  0
  placenta     0  0  0  0  0  0  0  6</code></pre>
</div>
<p>In both cases we do see that, with some exceptions, each tissue is
uniquely represented by one of the clusters. In some instances, the one
tissue is spread across two tissues, which is due to selecting too many
clusters. Selecting the number of clusters is generally a challenging
step in practice and an active area of research.</p>
</div>
<div class="section level4">
<h4 id="k-means">K-means<a class="anchor" aria-label="anchor" href="#k-means"></a>
</h4>
<p>We can also cluster with the <code>kmeans</code> function to perform
k-means clustering. As an example, let’s run k-means on the samples in
the space of the first two genes:</p>
<div class="codewrapper sourceCode" id="cb12">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">set.seed</span><span class="op">(</span><span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">km</span> <span class="op">&lt;-</span> <span class="fu">kmeans</span><span class="op">(</span><span class="fu">t</span><span class="op">(</span><span class="va">e</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span>,<span class="op">]</span><span class="op">)</span>, centers<span class="op">=</span><span class="fl">7</span><span class="op">)</span></span>
<span><span class="fu">names</span><span class="op">(</span><span class="va">km</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "cluster"      "centers"      "totss"        "withinss"     "tot.withinss"
[6] "betweenss"    "size"         "iter"         "ifault"      </code></pre>
</div>
<div class="codewrapper sourceCode" id="cb14">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">mypar</span><span class="op">(</span><span class="fl">1</span>,<span class="fl">2</span><span class="op">)</span></span>
<span><span class="fu">plot</span><span class="op">(</span><span class="va">e</span><span class="op">[</span><span class="fl">1</span>,<span class="op">]</span>, <span class="va">e</span><span class="op">[</span><span class="fl">2</span>,<span class="op">]</span>, col<span class="op">=</span><span class="fu">as.fumeric</span><span class="op">(</span><span class="va">tissue</span><span class="op">)</span>, pch<span class="op">=</span><span class="fl">16</span><span class="op">)</span></span>
<span><span class="fu">plot</span><span class="op">(</span><span class="va">e</span><span class="op">[</span><span class="fl">1</span>,<span class="op">]</span>, <span class="va">e</span><span class="op">[</span><span class="fl">2</span>,<span class="op">]</span>, col<span class="op">=</span><span class="va">km</span><span class="op">$</span><span class="va">cluster</span>, pch<span class="op">=</span><span class="fl">16</span><span class="op">)</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="../fig/clustering-rendered-kmeans-1.png" alt="Plot of gene expression for first two genes (order of appearance in data) with color representing tissue (left) and clusters found with kmeans (right)." class="figure mx-auto d-block"><figcaption>
Plot of gene expression for first two genes (order of appearance in
data) with color representing tissue (left) and clusters found with
kmeans (right).
</figcaption></figure><p>In the first plot, color represents the actual tissues, while in the
second, color represents the clusters that were defined by
<code>kmeans</code>. We can see from tabulating the results that this
particular clustering exercise did not perform well:</p>
<div class="codewrapper sourceCode" id="cb15">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">table</span><span class="op">(</span>true<span class="op">=</span><span class="va">tissue</span>,cluster<span class="op">=</span><span class="va">km</span><span class="op">$</span><span class="va">cluster</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>             cluster
true           1  2  3  4  5  6  7
  cerebellum   1  0  0 13  6  4 14
  colon        3  0 22  0  6  3  0
  endometrium  3  0  0  6  0  2  4
  hippocampus  0  0  0  0 16 15  0
  kidney      10  0  2  1  0  9 17
  liver        0 18  0  7  0  0  1
  placenta     4  0  0  1  0  0  1</code></pre>
</div>
<p>This is very likely due to the fact that the first two genes are not
informative regarding tissue type. We can see this in the first plot
above. If we instead perform k-means clustering using all of the genes,
we obtain a much improved result. To visualize this, we can use an MDS
plot:</p>
<div class="codewrapper sourceCode" id="cb17">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">km</span> <span class="op">&lt;-</span> <span class="fu">kmeans</span><span class="op">(</span><span class="fu">t</span><span class="op">(</span><span class="va">e</span><span class="op">)</span>, centers<span class="op">=</span><span class="fl">7</span><span class="op">)</span></span>
<span><span class="va">mds</span> <span class="op">&lt;-</span> <span class="fu">cmdscale</span><span class="op">(</span><span class="va">d</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">mypar</span><span class="op">(</span><span class="fl">1</span>,<span class="fl">2</span><span class="op">)</span></span>
<span><span class="fu">plot</span><span class="op">(</span><span class="va">mds</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span>, <span class="va">mds</span><span class="op">[</span>,<span class="fl">2</span><span class="op">]</span><span class="op">)</span> </span>
<span><span class="fu">plot</span><span class="op">(</span><span class="va">mds</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span>, <span class="va">mds</span><span class="op">[</span>,<span class="fl">2</span><span class="op">]</span>, col<span class="op">=</span><span class="va">km</span><span class="op">$</span><span class="va">cluster</span>, pch<span class="op">=</span><span class="fl">16</span><span class="op">)</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="../fig/clustering-rendered-kmeans_mds-1.png" alt="Plot of gene expression for first two PCs with color representing tissues (left) and clusters found using all genes (right)." class="figure mx-auto d-block"><figcaption>
Plot of gene expression for first two PCs with color representing
tissues (left) and clusters found using all genes (right).
</figcaption></figure><p>By tabulating the results, we see that we obtain a similar answer to
that obtained with hierarchical clustering.</p>
<div class="codewrapper sourceCode" id="cb18">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">table</span><span class="op">(</span>true<span class="op">=</span><span class="va">tissue</span>,cluster<span class="op">=</span><span class="va">km</span><span class="op">$</span><span class="va">cluster</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>             cluster
true           1  2  3  4  5  6  7
  cerebellum   0  2  0  5  0  0 31
  colon        0  0 34  0  0  0  0
  endometrium  0  0  0  0  0 15  0
  hippocampus  0  0  0 31  0  0  0
  kidney       0  2  0  0 19 18  0
  liver       24  2  0  0  0  0  0
  placenta     0  0  6  0  0  0  0</code></pre>
</div>
</div>
<div class="section level4">
<h4 id="heatmaps">Heatmaps<a class="anchor" aria-label="anchor" href="#heatmaps"></a>
</h4>
<p>Heatmaps are ubiquitous in the genomics literature. They are very
useful plots for visualizing the measurements for a subset of rows over
all the samples. A <em>dendrogram</em> is added on top and on the side
that is created with hierarchical clustering. We will demonstrate how to
create heatmaps from within R. Let’s begin by defining a color
palette:</p>
<div class="codewrapper sourceCode" id="cb20">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">RColorBrewer</span><span class="op">)</span> </span>
<span><span class="va">hmcol</span> <span class="op">&lt;-</span> <span class="fu">colorRampPalette</span><span class="op">(</span><span class="fu">brewer.pal</span><span class="op">(</span><span class="fl">9</span>, <span class="st">"GnBu"</span><span class="op">)</span><span class="op">)</span><span class="op">(</span><span class="fl">100</span><span class="op">)</span></span></code></pre>
</div>
<p>Now, pick the genes with the top variance over all samples:</p>
<div class="codewrapper sourceCode" id="cb21">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">genefilter</span><span class="op">)</span></span>
<span><span class="va">rv</span> <span class="op">&lt;-</span> <span class="fu">rowVars</span><span class="op">(</span><span class="va">e</span><span class="op">)</span></span>
<span><span class="va">idx</span> <span class="op">&lt;-</span> <span class="fu">order</span><span class="op">(</span><span class="op">-</span><span class="va">rv</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">40</span><span class="op">]</span></span></code></pre>
</div>
<p>While a <code>heatmap</code> function is included in R, we recommend
the <code>heatmap.2</code> function from the <code>gplots</code> package
on CRAN because it is a bit more customized. For example, it stretches
to fill the window. Here we add colors to indicate the tissue on the
top:</p>
<div class="codewrapper sourceCode" id="cb22">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">gplots</span><span class="op">)</span> <span class="co">##Available from CRAN</span></span>
<span><span class="va">cols</span> <span class="op">&lt;-</span> <span class="fu">palette</span><span class="op">(</span><span class="fu">brewer.pal</span><span class="op">(</span><span class="fl">8</span>, <span class="st">"Dark2"</span><span class="op">)</span><span class="op">)</span><span class="op">[</span><span class="fu">as.fumeric</span><span class="op">(</span><span class="va">tissue</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="fu">head</span><span class="op">(</span><span class="fu">cbind</span><span class="op">(</span><span class="fu">colnames</span><span class="op">(</span><span class="va">e</span><span class="op">)</span>,<span class="va">cols</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>                       cols
[1,] "GSM11805.CEL.gz" "#1B9E77"
[2,] "GSM11814.CEL.gz" "#1B9E77"
[3,] "GSM11823.CEL.gz" "#1B9E77"
[4,] "GSM11830.CEL.gz" "#1B9E77"
[5,] "GSM12067.CEL.gz" "#1B9E77"
[6,] "GSM12075.CEL.gz" "#1B9E77"</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb24">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">heatmap.2</span><span class="op">(</span><span class="va">e</span><span class="op">[</span><span class="va">idx</span>,<span class="op">]</span>, labCol<span class="op">=</span><span class="va">tissue</span>,</span>
<span>          trace<span class="op">=</span><span class="st">"none"</span>, </span>
<span>          ColSideColors<span class="op">=</span><span class="va">cols</span>, </span>
<span>          col<span class="op">=</span><span class="va">hmcol</span><span class="op">)</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="../fig/clustering-rendered-heatmap.2-1.png" alt="Heatmap created using the 40 most variable genes and the function heatmap.2." class="figure mx-auto d-block"><figcaption>
Heatmap created using the 40 most variable genes and the function
heatmap.2.
</figcaption></figure><p>We did not use tissue information to create this heatmap, and we can
quickly see, with just 40 genes, good separation across tissues.</p>
<div id="exercise-1" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="exercise-1" class="callout-inner">
<h3 class="callout-title">Exercise 1</h3>
<div class="callout-content">
<p>Create a random matrix with no correlation in the following way:</p>
<div class="codewrapper sourceCode" id="cb25">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">set.seed</span><span class="op">(</span><span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">m</span> <span class="op">=</span> <span class="fl">10000</span></span>
<span><span class="va">n</span> <span class="op">=</span> <span class="fl">24</span></span>
<span><span class="va">x</span> <span class="op">=</span> <span class="fu">matrix</span><span class="op">(</span><span class="fu">rnorm</span><span class="op">(</span><span class="va">m</span> <span class="op">*</span> <span class="va">n</span><span class="op">)</span>, <span class="va">m</span>, <span class="va">n</span><span class="op">)</span> </span>
<span><span class="fu">colnames</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="va">n</span></span></code></pre>
</div>
<p>Run hierarchical clustering on this data with the <code>hclust</code>
function with default parameters to cluster the columns. Create a
dendrogram.<br>
In the dendrogram, which pairs of samples are the furthest away from
each other?<br>
A) 7 and 23<br>
B) 19 and 14<br>
C) 1 and 16<br>
D) 17 and 18</p>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1"> Show me the solution </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" aria-labelledby="headingSolution1" data-bs-parent="#accordionSolution1">
<div class="accordion-body">
<p><code>d &lt;- dist(t(x))</code><br><code>hc &lt;- hclust(d)</code><br><code>mypar()</code><br><code>plot(hc)</code></p>
<p>A hint to determine the disimilarity between two samples is to look
at how hight would be the dendogram be cut in order to have these two
samples in the same group. The higher the cut, the disimilar those
samples are. In this case the cut that would join sample 19 and 14 is
higher than any of the other pairs.</p>
<p>7 and 23 - 141<br>
19 and 14 - 143<br>
1 and 16 - 142<br>
17 and 18 - 142</p>
<p>The answer is B: 19 and 14. The answer might be different due to the
random numbers.</p>
</div>
</div>
</div>
</div>
<div id="exercise-2" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="exercise-2" class="callout-inner">
<h3 class="callout-title">Exercise 2</h3>
<div class="callout-content">
<p>Set the seed at 1, <code>set.seed(1)</code> and replicate the
creation of this matrix 100 times:</p>
<div class="codewrapper sourceCode" id="cb26">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">m</span> <span class="op">=</span> <span class="fl">10000</span></span>
<span><span class="va">n</span> <span class="op">=</span> <span class="fl">24</span></span>
<span><span class="va">x</span> <span class="op">=</span> <span class="fu">matrix</span><span class="op">(</span><span class="fu">rnorm</span><span class="op">(</span><span class="va">m</span> <span class="op">*</span> <span class="va">n</span><span class="op">)</span>, <span class="va">m</span>, <span class="va">n</span><span class="op">)</span></span></code></pre>
</div>
<p>then perform hierarchical clustering as in the solution to exercise
1, and find the number of clusters if you use <code>cuttree</code> at
height 143. This number is a random variable. Based on the Monte Carlo
simulation, what is the standard error of this random variable?</p>
</div>
</div>
</div>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2"> Show me the solution </h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" aria-labelledby="headingSolution2" data-bs-parent="#accordionSolution2">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb27">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">set.seed</span><span class="op">(</span><span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">res_list</span> <span class="op">&lt;-</span> <span class="fu">replicate</span><span class="op">(</span><span class="fl">100</span>, <span class="op">{</span></span>
<span>  <span class="va">m</span> <span class="op">=</span> <span class="fl">10000</span></span>
<span>  <span class="va">n</span> <span class="op">=</span> <span class="fl">24</span></span>
<span>  <span class="co"># Generate a new matrix every repetition.</span></span>
<span>  <span class="va">x</span> <span class="op">=</span> <span class="fu">matrix</span><span class="op">(</span><span class="fu">rnorm</span><span class="op">(</span><span class="va">m</span> <span class="op">*</span> <span class="va">n</span><span class="op">)</span>, <span class="va">m</span>, <span class="va">n</span><span class="op">)</span></span>
<span>  <span class="va">d</span> <span class="op">&lt;-</span> <span class="fu">dist</span><span class="op">(</span><span class="fu">t</span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="co"># Run the hierarchical clustering on the new matrix.</span></span>
<span>  <span class="va">hc</span> <span class="op">&lt;-</span> <span class="fu">hclust</span><span class="op">(</span><span class="va">d</span><span class="op">)</span></span>
<span>  <span class="co"># Compute how many clusters form if we cut the dendogram at a height of 143.</span></span>
<span>  <span class="va">hclusters</span> <span class="op">&lt;-</span> <span class="fu">cutree</span><span class="op">(</span><span class="va">hc</span>, h<span class="op">=</span><span class="fl">143</span><span class="op">)</span></span>
<span>  <span class="va">num_clus</span> <span class="op">&lt;-</span> <span class="fu">length</span><span class="op">(</span><span class="fu">unique</span><span class="op">(</span><span class="va">hclusters</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="kw">return</span><span class="op">(</span><span class="va">num_clus</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span>
<span><span class="co"># Compute the Standard Error of the number of clusters created by cutting</span></span>
<span><span class="co"># the dendograms at a height of 143 on each repetition.</span></span>
<span><span class="fu">popsd</span><span class="op">(</span><span class="va">res_list</span><span class="op">)</span></span></code></pre>
</div>
</div>
</div>
</div>
</div>
<div id="exercise-3" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="exercise-3" class="callout-inner">
<h3 class="callout-title">Exercise 3</h3>
<div class="callout-content">
<p>Run <code>kmeans</code> with 4 centers for the blood RNA data:</p>
<div class="codewrapper sourceCode" id="cb28">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">GSE5859Subset</span><span class="op">)</span></span>
<span><span class="fu">data</span><span class="op">(</span><span class="va">GSE5859Subset</span><span class="op">)</span></span></code></pre>
</div>
<p>Set the seed to 10, <code>set.seed(10)</code> right before running
kmeans with 5 centers. Explore the relationship of clusters and
information in <code>sampleInfo</code>. Which of the following best
describes what you find? A) <code>sampleInfo$group</code> is driving the
clusters as the 0s and 1s are in completely different clusters. B) The
year is driving the clusters. C) <code>Date</code> is driving the
clusters. D) The clusters don’t depend on any of the column of
<code>sampleInfo</code></p>
</div>
</div>
</div>
<div id="accordionSolution3" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution3" aria-expanded="false" aria-controls="collapseSolution3">
  <h4 class="accordion-header" id="headingSolution3"> Show me the solution </h4>
</button>
<div id="collapseSolution3" class="accordion-collapse collapse" aria-labelledby="headingSolution3" data-bs-parent="#accordionSolution3">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb29">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">km</span> <span class="op">&lt;-</span> <span class="fu">kmeans</span><span class="op">(</span><span class="fu">t</span><span class="op">(</span><span class="va">geneExpression</span><span class="op">)</span>, centers <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span>
<span><span class="va">km</span><span class="op">$</span><span class="va">cluster</span></span>
<span><span class="co"># Use the group as the true clusters of our data, and compare it agains the</span></span>
<span><span class="co"># clusters found by k-means.</span></span>
<span><span class="fu">table</span><span class="op">(</span>true <span class="op">=</span> <span class="va">sampleInfo</span><span class="op">$</span><span class="va">group</span>, cluster <span class="op">=</span> <span class="va">km</span><span class="op">$</span><span class="va">cluster</span><span class="op">)</span></span>
<span><span class="co"># Now use the date of acquisition of the samples as the true clusters and</span></span>
<span><span class="co"># compare it against the same clusters found by k-means.</span></span>
<span><span class="fu">table</span><span class="op">(</span>true <span class="op">=</span> <span class="va">sampleInfo</span><span class="op">$</span><span class="va">date</span>, cluster <span class="op">=</span> <span class="va">km</span><span class="op">$</span><span class="va">cluster</span><span class="op">)</span></span></code></pre>
</div>
<p>The answer is C: Date is driving the clusters.</p>
</div>
</div>
</div>
</div>
<div id="exercise-4" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="exercise-4" class="callout-inner">
<h3 class="callout-title">Exercise 4</h3>
<div class="callout-content">
<p>Load the data:</p>
<div class="codewrapper sourceCode" id="cb30">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">GSE5859Subset</span><span class="op">)</span></span>
<span><span class="fu">data</span><span class="op">(</span><span class="va">GSE5859Subset</span><span class="op">)</span></span></code></pre>
</div>
<p>Pick the 25 genes with the highest across sample variance. This
function might help:</p>
<div class="codewrapper sourceCode" id="cb31">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">install.packages</span><span class="op">(</span><span class="st">"matrixStats"</span><span class="op">)</span> </span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">matrixStats</span><span class="op">)</span></span>
<span><span class="op">?</span><span class="va">rowMads</span> <span class="co">## we use mads due to a outlier sample</span></span></code></pre>
</div>
<p>Use <code>heatmap.2</code> to make a heatmap showing the
<code>sampleInfo$group</code> with color, the date as labels, the rows
labelled with chromosome, and scaling the rows. What do we learn from
this heatmap? A) The data appears as if it was generated by
<code>rnorm</code>.<br>
B) Some genes in chr1 are very variable.<br>
C) A group of chrY genes are higher in group 0 and appear to drive the
clustering. Within those clusters there appears to be clustering by
<code>month</code>.<br>
D) A group of chrY genes are higher in October compared to June and
appear to drive the clustering. Within those clusters there appears to
be clustering by <code>samplInfo$group</code>.</p>
</div>
</div>
</div>
<div id="accordionSolution4" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution4" aria-expanded="false" aria-controls="collapseSolution4">
  <h4 class="accordion-header" id="headingSolution4"> Show me the solution </h4>
</button>
<div id="collapseSolution4" class="accordion-collapse collapse" aria-labelledby="headingSolution4" data-bs-parent="#accordionSolution4">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb32">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">hmcol</span> <span class="op">&lt;-</span> <span class="fu">colorRampPalette</span><span class="op">(</span><span class="fu">brewer.pal</span><span class="op">(</span><span class="fl">9</span>, <span class="st">"GnBu"</span><span class="op">)</span><span class="op">)</span><span class="op">(</span><span class="fl">100</span><span class="op">)</span></span>
<span><span class="va">month</span> <span class="op">=</span> <span class="fu">format</span><span class="op">(</span> <span class="va">sampleInfo</span><span class="op">$</span><span class="va">date</span>, <span class="st">"%m"</span><span class="op">)</span></span>
<span><span class="va">rv</span> <span class="op">&lt;-</span> <span class="fu">rowVars</span><span class="op">(</span><span class="va">geneExpression</span><span class="op">)</span></span>
<span><span class="va">idx</span> <span class="op">&lt;-</span> <span class="fu">order</span><span class="op">(</span><span class="op">-</span><span class="va">rv</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">25</span><span class="op">]</span></span>
<span><span class="va">cols</span> <span class="op">&lt;-</span> <span class="fu">palette</span><span class="op">(</span><span class="fu">brewer.pal</span><span class="op">(</span><span class="fl">8</span>, <span class="st">"Dark2"</span><span class="op">)</span><span class="op">)</span><span class="op">[</span><span class="fu">as.fumeric</span><span class="op">(</span><span class="fu">as.character</span><span class="op">(</span><span class="va">sampleInfo</span><span class="op">$</span><span class="va">group</span><span class="op">)</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="co"># Compute the heatmap using the intensity of the gene expressions to color</span></span>
<span><span class="co"># the graph. We are using the months of acquisition of the samples as the</span></span>
<span><span class="co"># columns in our plot, so we can see how the samples arrange according to the date.</span></span>
<span><span class="fu">heatmap.2</span><span class="op">(</span><span class="va">geneExpression</span><span class="op">[</span><span class="va">idx</span>,<span class="op">]</span>, </span>
<span>          trace <span class="op">=</span> <span class="st">'none'</span>, labRow <span class="op">=</span> <span class="va">geneAnnotation</span><span class="op">[</span><span class="va">idx</span>,<span class="op">]</span><span class="op">$</span><span class="va">CHR</span>,</span>
<span>          col <span class="op">=</span> <span class="va">hmcol</span>, labCol <span class="op">=</span> <span class="va">month</span>,</span>
<span>          ColSideColors <span class="op">=</span> <span class="va">cols</span><span class="op">)</span></span></code></pre>
</div>
<p>The correct answer is C: A group of chrY genes are higher in group 0
and appear to drive the clustering. Within those clusters there appears
to be clustering by month. Answer A is wrong because if the data were
generated by <code>norm</code>, the color distribution of the heatmap
would be entirely random. Answer B is wrong because the colors in the
row chr1 are more or less the same except for one column (one sample).
Answer D is wrong chrY genes are higher in June, not October.</p>
</div>
</div>
</div>
</div>
<div id="exercise-5" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="exercise-5" class="callout-inner">
<h3 class="callout-title">Exercise 5</h3>
<div class="callout-content">
<p>Create a large dataset of random data that is completely independent
of <code>sampleInfo$group</code> like this:</p>
<div class="codewrapper sourceCode" id="cb33">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">set.seed</span><span class="op">(</span><span class="fl">17</span><span class="op">)</span></span>
<span><span class="va">m</span> <span class="op">=</span> <span class="fu">nrow</span><span class="op">(</span><span class="va">geneExpression</span><span class="op">)</span> </span>
<span><span class="va">n</span> <span class="op">=</span> <span class="fu">ncol</span><span class="op">(</span><span class="va">geneExpression</span><span class="op">)</span> </span>
<span><span class="va">x</span> <span class="op">=</span> <span class="fu">matrix</span><span class="op">(</span><span class="fu">rnorm</span><span class="op">(</span><span class="va">m</span> <span class="op">*</span> <span class="va">n</span><span class="op">)</span>, <span class="va">m</span>, <span class="va">n</span><span class="op">)</span> </span>
<span><span class="va">g</span> <span class="op">=</span> <span class="fu">factor</span><span class="op">(</span><span class="va">sampleInfo</span><span class="op">$</span><span class="va">g</span><span class="op">)</span></span></code></pre>
</div>
<p>Create two heatmaps with these data. Show the group <code>g</code>
either with labels or colors. First, take the 50 genes with smallest
p-values obtained with <code>rowttests</code>. Then, take the 50 genes
with largest standard deviations. Which of the following statements is
true?<br>
A) There is no relationship between <code>g</code> and <code>x</code>,
but with 8,793 tests some will appear significant by chance. Selecting
genes with the t-test gives us a deceiving result.<br>
B) These two techniques produced similar heatmaps.<br>
C) Selecting genes with the t-test is a better technique since it
permits us to detect the two groups. It appears to find hidden
signals.<br>
D) The genes with the largest standard deviation add variability to the
plot and do not let us find the differences between the two groups.</p>
</div>
</div>
</div>
<div id="accordionSolution5" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution5" aria-expanded="false" aria-controls="collapseSolution5">
  <h4 class="accordion-header" id="headingSolution5"> Show me the solution </h4>
</button>
<div id="collapseSolution5" class="accordion-collapse collapse" aria-labelledby="headingSolution5" data-bs-parent="#accordionSolution5">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb34">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">genefilter</span><span class="op">)</span></span>
<span><span class="co"># p-value</span></span>
<span><span class="co"># Perform a t-test on the expression of each gene to determine if they have</span></span>
<span><span class="co"># an effect different from 0.</span></span>
<span><span class="va">pvals</span> <span class="op">&lt;-</span> <span class="fu">rowttests</span><span class="op">(</span><span class="va">x</span>, <span class="va">g</span><span class="op">)</span><span class="op">$</span><span class="va">p.value</span></span>
<span><span class="co"># Select the top 50 genes that show the most statistical significance of their</span></span>
<span><span class="co"># expression, obtained by the t-tests.</span></span>
<span><span class="va">idx</span> <span class="op">&lt;-</span> <span class="fu">order</span><span class="op">(</span><span class="va">pvals</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">50</span><span class="op">]</span></span>
<span><span class="va">cols</span> <span class="op">&lt;-</span> <span class="fu">palette</span><span class="op">(</span><span class="fu">brewer.pal</span><span class="op">(</span><span class="fl">8</span>, <span class="st">"Dark2"</span><span class="op">)</span><span class="op">)</span><span class="op">[</span><span class="fu">as.fumeric</span><span class="op">(</span><span class="fu">as.character</span><span class="op">(</span><span class="va">sampleInfo</span><span class="op">$</span><span class="va">g</span><span class="op">)</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="fu">heatmap.2</span><span class="op">(</span><span class="va">x</span><span class="op">[</span><span class="va">idx</span>,<span class="op">]</span>, </span>
<span>          trace <span class="op">=</span> <span class="st">'none'</span>, labRow <span class="op">=</span> <span class="va">geneAnnotation</span><span class="op">[</span><span class="va">idx</span>,<span class="op">]</span><span class="op">$</span><span class="va">CHR</span>,</span>
<span>          col <span class="op">=</span> <span class="va">hmcol</span>, labCol <span class="op">=</span> <span class="va">month</span>,</span>
<span>          ColSideColors <span class="op">=</span> <span class="va">cols</span><span class="op">)</span></span>
<span><span class="co"># std dev</span></span>
<span><span class="co"># Compute the standard deviation of the expression of each gene.</span></span>
<span><span class="va">sds</span> <span class="op">&lt;-</span> <span class="fu">rowSds</span><span class="op">(</span><span class="va">x</span>,<span class="va">g</span><span class="op">)</span></span>
<span><span class="co"># Select the top 50 genes that show a higher variance of their expression</span></span>
<span><span class="co"># acoss all the samples.</span></span>
<span><span class="va">idx</span> <span class="op">&lt;-</span> <span class="fu">order</span><span class="op">(</span><span class="op">-</span><span class="va">sds</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">50</span><span class="op">]</span></span>
<span><span class="va">cols</span> <span class="op">&lt;-</span> <span class="fu">palette</span><span class="op">(</span><span class="fu">brewer.pal</span><span class="op">(</span><span class="fl">8</span>, <span class="st">"Dark2"</span><span class="op">)</span><span class="op">)</span><span class="op">[</span><span class="fu">as.fumeric</span><span class="op">(</span><span class="fu">as.character</span><span class="op">(</span><span class="va">sampleInfo</span><span class="op">$</span><span class="va">g</span><span class="op">)</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="fu">heatmap.2</span><span class="op">(</span><span class="va">x</span><span class="op">[</span><span class="va">idx</span>,<span class="op">]</span>, </span>
<span>          trace <span class="op">=</span> <span class="st">'none'</span>, labRow <span class="op">=</span> <span class="va">geneAnnotation</span><span class="op">[</span><span class="va">idx</span>,<span class="op">]</span><span class="op">$</span><span class="va">CHR</span>,</span>
<span>          col <span class="op">=</span> <span class="va">hmcol</span>, labCol <span class="op">=</span> <span class="va">month</span>,</span>
<span>          ColSideColors <span class="op">=</span> <span class="va">cols</span><span class="op">)</span></span></code></pre>
</div>
<p>The answer is A: There is no relationship between g and x, but with
8,793 tests some will appear significant by chance. Selecting genes with
the t-test gives us a deceiving result.</p>
<p>Recall that we have already selected smallest p-values from a dataset
in which the null hypothesis is true. Therefore, we can see clusters
that indicate that there is a significant difference between sample
groups. However, this significance is not real because we know that the
null hypothesis is true.</p>
</div>
</div>
</div>
</div>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Key Points</h3>
<div class="callout-content">
<ul>
<li>We can use Euclidean distance to define the dissimilarity between
samples.</li>
<li>We can also use other metrics according to the prior knowledge we
have from our data.</li>
</ul>
</div>
</div>
</div>
<!--
Place links that you need to refer to multiple times across pages here. Delete
any links that you are not going to use.
 -->
</div>
</section></section><section id="aio-conditional-probabilities-expectations"><p>Content from <a href="conditional-probabilities-expectations.html">Conditional Probabilities and Expectations</a></p>
<hr>
<p>Last updated on 2025-07-18 |

        <a href="https://github.com/smcclatchy/machine-learning-biomedical-science/edit/main/episodes/conditional-probabilities-expectations.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<p>Estimated time: <i aria-hidden="true" data-feather="clock"></i> 60 minutes</p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>How do we get statistical information from specific subsets in our
data?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Calculate statistics from subsets of a dataset that satisfies
specific conditions.</li>
<li>Predict the expected outcome of an experiment given certain
conditions on a dataset.</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<section><h2 class="section-heading" id="conditional-probabilities-and-expectations">Conditional Probabilities and Expectations<a class="anchor" aria-label="anchor" href="#conditional-probabilities-and-expectations"></a>
</h2>
<hr class="half-width">
<p>Prediction problems can be divided into categorical and continuous
outcomes. However, many of the algorithms can be applied to both due to
the connection between <em>conditional probabilities</em> and
<em>conditional expectations</em>.</p>
<p>For categorical data, for example binary outcomes, if we know the
probability of <span class="math inline">\(Y\)</span> being any of the
possible outcomes <span class="math inline">\(k\)</span> given a set of
predictors <span class="math inline">\(X=(X_1,\dots,X_p)^\top\)</span>,</p>
<p><span class="math display">\[
f_k(x) = \mbox{Pr}(Y=k \mid X=x)
\]</span></p>
<p>we can optimize our predictions. Specifically, for any <span class="math inline">\(x\)</span> we predict the <span class="math inline">\(k\)</span> that has the largest probability <span class="math inline">\(f_k(x)\)</span>.</p>
<p>To simplify the exposition below, we will consider the case of binary
data. You can think of the probability <span class="math inline">\(\mbox{Pr}(Y=1 \mid X=x)\)</span> as the proportion
of 1s in the stratum of the population for which <span class="math inline">\(X=x\)</span>. Given that the expectation is the
average of all <span class="math inline">\(Y\)</span> values, in this
case the expectation is equivalent to the probability: <span class="math inline">\(f(x) \equiv \mbox{E}(Y \mid X=x)=\mbox{Pr}(Y=1
\mid X=x)\)</span>. We therefore use only the expectation in the
descriptions below as it is more general.</p>
<p>In general, the expected value has an attractive mathematical
property, which is that it minimizes the expected distance between the
predictor <span class="math inline">\(\hat{Y}\)</span> and <span class="math inline">\(Y\)</span>:</p>
<p><span class="math display">\[
\mbox{E}\{ (\hat{Y} - Y)^2  \mid  X=x \}
\]</span></p>
<div class="section level4">
<h4 id="regression-in-the-context-of-prediction">Regression in the context of prediction<a class="anchor" aria-label="anchor" href="#regression-in-the-context-of-prediction"></a>
</h4>
<p><a name="regression"></a></p>
<p>We use the son and father height example to illustrate how regression
can be interpreted as a machine learning technique. In our example, we
are trying to predict the son’s height <span class="math inline">\(Y\)</span> based on the father’s <span class="math inline">\(X\)</span>. Here we have only one predictor. Now
if we were asked to predict the height of a randomly selected son, we
would go with the average height:</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">rafalib</span><span class="op">)</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">UsingR</span><span class="op">)</span></span>
<span><span class="fu">mypar</span><span class="op">(</span><span class="fl">1</span>, <span class="fl">1</span><span class="op">)</span></span>
<span><span class="fu">data</span><span class="op">(</span><span class="va">father.son</span>, package<span class="op">=</span><span class="st">"UsingR"</span><span class="op">)</span></span>
<span><span class="va">x</span><span class="op">=</span><span class="fu">round</span><span class="op">(</span><span class="va">father.son</span><span class="op">$</span><span class="va">fheight</span><span class="op">)</span> <span class="co">## round to nearest inch</span></span>
<span><span class="va">y</span><span class="op">=</span><span class="fu">round</span><span class="op">(</span><span class="va">father.son</span><span class="op">$</span><span class="va">sheight</span><span class="op">)</span></span>
<span><span class="fu">hist</span><span class="op">(</span><span class="va">y</span>, breaks<span class="op">=</span><span class="fu">seq</span><span class="op">(</span><span class="fu">min</span><span class="op">(</span><span class="va">y</span><span class="op">)</span>, <span class="fu">max</span><span class="op">(</span><span class="va">y</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">abline</span><span class="op">(</span>v<span class="op">=</span><span class="fu">mean</span><span class="op">(</span><span class="va">y</span><span class="op">)</span>, col<span class="op">=</span><span class="st">"red"</span>, lwd<span class="op">=</span><span class="fl">2</span><span class="op">)</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="../fig/conditional-probabilities-expectations-rendered-height_hist-1.png" alt="Histogram of son heights." class="figure mx-auto d-block"><figcaption>
Histogram of son heights.
</figcaption></figure><p>In this case, we can also approximate the distribution of <span class="math inline">\(Y\)</span> as normal, which implies the mean
maximizes the probability density.</p>
<p>Let’s imagine that we are given more information. We are told that
the father of this randomly selected son has a height of 71 inches (1.25
SDs taller than the average). What is our prediction now?</p>
<div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">mypar</span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span><span class="op">)</span></span>
<span><span class="fu">plot</span><span class="op">(</span><span class="va">x</span>, <span class="va">y</span>, </span>
<span>     xlab<span class="op">=</span><span class="st">"Father's height in inches"</span>,</span>
<span>     ylab<span class="op">=</span><span class="st">"Son's height in inches"</span>,</span>
<span>     main<span class="op">=</span><span class="fu">paste</span><span class="op">(</span><span class="st">"correlation ="</span>,</span>
<span>                <span class="fu">signif</span><span class="op">(</span><span class="fu">cor</span><span class="op">(</span><span class="va">x</span>, <span class="va">y</span><span class="op">)</span>, <span class="fl">2</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">abline</span><span class="op">(</span>v<span class="op">=</span><span class="fu">c</span><span class="op">(</span><span class="op">-</span><span class="fl">0.35</span>, <span class="fl">0.35</span><span class="op">)</span> <span class="op">+</span> <span class="fl">71</span>,</span>
<span>       col<span class="op">=</span><span class="st">"red"</span><span class="op">)</span></span>
<span><span class="fu">hist</span><span class="op">(</span><span class="va">y</span><span class="op">[</span><span class="va">x</span><span class="op">==</span><span class="fl">71</span><span class="op">]</span>, </span>
<span>     xlab<span class="op">=</span><span class="st">"Heights"</span>, </span>
<span>     nc<span class="op">=</span><span class="fl">8</span>, </span>
<span>     main<span class="op">=</span><span class="st">""</span>, </span>
<span>     xlim<span class="op">=</span><span class="fu">range</span><span class="op">(</span><span class="va">y</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="../fig/conditional-probabilities-expectations-rendered-conditional_distribution-1.png" alt="Son versus father height (left) with the red lines denoting the stratum defined by conditioning on fathers being 71 inches tall. Conditional distribution: son height distribution of stratum defined by 71 inch fathers." class="figure mx-auto d-block"><figcaption>
Son versus father height (left) with the red lines denoting the stratum
defined by conditioning on fathers being 71 inches tall. Conditional
distribution: son height distribution of stratum defined by 71 inch
fathers.
</figcaption></figure><p>The best guess is still the expectation, but our strata has changed
from all the data, to only the <span class="math inline">\(Y\)</span>
with <span class="math inline">\(X=71\)</span>. So we can stratify and
take the average, which is the conditional expectation. Our prediction
for any <span class="math inline">\(x\)</span> is therefore:</p>
<p><span class="math display">\[
f(x) = E(Y \mid X=x)
\]</span></p>
<p>It turns out that because this data is approximated by a bivariate
normal distribution, using calculus, we can show that:</p>
<p><span class="math display">\[
f(x) = \mu_Y + \rho \frac{\sigma_Y}{\sigma_X} (X-\mu_X)
\]</span></p>
<p>and if we estimate these five parameters from the sample, we get the
regression line:</p>
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">mypar</span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span><span class="op">)</span></span>
<span><span class="fu">plot</span><span class="op">(</span><span class="va">x</span>, <span class="va">y</span>, </span>
<span>     xlab<span class="op">=</span><span class="st">"Father's height in inches"</span>, </span>
<span>     ylab<span class="op">=</span><span class="st">"Son's height in inches"</span>,</span>
<span>     main<span class="op">=</span><span class="fu">paste</span><span class="op">(</span><span class="st">"correlation ="</span>,</span>
<span>                <span class="fu">signif</span><span class="op">(</span><span class="fu">cor</span><span class="op">(</span><span class="va">x</span>,<span class="va">y</span><span class="op">)</span>,<span class="fl">2</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">abline</span><span class="op">(</span>v<span class="op">=</span><span class="fu">c</span><span class="op">(</span><span class="op">-</span><span class="fl">0.35</span>, <span class="fl">0.35</span><span class="op">)</span> <span class="op">+</span> <span class="fl">71</span>,</span>
<span>       col<span class="op">=</span><span class="st">"red"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">fit</span> <span class="op">&lt;-</span> <span class="fu">lm</span><span class="op">(</span><span class="va">y</span> <span class="op">~</span> <span class="va">x</span><span class="op">)</span></span>
<span><span class="fu">abline</span><span class="op">(</span><span class="va">fit</span>, col<span class="op">=</span><span class="fl">1</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">hist</span><span class="op">(</span><span class="va">y</span><span class="op">[</span><span class="va">x</span><span class="op">==</span><span class="fl">71</span><span class="op">]</span>,</span>
<span>     xlab<span class="op">=</span><span class="st">"Heights"</span>,</span>
<span>     nc<span class="op">=</span><span class="fl">8</span>,</span>
<span>     main<span class="op">=</span><span class="st">""</span>,</span>
<span>     xlim<span class="op">=</span><span class="fu">range</span><span class="op">(</span><span class="va">y</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">abline</span><span class="op">(</span>v <span class="op">=</span> <span class="va">fit</span><span class="op">$</span><span class="va">coef</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span> <span class="op">+</span> <span class="va">fit</span><span class="op">$</span><span class="va">coef</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span><span class="op">*</span><span class="fl">71</span>, col<span class="op">=</span><span class="fl">1</span><span class="op">)</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="../fig/conditional-probabilities-expectations-rendered-regression-1.png" alt="Son versus father height showing predicted heights based on regression line (left). Conditional distribution with vertical line representing regression prediction." class="figure mx-auto d-block"><figcaption>
Son versus father height showing predicted heights based on regression
line (left). Conditional distribution with vertical line representing
regression prediction.
</figcaption></figure><p>In this particular case, the regression line provides an optimal
prediction function for <span class="math inline">\(Y\)</span>. But this
is not generally true because, in the typical machine learning problems,
the optimal <span class="math inline">\(f(x)\)</span> is rarely a simple
line.</p>
<div id="exercise-1" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="exercise-1" class="callout-inner">
<h3 class="callout-title">Exercise 1</h3>
<div class="callout-content">
<p>Throughout these exercises it will be useful to remember that when
our data are 0s and 1s, probabilities and expectations are the same
thing. We can do the math, but here is some R code:</p>
<div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">n</span> <span class="op">=</span> <span class="fl">1000</span></span>
<span><span class="va">y</span> <span class="op">=</span> <span class="fu">rbinom</span><span class="op">(</span><span class="va">n</span>, <span class="fl">1</span>, <span class="fl">0.25</span><span class="op">)</span> <span class="co">## proportion of ones Pr(Y)</span></span>
<span><span class="fu">sum</span><span class="op">(</span><span class="va">y</span><span class="op">==</span><span class="fl">1</span><span class="op">)</span><span class="op">/</span><span class="fu">length</span><span class="op">(</span><span class="va">y</span><span class="op">)</span> <span class="co">## expectation of Y mean(y)</span></span></code></pre>
</div>
<p>Generate some random data to imitate heights for men (0) and women
(1):</p>
<div class="codewrapper sourceCode" id="cb5">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">n</span> <span class="op">=</span> <span class="fl">10000</span></span>
<span><span class="fu">set.seed</span><span class="op">(</span><span class="fl">1</span><span class="op">)</span></span>
<span><span class="co"># Generate a sample of heights for a mixed population of men and women</span></span>
<span><span class="va">men</span> <span class="op">=</span> <span class="fu">rnorm</span><span class="op">(</span><span class="va">n</span>, <span class="fl">176</span>, <span class="fl">7</span><span class="op">)</span> <span class="co"># height in centimeters </span></span>
<span><span class="va">women</span> <span class="op">=</span> <span class="fu">rnorm</span><span class="op">(</span><span class="va">n</span>, <span class="fl">162</span>, <span class="fl">7</span><span class="op">)</span> <span class="co"># height in centimeters</span></span>
<span><span class="co"># Assign a class label to each height generated above (0: men, 1:women)</span></span>
<span><span class="va">y</span> <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="fu">rep</span><span class="op">(</span><span class="fl">0</span>, <span class="va">n</span><span class="op">)</span>, <span class="fu">rep</span><span class="op">(</span><span class="fl">1</span>, <span class="va">n</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">x</span> <span class="op">=</span> <span class="fu">round</span><span class="op">(</span><span class="fu">c</span><span class="op">(</span><span class="va">men</span>, <span class="va">women</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">## mix it up</span></span>
<span><span class="va">ind</span> <span class="op">=</span> <span class="fu">sample</span><span class="op">(</span><span class="fu">seq</span><span class="op">(</span>along<span class="op">=</span><span class="va">y</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">y</span> <span class="op">=</span> <span class="va">y</span><span class="op">[</span><span class="va">ind</span><span class="op">]</span></span>
<span><span class="va">x</span> <span class="op">=</span> <span class="va">x</span><span class="op">[</span><span class="va">ind</span><span class="op">]</span></span></code></pre>
</div>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1"> Show me the solution </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" data-bs-parent="#accordionSolution1" aria-labelledby="headingSolution1">
<div class="accordion-body">
<p>NA</p>
</div>
</div>
</div>
</div>
<div id="exercise-2" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="exercise-2" class="callout-inner">
<h3 class="callout-title">Exercise 2</h3>
<div class="callout-content">
<p>Using the data generated above, what is the E(Y | X = 176)?</p>
</div>
</div>
</div>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2"> Show me the solution </h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" data-bs-parent="#accordionSolution2" aria-labelledby="headingSolution2">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb6">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Take all labels `y` that correspond to individuals whose height is 176 cm.</span></span>
<span><span class="co"># Then compute the mean value as an insight about an individual with</span></span>
<span><span class="co"># that height could be a woman or a man.</span></span>
<span><span class="fu">mean</span><span class="op">(</span><span class="va">y</span><span class="op">[</span><span class="va">x</span><span class="op">==</span><span class="fl">176</span><span class="op">]</span><span class="op">)</span></span></code></pre>
</div>
</div>
</div>
</div>
</div>
<div id="exercise-3" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="exercise-3" class="callout-inner">
<h3 class="callout-title">Exercise 3</h3>
<div class="callout-content">
<p>Now make a plot of E(Y|X=x) for <code>x=seq(160, 178)</code> using
the data generated in exercise 1. If you are predicting female or male
based on height and want your probability of success to be larger than
0.5, what is the largest height where you predict female ?</p>
</div>
</div>
</div>
<div id="accordionSolution3" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution3" aria-expanded="false" aria-controls="collapseSolution3">
  <h4 class="accordion-header" id="headingSolution3"> Show me the solution </h4>
</button>
<div id="collapseSolution3" class="accordion-collapse collapse" data-bs-parent="#accordionSolution3" aria-labelledby="headingSolution3">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb7">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">mypar</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu">plot</span><span class="op">(</span><span class="va">x</span>,<span class="va">y</span><span class="op">)</span></span>
<span><span class="co"># This time we test a single height at a time from a set of heights,</span></span>
<span><span class="co"># between 160 cm and 178 cm.</span></span>
<span><span class="va">x_list</span> <span class="op">&lt;-</span> <span class="fu">seq</span><span class="op">(</span><span class="fl">160</span>,<span class="fl">178</span><span class="op">)</span></span>
<span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu">vector</span><span class="op">(</span><span class="st">'double'</span>, <span class="fu">length</span><span class="op">(</span><span class="va">x_list</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fu">seq_along</span><span class="op">(</span><span class="va">x_list</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="co"># Compute the expected label (woman or man) of an individual whose height</span></span>
<span>  <span class="co"># is the same as x_list[i]</span></span>
<span>  <span class="va">res</span><span class="op">[</span><span class="va">i</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu">mean</span><span class="op">(</span><span class="va">y</span><span class="op">[</span><span class="va">x</span><span class="op">==</span><span class="va">x_list</span><span class="op">[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="co"># Verify that the probability to identify an individual as a woman given their</span></span>
<span><span class="co"># height is higher than 50 %</span></span>
<span><span class="fu">mean</span><span class="op">(</span><span class="va">y</span><span class="op">[</span><span class="va">x</span><span class="op">==</span><span class="va">x_list</span><span class="op">[</span><span class="va">ind</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># And if we move one centimeter higher, our probability falls under 50 %</span></span>
<span><span class="fu">mean</span><span class="op">(</span><span class="va">y</span><span class="op">[</span><span class="va">x</span><span class="op">==</span><span class="va">x_list</span><span class="op">[</span><span class="va">ind</span> <span class="op">+</span> <span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span></code></pre>
</div>
</div>
</div>
</div>
</div>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Key Points</h3>
<div class="callout-content">
<ul>
<li>For categorical/discrete variables we have used strict conditions
(i.e. X=x); however, conditioning can be applied to continuous variables
by using ranges instead (e.g. X=x, X&lt;=x, or a&lt;X&lt;b)</li>
</ul>
</div>
</div>
</div>
<!--
Place links that you need to refer to multiple times across pages here. Delete
any links that you are not going to use.
 -->
</div>
</section></section><section id="aio-smoothing"><p>Content from <a href="smoothing.html">Smoothing</a></p>
<hr>
<p>Last updated on 2025-07-18 |

        <a href="https://github.com/smcclatchy/machine-learning-biomedical-science/edit/main/episodes/smoothing.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<p>Estimated time: <i aria-hidden="true" data-feather="clock"></i> 60 minutes</p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>Can a model be fitted to a dataset which shape is unknown but
smooth?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Fit a smooth regression model to data which behavior depends
conditionally on a set of predictors.</li>
<li>Predict the expected value of a smooth model given the value of the
predictors.</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<section><h2 class="section-heading" id="smoothing">Smoothing<a class="anchor" aria-label="anchor" href="#smoothing"></a>
</h2>
<hr class="half-width">
<p>Smoothing is a very powerful technique used all across data analysis.
It is designed to estimate <span class="math inline">\(f(x)\)</span>
when the shape is unknown, but assumed to be <em>smooth</em>. The
general idea is to group data points that are expected to have similar
expectations and compute the average, or fit a simple parametric model.
We illustrate two smoothing techniques using a gene expression
example.</p>
<p>The following data are gene expression measurements from replicated
RNA samples.</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Following three packages are available from Bioconductor</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">Biobase</span><span class="op">)</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">SpikeIn</span><span class="op">)</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">hgu95acdf</span><span class="op">)</span></span>
<span><span class="fu">data</span><span class="op">(</span><span class="va">SpikeIn95</span><span class="op">)</span></span></code></pre>
</div>
<p>We consider the data used in an MA-plot comparing two replicated
samples (<span class="math inline">\(Y\)</span> = log ratios and <span class="math inline">\(X\)</span> = averages) and take down-sample in a
way that balances the number of points for different strata of <span class="math inline">\(X\)</span> (code not shown):</p>
<div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">rafalib</span><span class="op">)</span></span>
<span><span class="fu">mypar</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu">plot</span><span class="op">(</span><span class="va">X</span>,<span class="va">Y</span><span class="op">)</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="../fig/smoothing-rendered-MAplot-1.png" alt="MA-plot comparing gene expression from two arrays." class="figure mx-auto d-block"><figcaption>
MA-plot comparing gene expression from two arrays.
</figcaption></figure><p>In the MA plot we see that <span class="math inline">\(Y\)</span>
depends on <span class="math inline">\(X\)</span>. This dependence must
be a bias because these are based on replicates, which means <span class="math inline">\(Y\)</span> should be 0 on average regardless of
<span class="math inline">\(X\)</span>. We want to predict <span class="math inline">\(f(x)=\mbox{E}(Y \mid X=x)\)</span> so that we can
remove this bias. Linear regression does not capture the apparent
curvature in <span class="math inline">\(f(x)\)</span>:</p>
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">mypar</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu">plot</span><span class="op">(</span><span class="va">X</span>,<span class="va">Y</span><span class="op">)</span></span>
<span><span class="va">fit</span> <span class="op">&lt;-</span> <span class="fu">lm</span><span class="op">(</span><span class="va">Y</span><span class="op">~</span><span class="va">X</span><span class="op">)</span></span>
<span><span class="fu">points</span><span class="op">(</span><span class="va">X</span>,<span class="va">Y</span>,pch<span class="op">=</span><span class="fl">21</span>,bg<span class="op">=</span><span class="fu">ifelse</span><span class="op">(</span><span class="va">Y</span><span class="op">&gt;</span><span class="va">fit</span><span class="op">$</span><span class="va">fitted</span>,<span class="fl">1</span>,<span class="fl">3</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">abline</span><span class="op">(</span><span class="va">fit</span>,col<span class="op">=</span><span class="fl">2</span>,lwd<span class="op">=</span><span class="fl">4</span>,lty<span class="op">=</span><span class="fl">2</span><span class="op">)</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="../fig/smoothing-rendered-MAplot_with_regression_line-1.png" alt="MA-plot comparing gene expression from two arrays with fitted regression line. The two colors represent positive and negative residuals." class="figure mx-auto d-block"><figcaption>
MA-plot comparing gene expression from two arrays with fitted regression
line. The two colors represent positive and negative residuals.
</figcaption></figure><p>The points above the fitted line (green) and those below (purple) are
not evenly distributed. We therefore need an alternative more flexible
approach.</p>
</section><section><h2 class="section-heading" id="bin-smoothing">Bin Smoothing<a class="anchor" aria-label="anchor" href="#bin-smoothing"></a>
</h2>
<hr class="half-width">
<p>Instead of fitting a line, let’s go back to the idea of stratifying
and computing the mean. This is referred to as <em>bin smoothing</em>.
The general idea is that the underlying curve is “smooth” enough so
that, in small bins, the curve is approximately constant. If we assume
the curve is constant, then all the <span class="math inline">\(Y\)</span> in that bin have the same expected
value. For example, in the plot below, we highlight points in a bin
centered at 8.6, as well as the points of a bin centered at 12.1, if we
use bins of size 1. We also show the fitted mean values for the <span class="math inline">\(Y\)</span> in those bins with dashed lines (code
not shown):</p>
<figure style="text-align: center"><img src="../fig/smoothing-rendered-binsmoother-1.png" alt="MAplot comparing gene expression from two arrays with bin smoother fit shown for two points." class="figure mx-auto d-block"><figcaption>
MAplot comparing gene expression from two arrays with bin smoother fit
shown for two points.
</figcaption></figure><p>By computing this mean for bins around every point, we form an
estimate of the underlying curve <span class="math inline">\(f(x)\)</span>. Below we show the procedure
happening as we move from the smallest value of <span class="math inline">\(x\)</span> to the largest. We show 10 intermediate
cases as well (code not shown):</p>
<figure style="text-align: center"><img src="../fig/smoothing-rendered-bin_smoothing_demo-1.png" alt="Illustration of how bin smoothing estimates a curve. Showing 12 steps of process." class="figure mx-auto d-block"><figcaption>
Illustration of how bin smoothing estimates a curve. Showing 12 steps of
process.
</figcaption></figure><p>The final result looks like this (code not shown):</p>
<figure style="text-align: center"><img src="../fig/smoothing-rendered-bin_smooth_final-1.png" alt="MA-plot with curve obtained with bin-smoothed curve shown." class="figure mx-auto d-block"><figcaption>
MA-plot with curve obtained with bin-smoothed curve shown.
</figcaption></figure><p>There are several functions in R that implement bin smoothers. One
example is <code>ksmooth</code>. However, in practice, we typically
prefer methods that use slightly more complicated models than fitting a
constant. The final result above, for example, is somewhat wiggly.
Methods such as <code>loess</code>, which we explain next, improve on
this.</p>
</section><section><h2 class="section-heading" id="loess">Loess<a class="anchor" aria-label="anchor" href="#loess"></a>
</h2>
<hr class="half-width">
<p>Local weighted regression (loess) is similar to bin smoothing in
principle. The main difference is that we approximate the local behavior
with a line or a parabola. This permits us to expand the bin sizes,
which stabilizes the estimates. Below we see lines fitted to two bins
that are slightly larger than those we used for the bin smoother (code
not shown). We can use larger bins because fitting lines provide
slightly more flexibility.</p>
<figure style="text-align: center"><img src="../fig/smoothing-rendered-loess-1.png" alt="MA-plot comparing gene expression from two arrays with bin local regression fit shown for two points." class="figure mx-auto d-block"><figcaption>
MA-plot comparing gene expression from two arrays with bin local
regression fit shown for two points.
</figcaption></figure><p>As we did for the bin smoother, we show 12 steps of the process that
leads to a loess fit (code not shown):</p>
<figure style="text-align: center"><img src="../fig/smoothing-rendered-loess_demo-1.png" alt="Illustration of how loess estimates a curve. Showing 12 steps of the process." class="figure mx-auto d-block"><figcaption>
Illustration of how loess estimates a curve. Showing 12 steps of the
process.
</figcaption></figure><p>The final result is a smoother fit than the bin smoother since we use
larger sample sizes to estimate our local parameters (code not
shown):</p>
<figure style="text-align: center"><img src="../fig/smoothing-rendered-loess_final-1.png" alt="MA-plot with curve obtained with loess." class="figure mx-auto d-block"><figcaption>
MA-plot with curve obtained with loess.
</figcaption></figure><p>The function <code>loess</code> performs this analysis for us:</p>
<div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fit</span> <span class="op">&lt;-</span> <span class="fu">loess</span><span class="op">(</span><span class="va">Y</span><span class="op">~</span><span class="va">X</span>, degree<span class="op">=</span><span class="fl">1</span>, span<span class="op">=</span><span class="fl">1</span><span class="op">/</span><span class="fl">3</span><span class="op">)</span></span>
<span></span>
<span><span class="va">newx</span> <span class="op">&lt;-</span> <span class="fu">seq</span><span class="op">(</span><span class="fu">min</span><span class="op">(</span><span class="va">X</span><span class="op">)</span>,<span class="fu">max</span><span class="op">(</span><span class="va">X</span><span class="op">)</span>,len<span class="op">=</span><span class="fl">100</span><span class="op">)</span> </span>
<span><span class="va">smooth</span> <span class="op">&lt;-</span> <span class="fu">predict</span><span class="op">(</span><span class="va">fit</span>,newdata<span class="op">=</span><span class="fu">data.frame</span><span class="op">(</span>X<span class="op">=</span><span class="va">newx</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">mypar</span> <span class="op">(</span><span class="op">)</span></span>
<span><span class="fu">plot</span><span class="op">(</span><span class="va">X</span>,<span class="va">Y</span>,col<span class="op">=</span><span class="st">"darkgrey"</span>,pch<span class="op">=</span><span class="fl">16</span><span class="op">)</span></span>
<span><span class="fu">lines</span><span class="op">(</span><span class="va">newx</span>,<span class="va">smooth</span>,col<span class="op">=</span><span class="st">"black"</span>,lwd<span class="op">=</span><span class="fl">3</span><span class="op">)</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="../fig/smoothing-rendered-loess2-1.png" alt="Loess fitted with the loess function." class="figure mx-auto d-block"><figcaption>
Loess fitted with the loess function.
</figcaption></figure><p>There are three other important differences between
<code>loess</code> and the typical bin smoother. The first is that
rather than keeping the bin size the same, <code>loess</code> keeps the
number of points used in the local fit the same. This number is
controlled via the <code>span</code> argument which expects a
proportion. For example, if <code>N</code> is the number of data points
and <code>span=0.5</code>, then for a given <span class="math inline">\(x\)</span> , <code>loess</code> will use the
<code>0.5*N</code> closest points to <span class="math inline">\(x\)</span> for the fit. The second difference is
that, when fitting the parametric model to obtain <span class="math inline">\(f(x)\)</span>, <code>loess</code> uses weighted
least squares, with higher weights for points that are closer to <span class="math inline">\(x\)</span>. The third difference is that
<code>loess</code> has the option of fitting the local model robustly.
An iterative algorithm is implemented in which, after fitting a model in
one iteration, outliers are detected and downweighted for the next
iteration. To use this option, we use the argument
<code>family="symmetric"</code>.</p>
<div id="exercise-1" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="exercise-1" class="callout-inner">
<h3 class="callout-title">Exercise 1</h3>
<div class="callout-content">
<p>Generate the following data:</p>
<div class="codewrapper sourceCode" id="cb5">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">n</span> <span class="op">=</span> <span class="fl">10000</span></span>
<span><span class="fu">set.seed</span><span class="op">(</span><span class="fl">1</span><span class="op">)</span></span>
<span><span class="co"># Generate a sample of heights for a mixed population of men and women</span></span>
<span><span class="va">men</span> <span class="op">=</span> <span class="fu">rnorm</span><span class="op">(</span><span class="va">n</span>,<span class="fl">176</span>,<span class="fl">7</span><span class="op">)</span> <span class="co">#height in centimeters</span></span>
<span><span class="va">women</span> <span class="op">=</span> <span class="fu">rnorm</span><span class="op">(</span><span class="va">n</span>,<span class="fl">162</span>,<span class="fl">7</span><span class="op">)</span> <span class="co">#height in centimeters</span></span>
<span><span class="co"># Assign a class label to each height generated above (0: men, 1:women)</span></span>
<span><span class="va">y</span> <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="fu">rep</span><span class="op">(</span><span class="fl">0</span>,<span class="va">n</span><span class="op">)</span>,<span class="fu">rep</span><span class="op">(</span><span class="fl">1</span>,<span class="va">n</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">x</span> <span class="op">=</span> <span class="fu">round</span><span class="op">(</span><span class="fu">c</span><span class="op">(</span><span class="va">men</span>,<span class="va">women</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">## mix it up</span></span>
<span><span class="va">ind</span> <span class="op">=</span> <span class="fu">sample</span><span class="op">(</span><span class="fu">seq</span><span class="op">(</span>along<span class="op">=</span><span class="va">y</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">y</span> <span class="op">=</span> <span class="va">y</span><span class="op">[</span><span class="va">ind</span><span class="op">]</span></span>
<span><span class="va">x</span> <span class="op">=</span> <span class="va">x</span><span class="op">[</span><span class="va">ind</span><span class="op">]</span></span></code></pre>
</div>
<p>Set the seed at 5, <code>set.seed(5)</code> and take a random sample
of 250 from:</p>
<div class="codewrapper sourceCode" id="cb6">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">set.seed</span><span class="op">(</span><span class="fl">5</span><span class="op">)</span></span>
<span><span class="va">N</span> <span class="op">=</span> <span class="fl">250</span></span>
<span><span class="co"># Take a sample of size N=250 individuals from our mixed population</span></span>
<span><span class="va">ind</span> <span class="op">=</span> <span class="fu">sample</span><span class="op">(</span><span class="fu">length</span><span class="op">(</span><span class="va">y</span><span class="op">)</span>, <span class="va">N</span><span class="op">)</span></span>
<span><span class="co"># Remember that `Y` contains the labels that identify if the individual is a</span></span>
<span><span class="co"># man or a woman.</span></span>
<span><span class="va">Y</span> <span class="op">=</span> <span class="va">y</span><span class="op">[</span><span class="va">ind</span><span class="op">]</span></span>
<span><span class="co"># And `X` contains the heights if those individuals.</span></span>
<span><span class="va">X</span> <span class="op">=</span> <span class="va">x</span><span class="op">[</span><span class="va">ind</span><span class="op">]</span></span></code></pre>
</div>
<p>Use loess to estimate f(x) = E(Y |X = x) using the default
parameters. What is the predicted f(168)?</p>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1"> Show me the solution </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" aria-labelledby="headingSolution1" data-bs-parent="#accordionSolution1">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb7">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Fit a LOESS model to predict if an individual is a man or a woman using</span></span>
<span><span class="co"># its height as predictor.</span></span>
<span><span class="va">fit</span> <span class="op">&lt;-</span> <span class="fu">loess</span><span class="op">(</span><span class="va">Y</span><span class="op">~</span><span class="va">X</span><span class="op">)</span></span>
<span><span class="co"># Generate a grid on the height axis to plot the model fitted above</span></span>
<span><span class="va">newx</span> <span class="op">&lt;-</span> <span class="fu">seq</span><span class="op">(</span><span class="fu">min</span><span class="op">(</span><span class="va">X</span><span class="op">)</span>,<span class="fu">max</span><span class="op">(</span><span class="va">X</span><span class="op">)</span>,len<span class="op">=</span><span class="fl">45</span><span class="op">)</span></span>
<span><span class="co"># Predict if the individual is a man or a woman according to the heights on</span></span>
<span><span class="co"># our `newx` grid</span></span>
<span><span class="va">hat</span> <span class="op">&lt;-</span> <span class="fu">predict</span><span class="op">(</span><span class="va">fit</span>, newdata<span class="op">=</span><span class="fu">data.frame</span><span class="op">(</span>X<span class="op">=</span><span class="va">newx</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">mypar</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu">plot</span><span class="op">(</span><span class="va">X</span>,<span class="va">Y</span><span class="op">)</span></span>
<span><span class="fu">names</span><span class="op">(</span><span class="va">hat</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu">round</span><span class="op">(</span><span class="va">newx</span>,<span class="fl">1</span><span class="op">)</span></span>
<span><span class="fu">lines</span><span class="op">(</span><span class="va">newx</span>,<span class="va">hat</span><span class="op">)</span></span>
<span><span class="co"># Lets check what is the predicted label for an individual whos height is</span></span>
<span><span class="co"># 168 cm. A label closer to 0 (&lt; 0.5) would be an insight that the</span></span>
<span><span class="co"># individual is a man, whereas a label closer to 1 (0.5) would indicate</span></span>
<span><span class="co"># that the individual is a woman.</span></span>
<span><span class="va">hat</span><span class="op">[</span><span class="st">'168'</span><span class="op">]</span></span></code></pre>
</div>
</div>
</div>
</div>
</div>
<div id="exercise-2" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="exercise-2" class="callout-inner">
<h3 class="callout-title">Exercise 2</h3>
<div class="callout-content">
<p>The loess estimate above is a random variable. We can compute
standard errors for it. Here we use Monte Carlo to demonstrate that it
is a random variable. Use Monte Carlo simulation to estimate the
standard error of your estimate of f (168).</p>
<p>Set the seed to 5, set.seed(5) and perform 10000 simulations and
report the SE of the loess-based estimate.</p>
</div>
</div>
</div>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2"> Show me the solution </h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" aria-labelledby="headingSolution2" data-bs-parent="#accordionSolution2">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb8">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">set.seed</span><span class="op">(</span><span class="fl">5</span><span class="op">)</span></span>
<span><span class="va">B</span> <span class="op">&lt;-</span> <span class="fl">10000</span></span>
<span><span class="va">N</span> <span class="op">&lt;-</span> <span class="fl">250</span></span>
<span><span class="va">newx</span> <span class="op">&lt;-</span> <span class="fu">seq</span><span class="op">(</span><span class="fu">min</span><span class="op">(</span><span class="va">X</span><span class="op">)</span>,<span class="fu">max</span><span class="op">(</span><span class="va">X</span><span class="op">)</span>,len<span class="op">=</span><span class="fl">45</span><span class="op">)</span></span>
<span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu">replicate</span><span class="op">(</span><span class="va">B</span>, <span class="op">{</span></span>
<span>  <span class="va">ind</span> <span class="op">=</span> <span class="fu">sample</span><span class="op">(</span><span class="fu">length</span><span class="op">(</span><span class="va">y</span><span class="op">)</span>,<span class="va">N</span><span class="op">)</span></span>
<span>  <span class="va">Y</span> <span class="op">=</span> <span class="va">y</span><span class="op">[</span><span class="va">ind</span><span class="op">]</span></span>
<span>  <span class="va">X</span> <span class="op">=</span> <span class="va">x</span><span class="op">[</span><span class="va">ind</span><span class="op">]</span></span>
<span>  <span class="co"># The model fitted by LOESS will be different according to the data used</span></span>
<span>  <span class="co"># to fit it, so we need to fit it again to each new random sample.</span></span>
<span>  <span class="va">fit</span> <span class="op">&lt;-</span> <span class="fu">loess</span><span class="op">(</span><span class="va">Y</span><span class="op">~</span><span class="va">X</span><span class="op">)</span></span>
<span>  <span class="va">hat</span> <span class="op">&lt;-</span> <span class="fu">predict</span><span class="op">(</span><span class="va">fit</span>, newdata<span class="op">=</span><span class="fu">data.frame</span><span class="op">(</span>X<span class="op">=</span><span class="va">newx</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="fu">names</span><span class="op">(</span><span class="va">hat</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu">round</span><span class="op">(</span><span class="va">newx</span>,<span class="fl">1</span><span class="op">)</span></span>
<span>  <span class="co"># Because the model is different, the predicted label for a specific</span></span>
<span>  <span class="co"># height will be different too. We are focused to know how much that</span></span>
<span>  <span class="co"># prediction will vary.</span></span>
<span>  <span class="kw">return</span><span class="op">(</span><span class="va">hat</span><span class="op">[</span><span class="st">'168'</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span>
<span><span class="fu">names</span><span class="op">(</span><span class="va">res</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="cn">NULL</span></span>
<span><span class="co"># Compute the Standard Error (SE) of the label estimation</span></span>
<span><span class="fu">popsd</span><span class="op">(</span><span class="va">res</span><span class="op">)</span></span></code></pre>
</div>
</div>
</div>
</div>
</div>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Key Points</h3>
<div class="callout-content">
<ul>
<li>The smoothing methods work well when used inside the range of
predictor values seen in the training set, however them are not suitable
for extrapolation the prediction outside those ranges.</li>
</ul>
</div>
</div>
</div>
<!--
Place links that you need to refer to multiple times across pages here. Delete
any links that you are not going to use.
 -->
</section></section><section id="aio-class-prediction"><p>Content from <a href="class-prediction.html">Class Prediction</a></p>
<hr>
<p>Last updated on 2025-07-18 |

        <a href="https://github.com/smcclatchy/machine-learning-biomedical-science/edit/main/episodes/class-prediction.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<p>Estimated time: <i aria-hidden="true" data-feather="clock"></i> 60 minutes</p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>What is machine learning (ML)?</li>
<li>Why should we learn ML?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>In brief, know two major types of ML.</li>
<li>Learn training and test data and their relevance.</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<section><h2 class="section-heading" id="class-prediction">Class Prediction<a class="anchor" aria-label="anchor" href="#class-prediction"></a>
</h2>
<hr class="half-width">
<p>Here we give a brief introduction to the main task of machine
learning: class prediction. In fact, many refer to class prediction as
machine learning and we sometimes use the two terms interchangeably. We
give a very brief introduction to this vast topic, focusing on some
specific examples.</p>
<p>Some of the examples we give here are motivated by those in the
excellent textbook <em>The Elements of Statistical Learning: Data
Mining, Inference, and Prediction</em>, by Trevor Hastie, Robert
Tibshirani and Jerome Friedman, which can be found <a href="http://statweb.stanford.edu/~tibs/ElemStatLearn/" class="external-link">here</a>.</p>
<p>Similar to inference in the context of regression, Machine Learning
(ML) studies the relationships between outcomes <span class="math inline">\(Y\)</span> and covariates <span class="math inline">\(X\)</span>. In ML, we call <span class="math inline">\(X\)</span> the predictors or features. The main
difference between ML and inference is that, in ML, we are interested
mainly in predicting <span class="math inline">\(Y\)</span> using <span class="math inline">\(X\)</span>. Statistical models are used, but while
in inference we estimate and interpret model parameters, in ML they are
mainly a means to an end: predicting <span class="math inline">\(Y\)</span>.</p>
<p>Here we introduce the main concepts needed to understand ML, along
with two specific algorithms: regression and k nearest neighbors (kNN).
Keep in mind that there are dozens of popular algorithms that we do not
cover here.</p>
<p>In a previous section, we covered the very simple one-predictor case.
However, most of ML is concerned with cases with more than one
predictor. For illustration purposes, we move to a case in which <span class="math inline">\(X\)</span> is two dimensional and <span class="math inline">\(Y\)</span> is binary. We simulate a situation with
a non-linear relationship using an example from the Hastie, Tibshirani
and Friedman book. In the plot below, we show the actual values of <span class="math inline">\(f(x_1,x_2)=E(Y \mid X_1=x_1,X_2=x_2)\)</span>
using colors. The following code is used to create a relatively complex
conditional probability function. We create the test and train data we
use later (code not shown). Here is the plot of <span class="math inline">\(f(x_1,x_2)\)</span> with red representing values
close to 1, blue representing values close to 0, and yellow values in
between.</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">rafalib</span><span class="op">)</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">RColorBrewer</span><span class="op">)</span></span>
<span><span class="va">hmcol</span> <span class="op">&lt;-</span> <span class="fu">colorRampPalette</span><span class="op">(</span><span class="fu">rev</span><span class="op">(</span><span class="fu">brewer.pal</span><span class="op">(</span><span class="fl">11</span>,  <span class="st">"Spectral"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">(</span><span class="fl">100</span><span class="op">)</span></span>
<span><span class="va">mycols</span><span class="op">=</span><span class="fu">c</span><span class="op">(</span><span class="va">hmcol</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, <span class="va">hmcol</span><span class="op">[</span><span class="fl">100</span><span class="op">]</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">set.seed</span><span class="op">(</span><span class="fl">1</span><span class="op">)</span></span>
<span><span class="co">## create covariates and outcomes</span></span>
<span><span class="co">## outcomes are always 50 0s and 50 1s</span></span>
<span><span class="va">s2</span><span class="op">=</span><span class="fl">0.15</span></span>
<span></span>
<span><span class="co">## pick means to create a non linear conditional expectation</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">MASS</span><span class="op">)</span></span>
<span><span class="va">M0</span> <span class="op">&lt;-</span> <span class="fu">mvrnorm</span><span class="op">(</span><span class="fl">10</span>, <span class="fu">c</span><span class="op">(</span><span class="fl">1</span>,<span class="fl">0</span><span class="op">)</span>, <span class="va">s2</span><span class="op">*</span><span class="fu">diag</span><span class="op">(</span><span class="fl">2</span><span class="op">)</span><span class="op">)</span> <span class="co">## generate 10 means</span></span>
<span><span class="va">M1</span> <span class="op">&lt;-</span> <span class="fu">rbind</span><span class="op">(</span><span class="fu">mvrnorm</span><span class="op">(</span><span class="fl">3</span>, <span class="fu">c</span><span class="op">(</span><span class="fl">1</span>,<span class="fl">1</span><span class="op">)</span>, <span class="va">s2</span><span class="op">*</span><span class="fu">diag</span><span class="op">(</span><span class="fl">2</span><span class="op">)</span><span class="op">)</span>,</span>
<span>            <span class="fu">mvrnorm</span><span class="op">(</span><span class="fl">3</span>, <span class="fu">c</span><span class="op">(</span><span class="fl">0</span>,<span class="fl">1</span><span class="op">)</span>, <span class="va">s2</span><span class="op">*</span><span class="fu">diag</span><span class="op">(</span><span class="fl">2</span><span class="op">)</span><span class="op">)</span>,</span>
<span>            <span class="fu">mvrnorm</span><span class="op">(</span><span class="fl">4</span>, <span class="fu">c</span><span class="op">(</span><span class="fl">0</span>,<span class="fl">0</span><span class="op">)</span>,<span class="va">s2</span><span class="op">*</span><span class="fu">diag</span><span class="op">(</span><span class="fl">2</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">### function to generate random pairs</span></span>
<span><span class="va">s</span> <span class="op">&lt;-</span> <span class="fu">sqrt</span><span class="op">(</span><span class="fl">1</span><span class="op">/</span><span class="fl">5</span><span class="op">)</span></span>
<span><span class="va">N</span><span class="op">=</span><span class="fl">200</span></span>
<span><span class="va">makeX</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">M</span>, <span class="va">n</span><span class="op">=</span><span class="va">N</span>, <span class="va">sigma</span><span class="op">=</span><span class="va">s</span><span class="op">*</span><span class="fu">diag</span><span class="op">(</span><span class="fl">2</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="va">z</span> <span class="op">&lt;-</span> <span class="fu">sample</span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span>, <span class="va">n</span>, replace<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span> <span class="co">## pick n at random from above 10</span></span>
<span>  <span class="va">m</span> <span class="op">&lt;-</span> <span class="va">M</span><span class="op">[</span><span class="va">z</span>,<span class="op">]</span> <span class="co">## these are the n vectors (2 components)</span></span>
<span>  <span class="kw">return</span><span class="op">(</span><span class="fu">t</span><span class="op">(</span><span class="fu">apply</span><span class="op">(</span><span class="va">m</span>, <span class="fl">1</span>, <span class="kw">function</span><span class="op">(</span><span class="va">mu</span><span class="op">)</span> <span class="fu">mvrnorm</span><span class="op">(</span><span class="fl">1</span>, <span class="va">mu</span>, <span class="va">sigma</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="co">## the final values</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="co">### create the training set and the test set</span></span>
<span><span class="va">x0</span> <span class="op">&lt;-</span> <span class="fu">makeX</span><span class="op">(</span><span class="va">M0</span><span class="op">)</span> <span class="co">##the final values for y=0 (green)</span></span>
<span><span class="va">testx0</span> <span class="op">&lt;-</span> <span class="fu">makeX</span><span class="op">(</span><span class="va">M0</span><span class="op">)</span></span>
<span><span class="va">x1</span> <span class="op">&lt;-</span> <span class="fu">makeX</span><span class="op">(</span><span class="va">M1</span><span class="op">)</span></span>
<span><span class="va">testx1</span> <span class="op">&lt;-</span> <span class="fu">makeX</span><span class="op">(</span><span class="va">M1</span><span class="op">)</span></span>
<span><span class="va">x</span> <span class="op">&lt;-</span> <span class="fu">rbind</span><span class="op">(</span><span class="va">x0</span>, <span class="va">x1</span><span class="op">)</span> <span class="co">## one matrix with everything</span></span>
<span><span class="va">test</span> <span class="op">&lt;-</span> <span class="fu">rbind</span><span class="op">(</span><span class="va">testx0</span>, <span class="va">testx1</span><span class="op">)</span></span>
<span><span class="va">y</span> <span class="op">&lt;-</span> <span class="fu">c</span><span class="op">(</span><span class="fu">rep</span><span class="op">(</span><span class="fl">0</span>, <span class="va">N</span><span class="op">)</span>, <span class="fu">rep</span><span class="op">(</span><span class="fl">1</span>, <span class="va">N</span><span class="op">)</span><span class="op">)</span> <span class="co"># the outcomes</span></span>
<span><span class="va">ytest</span> <span class="op">&lt;-</span> <span class="fu">c</span><span class="op">(</span><span class="fu">rep</span><span class="op">(</span><span class="fl">0</span>, <span class="va">N</span><span class="op">)</span>, <span class="fu">rep</span><span class="op">(</span><span class="fl">1</span>, <span class="va">N</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">cols</span> <span class="op">&lt;-</span> <span class="va">mycols</span><span class="op">[</span><span class="fu">c</span><span class="op">(</span><span class="fu">rep</span><span class="op">(</span><span class="fl">1</span>, <span class="va">N</span><span class="op">)</span>, <span class="fu">rep</span><span class="op">(</span><span class="fl">2</span>, <span class="va">N</span><span class="op">)</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="va">colstest</span> <span class="op">&lt;-</span> <span class="va">cols</span></span>
<span></span>
<span><span class="co">## Create a grid so we can predict all of X,Y</span></span>
<span><span class="va">GS</span> <span class="op">&lt;-</span> <span class="fl">150</span> <span class="co">## grid size is GS x GS</span></span>
<span><span class="va">XLIM</span> <span class="op">&lt;-</span> <span class="fu">c</span><span class="op">(</span><span class="fu">min</span><span class="op">(</span><span class="fu">c</span><span class="op">(</span><span class="va">x</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span>, <span class="va">test</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span><span class="op">)</span><span class="op">)</span>, <span class="fu">max</span><span class="op">(</span><span class="fu">c</span><span class="op">(</span><span class="va">x</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span>, <span class="va">test</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">tmpx</span> <span class="op">&lt;-</span> <span class="fu">seq</span><span class="op">(</span><span class="va">XLIM</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, <span class="va">XLIM</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span>, len<span class="op">=</span><span class="va">GS</span><span class="op">)</span></span>
<span><span class="va">YLIM</span> <span class="op">&lt;-</span> <span class="fu">c</span><span class="op">(</span><span class="fu">min</span><span class="op">(</span><span class="fu">c</span><span class="op">(</span><span class="va">x</span><span class="op">[</span>,<span class="fl">2</span><span class="op">]</span>, <span class="va">test</span><span class="op">[</span>,<span class="fl">2</span><span class="op">]</span><span class="op">)</span><span class="op">)</span>, <span class="fu">max</span><span class="op">(</span><span class="fu">c</span><span class="op">(</span><span class="va">x</span><span class="op">[</span>,<span class="fl">2</span><span class="op">]</span>, <span class="va">test</span><span class="op">[</span>,<span class="fl">2</span><span class="op">]</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">tmpy</span> <span class="op">&lt;-</span> <span class="fu">seq</span><span class="op">(</span><span class="va">YLIM</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, <span class="va">YLIM</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span>, len<span class="op">=</span><span class="va">GS</span><span class="op">)</span></span>
<span><span class="va">newx</span> <span class="op">&lt;-</span> <span class="fu">expand.grid</span><span class="op">(</span><span class="va">tmpx</span>, <span class="va">tmpy</span><span class="op">)</span> <span class="co">#grid used to show color contour of predictions</span></span>
<span></span>
<span><span class="co">### Bayes rule: best possible answer</span></span>
<span><span class="va">p</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">{</span> <span class="co">##probability of Y given X</span></span>
<span>  <span class="va">p0</span> <span class="op">&lt;-</span> <span class="fu">mean</span><span class="op">(</span><span class="fu">dnorm</span><span class="op">(</span><span class="va">x</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, <span class="va">M0</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span>, <span class="va">s</span><span class="op">)</span> <span class="op">*</span> <span class="fu">dnorm</span><span class="op">(</span><span class="va">x</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span>, <span class="va">M0</span><span class="op">[</span>,<span class="fl">2</span><span class="op">]</span>, <span class="va">s</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu">mean</span><span class="op">(</span><span class="fu">dnorm</span><span class="op">(</span><span class="va">x</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, <span class="va">M1</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span>, <span class="va">s</span><span class="op">)</span> <span class="op">*</span> <span class="fu">dnorm</span><span class="op">(</span><span class="va">x</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span>, <span class="va">M1</span><span class="op">[</span>,<span class="fl">2</span><span class="op">]</span>, <span class="va">s</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="va">p1</span><span class="op">/</span><span class="op">(</span><span class="va">p0</span><span class="op">+</span><span class="va">p1</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="co">### Create the bayesrule prediction</span></span>
<span><span class="va">bayesrule</span> <span class="op">&lt;-</span> <span class="fu">apply</span><span class="op">(</span><span class="va">newx</span>, <span class="fl">1</span>, <span class="va">p</span><span class="op">)</span></span>
<span><span class="va">colshat</span> <span class="op">&lt;-</span> <span class="va">bayesrule</span></span>
<span></span>
<span><span class="va">colshat</span> <span class="op">&lt;-</span> <span class="va">hmcol</span><span class="op">[</span><span class="fu">floor</span><span class="op">(</span><span class="va">bayesrule</span> <span class="op">*</span> <span class="fl">100</span><span class="op">)</span> <span class="op">+</span> <span class="fl">1</span><span class="op">]</span></span>
<span></span>
<span><span class="fu">mypar</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu">plot</span><span class="op">(</span><span class="va">x</span>, type<span class="op">=</span><span class="st">"n"</span>, xlab<span class="op">=</span><span class="st">"X1"</span>, ylab<span class="op">=</span><span class="st">"X2"</span>, xlim<span class="op">=</span><span class="va">XLIM</span>, ylim<span class="op">=</span><span class="va">YLIM</span><span class="op">)</span></span>
<span><span class="fu">points</span><span class="op">(</span><span class="va">newx</span>, col<span class="op">=</span><span class="va">colshat</span>, pch<span class="op">=</span><span class="fl">16</span>, cex<span class="op">=</span><span class="fl">0.35</span><span class="op">)</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="../fig/class-prediction-rendered-conditional_prob-1.png" alt="Probability of Y=1 as a function of X1 and X2. Red is close to 1, yellow close to 0.5, and blue close to 0." class="figure mx-auto d-block"><figcaption>
Probability of Y=1 as a function of X1 and X2. Red is close to 1, yellow
close to 0.5, and blue close to 0.
</figcaption></figure><p>If we show points for which <span class="math inline">\(E(Y \mid X=x)
&gt; 0.5\)</span> in red and the rest in blue, we see the boundary
region that denotes the boundary in which we switch from predicting 0 to
1.</p>
<div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">mypar</span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="va">colshat</span><span class="op">[</span><span class="va">bayesrule</span><span class="op">&gt;=</span><span class="fl">0.5</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">mycols</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span></span>
<span><span class="va">colshat</span><span class="op">[</span><span class="va">bayesrule</span><span class="op">&lt;</span><span class="fl">0.5</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">mycols</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span></span>
<span></span>
<span><span class="fu">plot</span><span class="op">(</span><span class="va">x</span>,type<span class="op">=</span><span class="st">"n"</span>,xlab<span class="op">=</span><span class="st">"X1"</span>,ylab<span class="op">=</span><span class="st">"X2"</span>,xlim<span class="op">=</span><span class="va">XLIM</span>,ylim<span class="op">=</span><span class="va">YLIM</span><span class="op">)</span></span>
<span><span class="fu">points</span><span class="op">(</span><span class="va">newx</span>,col<span class="op">=</span><span class="va">colshat</span>,pch<span class="op">=</span><span class="fl">16</span>,cex<span class="op">=</span><span class="fl">0.35</span><span class="op">)</span></span>
<span><span class="fu">contour</span><span class="op">(</span><span class="va">tmpx</span>,<span class="va">tmpy</span>,<span class="fu">matrix</span><span class="op">(</span><span class="fu">round</span><span class="op">(</span><span class="va">bayesrule</span><span class="op">)</span>,<span class="va">GS</span>,<span class="va">GS</span><span class="op">)</span>,levels<span class="op">=</span><span class="fu">c</span><span class="op">(</span><span class="fl">1</span>,<span class="fl">2</span><span class="op">)</span>,</span>
<span>        add<span class="op">=</span><span class="cn">TRUE</span>,drawlabels<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="../fig/class-prediction-rendered-bayes_rule-1.png" alt="Bayes rule. The line divides part of the space for which probability is larger than 0.5 (red) and lower than 0.5 (blue)." class="figure mx-auto d-block"><figcaption>
Bayes rule. The line divides part of the space for which probability is
larger than 0.5 (red) and lower than 0.5 (blue).
</figcaption></figure><p>The above plots relate to the “truth” that we do not get to see. Most
ML methodology is concerned with estimating <span class="math inline">\(f(x)\)</span>. A typical first step is usually to
consider a sample, referred to as the training set, to estimate <span class="math inline">\(f(x)\)</span>. We will review two specific ML
techniques. First, we need to review the main concept we use to evaluate
the performance of these methods.</p>
<div class="section level4">
<h4 id="training-and-test-sets">Training and test sets<a class="anchor" aria-label="anchor" href="#training-and-test-sets"></a>
</h4>
<p>In the code (not shown) for the first plot in this chapter, we
created a test and a training set. We plot them here:</p>
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#x, test, cols, and coltest were created in code that was not shown</span></span>
<span><span class="co">#x is training x1 and x2, test is test x1 and x2</span></span>
<span><span class="co">#cols (0=blue, 1=red) are training observations</span></span>
<span><span class="co">#coltests are test observations</span></span>
<span><span class="fu">mypar</span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span><span class="op">)</span></span>
<span><span class="fu">plot</span><span class="op">(</span><span class="va">x</span>, pch<span class="op">=</span><span class="fl">21</span>, bg<span class="op">=</span><span class="va">cols</span>, xlab<span class="op">=</span><span class="st">"X1"</span>, ylab<span class="op">=</span><span class="st">"X2"</span>, xlim<span class="op">=</span><span class="va">XLIM</span>, ylim<span class="op">=</span><span class="va">YLIM</span><span class="op">)</span></span>
<span><span class="fu">plot</span><span class="op">(</span><span class="va">test</span>, pch<span class="op">=</span><span class="fl">21</span>, bg<span class="op">=</span><span class="va">colstest</span>, xlab<span class="op">=</span><span class="st">"X1"</span>, ylab<span class="op">=</span><span class="st">"X2"</span>, xlim<span class="op">=</span><span class="va">XLIM</span>, ylim<span class="op">=</span><span class="va">YLIM</span><span class="op">)</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="../fig/class-prediction-rendered-test_train-1.png" alt="Training data (left) and test data (right)." class="figure mx-auto d-block"><figcaption>
Training data (left) and test data (right).
</figcaption></figure><p>You will notice that the test and train set have similar global
properties since they were generated by the same random variables (more
blue towards the bottom right), but are, by construction, different.
<strong>The reason we create test and training sets</strong> is to
detect over-training by testing on a different data than the one used to
fit models or train algorithms. We will see how important this is
below.</p>
</div>
<div class="section level4">
<h4 id="predicting-with-regression">Predicting with regression<a class="anchor" aria-label="anchor" href="#predicting-with-regression"></a>
</h4>
<p>A first naive approach to this ML problem is to fit a <strong>two
variable linear regression model</strong>:</p>
<div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">##x and y were created in the code (not shown) for the first plot</span></span>
<span><span class="co">#y is outcome for the training set</span></span>
<span><span class="va">X1</span> <span class="op">&lt;-</span> <span class="va">x</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span> <span class="co">##these are the covariates</span></span>
<span><span class="va">X2</span> <span class="op">&lt;-</span> <span class="va">x</span><span class="op">[</span>,<span class="fl">2</span><span class="op">]</span> </span>
<span><span class="va">fit1</span> <span class="op">&lt;-</span> <span class="fu">lm</span><span class="op">(</span><span class="va">y</span><span class="op">~</span><span class="va">X1</span><span class="op">+</span><span class="va">X2</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## get summary of a fitted model</span></span>
<span><span class="fu">summary</span><span class="op">(</span><span class="va">fit1</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>
Call:
lm(formula = y ~ X1 + X2)

Residuals:
     Min       1Q   Median       3Q      Max
-0.99350 -0.36244  0.00274  0.37927  0.92329

Coefficients:
            Estimate Std. Error t value Pr(&gt;|t|)
(Intercept)  0.55305    0.02559  21.608  &lt; 2e-16 ***
X1          -0.20902    0.02378  -8.790  &lt; 2e-16 ***
X2           0.22230    0.02610   8.517 3.42e-16 ***
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

Residual standard error: 0.4382 on 397 degrees of freedom
Multiple R-squared:  0.2377,	Adjusted R-squared:  0.2339
F-statistic: 61.91 on 2 and 397 DF,  p-value: &lt; 2.2e-16</code></pre>
</div>
<p>Once we the have fitted values, we can estimate <span class="math inline">\(f(x_1,x_2)\)</span> with <span class="math inline">\(\hat{f}(x_1,x_2)=\hat{\beta}_0 + \hat{\beta}_1x_1
+\hat{\beta}_2 x_2\)</span>. To provide an actual prediction, we simply
predict 1 when <span class="math inline">\(\hat{f}(x_1,x_2)&gt;0.5\)</span>. We now examine
the error rates in the test and training sets and also plot the boundary
region:</p>
<div class="codewrapper sourceCode" id="cb6">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">##prediction on train</span></span>
<span><span class="va">yhat</span> <span class="op">&lt;-</span> <span class="fu">predict</span><span class="op">(</span><span class="va">fit1</span><span class="op">)</span></span>
<span><span class="va">yhat</span> <span class="op">&lt;-</span> <span class="fu">as.numeric</span><span class="op">(</span><span class="va">yhat</span><span class="op">&gt;</span><span class="fl">0.5</span><span class="op">)</span></span>
<span><span class="fu">cat</span><span class="op">(</span><span class="st">"Linear regression prediction error in train:"</span>, <span class="fl">1</span><span class="op">-</span><span class="fu">mean</span><span class="op">(</span><span class="va">yhat</span><span class="op">==</span><span class="va">y</span><span class="op">)</span>, <span class="st">"\n"</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Linear regression prediction error in train: 0.2975 </code></pre>
</div>
<p>We can quickly obtain predicted values for any set of values using
the <code>predict</code> function:</p>
<div class="codewrapper sourceCode" id="cb8">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">yhat</span> <span class="op">&lt;-</span> <span class="fu">predict</span><span class="op">(</span><span class="va">fit1</span>, newdata<span class="op">=</span><span class="fu">data.frame</span><span class="op">(</span>X1<span class="op">=</span><span class="va">newx</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span>, X2<span class="op">=</span><span class="va">newx</span><span class="op">[</span>,<span class="fl">2</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<p>Now we can create a plot showing where we predict 1s and where we
predict 0s, as well as the boundary. We can also use the
<code>predict</code> function to obtain predicted values for our test
set. <strong>Note that nowhere do we fit the model on the test
set:</strong></p>
<div class="codewrapper sourceCode" id="cb9">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">colshat</span> <span class="op">&lt;-</span> <span class="va">yhat</span></span>
<span><span class="va">colshat</span><span class="op">[</span><span class="va">yhat</span><span class="op">&gt;=</span><span class="fl">0.5</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">mycols</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span></span>
<span><span class="va">colshat</span><span class="op">[</span><span class="va">yhat</span><span class="op">&lt;</span><span class="fl">0.5</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">mycols</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span></span>
<span><span class="va">m</span> <span class="op">&lt;-</span> <span class="op">-</span><span class="va">fit1</span><span class="op">$</span><span class="va">coef</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span><span class="op">/</span><span class="va">fit1</span><span class="op">$</span><span class="va">coef</span><span class="op">[</span><span class="fl">3</span><span class="op">]</span> <span class="co">#boundary slope</span></span>
<span><span class="va">b</span> <span class="op">&lt;-</span> <span class="op">(</span><span class="fl">0.5</span> <span class="op">-</span> <span class="va">fit1</span><span class="op">$</span><span class="va">coef</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span><span class="op">/</span><span class="va">fit1</span><span class="op">$</span><span class="va">coef</span><span class="op">[</span><span class="fl">3</span><span class="op">]</span> <span class="co">#boundary intercept</span></span>
<span></span>
<span><span class="co">##prediction on test</span></span>
<span><span class="va">yhat</span> <span class="op">&lt;-</span> <span class="fu">predict</span><span class="op">(</span><span class="va">fit1</span>,newdata<span class="op">=</span><span class="fu">data.frame</span><span class="op">(</span>X1<span class="op">=</span><span class="va">test</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span>,X2<span class="op">=</span><span class="va">test</span><span class="op">[</span>,<span class="fl">2</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">yhat</span> <span class="op">&lt;-</span> <span class="fu">as.numeric</span><span class="op">(</span><span class="va">yhat</span><span class="op">&gt;</span><span class="fl">0.5</span><span class="op">)</span></span>
<span><span class="fu">cat</span><span class="op">(</span><span class="st">"Linear regression prediction error in test:"</span>,<span class="fl">1</span><span class="op">-</span><span class="fu">mean</span><span class="op">(</span><span class="va">yhat</span><span class="op">==</span><span class="va">ytest</span><span class="op">)</span>,<span class="st">"\n"</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Linear regression prediction error in test: 0.32 </code></pre>
</div>
<div class="codewrapper sourceCode" id="cb11">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">plot</span><span class="op">(</span><span class="va">test</span>,type<span class="op">=</span><span class="st">"n"</span>,xlab<span class="op">=</span><span class="st">"X1"</span>,ylab<span class="op">=</span><span class="st">"X2"</span>,xlim<span class="op">=</span><span class="va">XLIM</span>,ylim<span class="op">=</span><span class="va">YLIM</span><span class="op">)</span></span>
<span><span class="fu">abline</span><span class="op">(</span><span class="va">b</span>,<span class="va">m</span><span class="op">)</span></span>
<span><span class="fu">points</span><span class="op">(</span><span class="va">newx</span>,col<span class="op">=</span><span class="va">colshat</span>,pch<span class="op">=</span><span class="fl">16</span>,cex<span class="op">=</span><span class="fl">0.35</span><span class="op">)</span></span>
<span></span>
<span><span class="co">##test was created in the code (not shown) for the first plot</span></span>
<span><span class="fu">points</span><span class="op">(</span><span class="va">test</span>,bg<span class="op">=</span><span class="va">cols</span>,pch<span class="op">=</span><span class="fl">21</span><span class="op">)</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="../fig/class-prediction-rendered-regression_prediction-1.png" alt="We estimate the probability of 1 with a linear regression model with X1 and X2 as predictors. The resulting prediction map is divided into parts that are larger than 0.5 (red) and lower than 0.5 (blue)." class="figure mx-auto d-block"><figcaption>
We estimate the probability of 1 with a linear regression model with X1
and X2 as predictors. The resulting prediction map is divided into parts
that are larger than 0.5 (red) and lower than 0.5 (blue).
</figcaption></figure><p>The error rates in the test and train sets are quite similar. Thus,
we do not seem to be over-training. This is not surprising as we are
fitting a 2 parameter model to 400 data points. However, note that the
boundary is a line. Because we are fitting a plane to the data, there is
no other option here. <strong>The linear regression method is too
rigid.</strong> The rigidity makes it stable and avoids over-training,
but it also keeps the model from adapting to the non-linear relationship
between <span class="math inline">\(Y\)</span> and <span class="math inline">\(X\)</span>. We saw this before in the smoothing
section. The next ML technique we consider is similar to the smoothing
techniques described before.</p>
</div>
<div class="section level4">
<h4 id="k-nearest-neighbor">K-nearest neighbor<a class="anchor" aria-label="anchor" href="#k-nearest-neighbor"></a>
</h4>
<p>K-nearest neighbors (kNN) is similar to bin smoothing, but it is
easier to adapt to multiple dimensions. Basically, for any point <span class="math inline">\(x\)</span> for which we want an estimate, we look
for the k nearest points and then take an average of these points. This
gives us an estimate of <span class="math inline">\(f(x_1,x_2)\)</span>,
just like the bin smoother gave us an estimate of a curve. We can now
control flexibility through <span class="math inline">\(k\)</span>. Here
we compare <span class="math inline">\(k=1\)</span> and <span class="math inline">\(k=100\)</span>.</p>
<div class="codewrapper sourceCode" id="cb12">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">class</span><span class="op">)</span></span>
<span><span class="fu">mypar</span><span class="op">(</span><span class="fl">2</span>,<span class="fl">2</span><span class="op">)</span></span>
<span><span class="kw">for</span><span class="op">(</span><span class="va">k</span> <span class="kw">in</span> <span class="fu">c</span><span class="op">(</span><span class="fl">1</span>, <span class="fl">100</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="co">## predict on train</span></span>
<span>  <span class="va">yhat</span> <span class="op">&lt;-</span> <span class="fu">knn</span><span class="op">(</span><span class="va">x</span>, <span class="va">x</span>, <span class="va">y</span>, k<span class="op">=</span><span class="va">k</span><span class="op">)</span></span>
<span>  <span class="fu">cat</span><span class="op">(</span><span class="st">"KNN prediction error in train:"</span>, </span>
<span>      <span class="fl">1</span><span class="op">-</span><span class="fu">mean</span><span class="op">(</span><span class="op">(</span><span class="fu">as.numeric</span><span class="op">(</span><span class="va">yhat</span><span class="op">)</span><span class="op">-</span><span class="fl">1</span><span class="op">)</span><span class="op">==</span><span class="va">y</span><span class="op">)</span>, <span class="st">"\n"</span><span class="op">)</span></span>
<span>  <span class="co">## make plot</span></span>
<span>  <span class="va">yhat</span> <span class="op">&lt;-</span> <span class="fu">knn</span><span class="op">(</span><span class="va">x</span>,<span class="va">test</span>,<span class="va">y</span>,k<span class="op">=</span><span class="va">k</span><span class="op">)</span></span>
<span>  <span class="fu">cat</span><span class="op">(</span><span class="st">"KNN prediction error in test:"</span>, </span>
<span>      <span class="fl">1</span><span class="op">-</span><span class="fu">mean</span><span class="op">(</span><span class="op">(</span><span class="fu">as.numeric</span><span class="op">(</span><span class="va">yhat</span><span class="op">)</span><span class="op">-</span><span class="fl">1</span><span class="op">)</span><span class="op">==</span><span class="va">ytest</span><span class="op">)</span>, <span class="st">"\n"</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>KNN prediction error in train: 0
KNN prediction error in test: 0.3375
KNN prediction error in train: 0.2725
KNN prediction error in test: 0.3125 </code></pre>
</div>
<p>To visualize why we make no errors in the train set and many errors
in the test set when <span class="math inline">\(k=1\)</span> and obtain
more stable results from <span class="math inline">\(k=100\)</span>, we
show the prediction regions (code not shown):</p>
<div class="codewrapper sourceCode" id="cb14">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">class</span><span class="op">)</span></span>
<span><span class="fu">mypar</span><span class="op">(</span><span class="fl">2</span>,<span class="fl">2</span><span class="op">)</span></span>
<span><span class="kw">for</span><span class="op">(</span><span class="va">k</span> <span class="kw">in</span> <span class="fu">c</span><span class="op">(</span><span class="fl">1</span>,<span class="fl">100</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="co">##predict on train</span></span>
<span>  <span class="va">yhat</span> <span class="op">&lt;-</span> <span class="fu">knn</span><span class="op">(</span><span class="va">x</span>,<span class="va">x</span>,<span class="va">y</span>,k<span class="op">=</span><span class="va">k</span><span class="op">)</span></span>
<span>  <span class="co">##make plot</span></span>
<span>  <span class="va">yhat</span> <span class="op">&lt;-</span> <span class="fu">knn</span><span class="op">(</span><span class="va">x</span>,<span class="va">newx</span>,<span class="va">y</span>,k<span class="op">=</span><span class="va">k</span><span class="op">)</span></span>
<span>  <span class="va">colshat</span> <span class="op">&lt;-</span> <span class="va">mycols</span><span class="op">[</span><span class="fu">as.numeric</span><span class="op">(</span><span class="va">yhat</span><span class="op">)</span><span class="op">]</span></span>
<span>  <span class="fu">plot</span><span class="op">(</span><span class="va">x</span>,type<span class="op">=</span><span class="st">"n"</span>,xlab<span class="op">=</span><span class="st">"X1"</span>,ylab<span class="op">=</span><span class="st">"X2"</span>,xlim<span class="op">=</span><span class="va">XLIM</span>,ylim<span class="op">=</span><span class="va">YLIM</span><span class="op">)</span></span>
<span>  <span class="fu">points</span><span class="op">(</span><span class="va">newx</span>,col<span class="op">=</span><span class="va">colshat</span>,cex<span class="op">=</span><span class="fl">0.35</span>,pch<span class="op">=</span><span class="fl">16</span><span class="op">)</span></span>
<span>  <span class="fu">contour</span><span class="op">(</span><span class="va">tmpx</span>,<span class="va">tmpy</span>,<span class="fu">matrix</span><span class="op">(</span><span class="fu">as.numeric</span><span class="op">(</span><span class="va">yhat</span><span class="op">)</span>,<span class="va">GS</span>,<span class="va">GS</span><span class="op">)</span>,levels<span class="op">=</span><span class="fu">c</span><span class="op">(</span><span class="fl">1</span>,<span class="fl">2</span><span class="op">)</span>,</span>
<span>          add<span class="op">=</span><span class="cn">TRUE</span>,drawlabels<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span>
<span>  <span class="fu">points</span><span class="op">(</span><span class="va">x</span>,bg<span class="op">=</span><span class="va">cols</span>,pch<span class="op">=</span><span class="fl">21</span><span class="op">)</span></span>
<span>  <span class="fu">title</span><span class="op">(</span><span class="fu">paste</span><span class="op">(</span><span class="st">"Train: KNN ("</span>,<span class="va">k</span>,<span class="st">")"</span>,sep<span class="op">=</span><span class="st">""</span><span class="op">)</span><span class="op">)</span></span>
<span>  </span>
<span>  <span class="fu">plot</span><span class="op">(</span><span class="va">test</span>,type<span class="op">=</span><span class="st">"n"</span>,xlab<span class="op">=</span><span class="st">"X1"</span>,ylab<span class="op">=</span><span class="st">"X2"</span>,xlim<span class="op">=</span><span class="va">XLIM</span>,ylim<span class="op">=</span><span class="va">YLIM</span><span class="op">)</span></span>
<span>  <span class="fu">points</span><span class="op">(</span><span class="va">newx</span>,col<span class="op">=</span><span class="va">colshat</span>,cex<span class="op">=</span><span class="fl">0.35</span>,pch<span class="op">=</span><span class="fl">16</span><span class="op">)</span></span>
<span>  <span class="fu">contour</span><span class="op">(</span><span class="va">tmpx</span>,<span class="va">tmpy</span>,<span class="fu">matrix</span><span class="op">(</span><span class="fu">as.numeric</span><span class="op">(</span><span class="va">yhat</span><span class="op">)</span>,<span class="va">GS</span>,<span class="va">GS</span><span class="op">)</span>,levels<span class="op">=</span><span class="fu">c</span><span class="op">(</span><span class="fl">1</span>,<span class="fl">2</span><span class="op">)</span>,</span>
<span>          add<span class="op">=</span><span class="cn">TRUE</span>,drawlabels<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span>
<span>  <span class="fu">points</span><span class="op">(</span><span class="va">test</span>,bg<span class="op">=</span><span class="va">cols</span>,pch<span class="op">=</span><span class="fl">21</span><span class="op">)</span></span>
<span>  <span class="fu">title</span><span class="op">(</span><span class="fu">paste</span><span class="op">(</span><span class="st">"Test: KNN ("</span>,<span class="va">k</span>,<span class="st">")"</span>,sep<span class="op">=</span><span class="st">""</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="../fig/class-prediction-rendered-knn-1.png" alt="Prediction regions obtained with kNN for k=1 (top) and k=200 (bottom). We show both train (left) and test data (right)." class="figure mx-auto d-block"><figcaption>
Prediction regions obtained with kNN for k=1 (top) and k=200 (bottom).
We show both train (left) and test data (right).
</figcaption></figure><p>When <span class="math inline">\(k=1\)</span>, we make no mistakes in
the training test since every point is its closest neighbor and it is
equal to itself. However, we see some islands of blue in the red area
that, once we move to the test set, are more error prone. In the case
<span class="math inline">\(k=100\)</span>, we do not have this problem
and we also see that we improve the error rate over linear regression.
We can also see that our estimate of <span class="math inline">\(f(x_1,x_2)\)</span> is closer to the truth.</p>
</div>
<div class="section level4">
<h4 id="bayes-rule">Bayes rule<a class="anchor" aria-label="anchor" href="#bayes-rule"></a>
</h4>
<figure><img src=".././fig/bayes_rule_wikipedia.png" alt="Bayes Rule 101 - From Wikipedia" class="figure mx-auto d-block"><div class="figcaption">Bayes Rule 101 - From Wikipedia</div>
</figure><p>Here we include a comparison of the test and train set errors for
various values of <span class="math inline">\(k\)</span>. We also
include the error rate that we would make if we actually knew <span class="math inline">\({E}(Y | X_1=x1, X_2=x_2)\)</span> referred to as
<em>Bayes Rule</em>.</p>
<p>We start by computing the error rates…</p>
<div class="codewrapper sourceCode" id="cb15">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">###Bayes Rule</span></span>
<span><span class="va">yhat</span> <span class="op">&lt;-</span> <span class="fu">apply</span><span class="op">(</span><span class="va">test</span>,<span class="fl">1</span>,<span class="va">p</span><span class="op">)</span></span>
<span><span class="fu">cat</span><span class="op">(</span><span class="st">"Bayes rule prediction error in train"</span>, <span class="fl">1</span><span class="op">-</span><span class="fu">mean</span><span class="op">(</span><span class="fu">round</span><span class="op">(</span><span class="va">yhat</span><span class="op">)</span><span class="op">==</span><span class="va">y</span><span class="op">)</span>, <span class="st">"\n"</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Bayes rule prediction error in train 0.2825 </code></pre>
</div>
<div class="codewrapper sourceCode" id="cb17">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">bayes.error</span><span class="op">=</span><span class="fl">1</span><span class="op">-</span><span class="fu">mean</span><span class="op">(</span><span class="fu">round</span><span class="op">(</span><span class="va">yhat</span><span class="op">)</span><span class="op">==</span><span class="va">y</span><span class="op">)</span></span>
<span><span class="va">train.error</span> <span class="op">&lt;-</span> <span class="fu">rep</span><span class="op">(</span><span class="fl">0</span>,<span class="fl">16</span><span class="op">)</span></span>
<span><span class="va">test.error</span> <span class="op">&lt;-</span> <span class="fu">rep</span><span class="op">(</span><span class="fl">0</span>,<span class="fl">16</span><span class="op">)</span></span>
<span><span class="kw">for</span><span class="op">(</span><span class="va">k</span> <span class="kw">in</span> <span class="fu">seq</span><span class="op">(</span>along<span class="op">=</span><span class="va">train.error</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="co">##predict on train</span></span>
<span>  <span class="va">yhat</span> <span class="op">&lt;-</span> <span class="fu">knn</span><span class="op">(</span><span class="va">x</span>,<span class="va">x</span>,<span class="va">y</span>,k<span class="op">=</span><span class="fl">2</span><span class="op">^</span><span class="op">(</span><span class="va">k</span><span class="op">/</span><span class="fl">2</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="va">train.error</span><span class="op">[</span><span class="va">k</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fl">1</span><span class="op">-</span><span class="fu">mean</span><span class="op">(</span><span class="op">(</span><span class="fu">as.numeric</span><span class="op">(</span><span class="va">yhat</span><span class="op">)</span><span class="op">-</span><span class="fl">1</span><span class="op">)</span><span class="op">==</span><span class="va">y</span><span class="op">)</span></span>
<span>  <span class="co">##prediction on test    </span></span>
<span>  <span class="va">yhat</span> <span class="op">&lt;-</span> <span class="fu">knn</span><span class="op">(</span><span class="va">x</span>,<span class="va">test</span>,<span class="va">y</span>,k<span class="op">=</span><span class="fl">2</span><span class="op">^</span><span class="op">(</span><span class="va">k</span><span class="op">/</span><span class="fl">2</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="va">test.error</span><span class="op">[</span><span class="va">k</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fl">1</span><span class="op">-</span><span class="fu">mean</span><span class="op">(</span><span class="op">(</span><span class="fu">as.numeric</span><span class="op">(</span><span class="va">yhat</span><span class="op">)</span><span class="op">-</span><span class="fl">1</span><span class="op">)</span><span class="op">==</span><span class="va">y</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre>
</div>
<p>… and then plot the error rates against values of <span class="math inline">\(k\)</span>. We also show the Bayes rules error
rate as a horizontal line.</p>
<div class="codewrapper sourceCode" id="cb18">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ks</span> <span class="op">&lt;-</span> <span class="fl">2</span><span class="op">^</span><span class="op">(</span><span class="fu">seq</span><span class="op">(</span>along<span class="op">=</span><span class="va">train.error</span><span class="op">)</span><span class="op">/</span><span class="fl">2</span><span class="op">)</span></span>
<span><span class="fu">mypar</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu">plot</span><span class="op">(</span><span class="va">ks</span>, <span class="va">train.error</span>, type<span class="op">=</span><span class="st">"n"</span>, xlab<span class="op">=</span><span class="st">"K"</span>, ylab<span class="op">=</span><span class="st">"Prediction Error"</span>, log<span class="op">=</span><span class="st">"x"</span>,</span>
<span>     ylim<span class="op">=</span><span class="fu">range</span><span class="op">(</span><span class="fu">c</span><span class="op">(</span><span class="va">test.error</span>, <span class="va">train.error</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">lines</span><span class="op">(</span><span class="va">ks</span>, <span class="va">train.error</span>, type<span class="op">=</span><span class="st">"b"</span>, col<span class="op">=</span><span class="fl">4</span>, lty<span class="op">=</span><span class="fl">2</span>, lwd<span class="op">=</span><span class="fl">2</span><span class="op">)</span></span>
<span><span class="fu">lines</span><span class="op">(</span><span class="va">ks</span>, <span class="va">test.error</span>, type<span class="op">=</span><span class="st">"b"</span>, col<span class="op">=</span><span class="fl">5</span>, lty<span class="op">=</span><span class="fl">3</span>, lwd<span class="op">=</span><span class="fl">2</span><span class="op">)</span></span>
<span><span class="fu">abline</span><span class="op">(</span>h<span class="op">=</span><span class="va">bayes.error</span>, col<span class="op">=</span><span class="fl">6</span><span class="op">)</span></span>
<span><span class="fu">legend</span><span class="op">(</span><span class="st">"bottomright"</span>, <span class="fu">c</span><span class="op">(</span><span class="st">"Train"</span>,<span class="st">"Test"</span>,<span class="st">"Bayes"</span><span class="op">)</span>, col<span class="op">=</span><span class="fu">c</span><span class="op">(</span><span class="fl">4</span>,<span class="fl">5</span>,<span class="fl">6</span><span class="op">)</span>, lty<span class="op">=</span><span class="fu">c</span><span class="op">(</span><span class="fl">2</span>,<span class="fl">3</span>,<span class="fl">1</span><span class="op">)</span>, box.lwd<span class="op">=</span><span class="fl">0</span><span class="op">)</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="../fig/class-prediction-rendered-bayes_rule2-1.png" alt="Prediction error in train (pink) and test (green) versus number of neighbors. The yellow line represents what one obtains with Bayes Rule." class="figure mx-auto d-block"><figcaption>
Prediction error in train (pink) and test (green) versus number of
neighbors. The yellow line represents what one obtains with Bayes Rule.
</figcaption></figure><p>Note that these error rates are random variables and have standard
errors. In the next section we describe cross-validation which helps
reduce some of this variability. However, even with this variability,
the plot clearly shows the problem of over-fitting when using values
lower than 20 and under-fitting with values above 100.</p>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Key Points</h3>
<div class="callout-content">
<ul>
<li>Data quality matters. Garbage in, Garbage out!</li>
</ul>
</div>
</div>
</div>
<!--
Place links that you need to refer to multiple times across pages here. Delete
any links that you are not going to use.
 -->
</div>
</section></section><section id="aio-cross-validation"><p>Content from <a href="cross-validation.html">Cross-validation</a></p>
<hr>
<p>Last updated on 2025-07-18 |

        <a href="https://github.com/smcclatchy/machine-learning-biomedical-science/edit/main/episodes/cross-validation.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<p>Estimated time: <i aria-hidden="true" data-feather="clock"></i> 60 minutes</p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>How can the best configuration of parameters be selected for a
machine learning model using only the data available?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Create a set of fold indices for cross-validation.</li>
<li>Select the best configuration for a machine learning model using
cross-validation.</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<section><h2 class="section-heading" id="cross-validation">Cross-validation<a class="anchor" aria-label="anchor" href="#cross-validation"></a>
</h2>
<hr class="half-width">
<p>Here we describe <em>cross-validation</em>: one of the fundamental
methods in machine learning for method assessment and picking parameters
in a prediction or machine learning task. Suppose we have a set of
observations with many features and each observation is associated with
a label. We will call this set our training data. Our task is to predict
the label of any new samples by learning patterns from the training
data. For a concrete example, let’s consider gene expression values,
where each gene acts as a feature. We will be given a new set of
unlabeled data (the test data) with the task of predicting the tissue
type of the new samples.</p>
<p>If we choose a machine learning algorithm with a tunable parameter,
we have to come up with a strategy for picking an optimal value for this
parameter. We could try some values, and then just choose the one which
performs the best on our training data, in terms of the number of errors
the algorithm would make if we apply it to the samples we have been
given for training. However, we have seen how this leads to
over-fitting.</p>
<p>Let’s start by loading the tissue gene expression dataset:</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">load</span><span class="op">(</span><span class="st">"data/tissuesGeneExpression.rda"</span><span class="op">)</span></span></code></pre>
</div>
<p>For illustration purposes, we will drop one of the tissues which
doesn’t have many samples:</p>
<div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">table</span><span class="op">(</span><span class="va">tissue</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>tissue
 cerebellum       colon endometrium hippocampus      kidney       liver
         38          34          15          31          39          26
   placenta
          6 </code></pre>
</div>
<div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ind</span> <span class="op">&lt;-</span> <span class="fu">which</span><span class="op">(</span><span class="va">tissue</span> <span class="op">!=</span> <span class="st">"placenta"</span><span class="op">)</span></span>
<span><span class="va">y</span> <span class="op">&lt;-</span> <span class="va">tissue</span><span class="op">[</span><span class="va">ind</span><span class="op">]</span></span>
<span><span class="va">X</span> <span class="op">&lt;-</span> <span class="fu">t</span><span class="op">(</span> <span class="va">e</span><span class="op">[</span>,<span class="va">ind</span><span class="op">]</span> <span class="op">)</span></span></code></pre>
</div>
<p>This tissue will not form part of our example.</p>
<p>Now let’s try out k-nearest neighbors for classification, using <span class="math inline">\(k=5\)</span>. What is our average error in
predicting the tissue in the training set, when we’ve used the same data
for training and for testing?</p>
<div class="codewrapper sourceCode" id="cb5">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">class</span><span class="op">)</span></span>
<span><span class="va">pred</span> <span class="op">&lt;-</span> <span class="fu">knn</span><span class="op">(</span>train <span class="op">=</span>  <span class="va">X</span>, test <span class="op">=</span> <span class="va">X</span>, cl<span class="op">=</span><span class="va">y</span>, k<span class="op">=</span><span class="fl">5</span><span class="op">)</span></span>
<span><span class="fu">mean</span><span class="op">(</span><span class="va">y</span> <span class="op">!=</span> <span class="va">pred</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 0</code></pre>
</div>
<p>We have no errors in prediction in the training set with <span class="math inline">\(k=5\)</span>. What if we use <span class="math inline">\(k=1\)</span>?</p>
<div class="codewrapper sourceCode" id="cb7">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pred</span> <span class="op">&lt;-</span> <span class="fu">knn</span><span class="op">(</span>train<span class="op">=</span><span class="va">X</span>, test<span class="op">=</span><span class="va">X</span>, cl<span class="op">=</span><span class="va">y</span>, k<span class="op">=</span><span class="fl">1</span><span class="op">)</span></span>
<span><span class="fu">mean</span><span class="op">(</span><span class="va">y</span> <span class="op">!=</span> <span class="va">pred</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 0</code></pre>
</div>
<p>Trying to classify the same observations as we use to <em>train</em>
the model can be very misleading. In fact, for k-nearest neighbors,
using k=1 will always give 0 classification error in the training set,
because we use the single observation to classify itself. The reliable
way to get a sense of the performance of an algorithm is to make it give
a prediction for a sample it has never seen. Similarly, if we want to
know what the best value for a tunable parameter is, we need to see how
different values of the parameter perform on samples, which are not in
the training data.</p>
<p>Cross-validation is a widely-used method in machine learning, which
solves this training and test data problem, while still using all the
data for testing the predictive accuracy. It accomplishes this by
splitting the data into a number of <em>folds</em>. If we have <span class="math inline">\(N\)</span> folds, then the first step of the
algorithm is to train the algorithm using <span class="math inline">\((N-1)\)</span> of the folds, and test the
algorithm’s accuracy on the single left-out fold. This is then repeated
N times until each fold has been used as in the <em>test</em> set. If we
have <span class="math inline">\(M\)</span> parameter settings to try
out, then this is accomplished in an outer loop, so we have to fit the
algorithm a total of <span class="math inline">\(N \times M\)</span>
times.</p>
<p>We will use the <code>createFolds</code> function from the
<code>caret</code> package to make 5 folds of our gene expression data,
which are balanced over the tissues. Don’t be confused by the fact that
the <code>createFolds</code> function uses the same letter ‘k’ as the
‘k’ in k-nearest neighbors. These ‘k’ are totally unrelated. The caret
function <code>createFolds</code> is asking for how many folds to
create, the <span class="math inline">\(N\)</span> from above. The ‘k’
in the <code>knn</code> function is for how many closest observations to
use in classifying a new sample. Here we will create 10 folds:</p>
<div class="codewrapper sourceCode" id="cb9">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">caret</span><span class="op">)</span></span>
<span><span class="fu">set.seed</span><span class="op">(</span><span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">idx</span> <span class="op">&lt;-</span> <span class="fu">createFolds</span><span class="op">(</span><span class="va">y</span>, k<span class="op">=</span><span class="fl">10</span><span class="op">)</span></span>
<span><span class="fu">sapply</span><span class="op">(</span><span class="va">idx</span>, <span class="va">length</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Fold01 Fold02 Fold03 Fold04 Fold05 Fold06 Fold07 Fold08 Fold09 Fold10
    19     19     18     18     19     18     20     16     16     20 </code></pre>
</div>
<p>The folds are returned as a list of numeric indices. The first fold
of data is therefore:</p>
<div class="codewrapper sourceCode" id="cb11">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">y</span><span class="op">[</span><span class="va">idx</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">]</span> <span class="co">##the labels</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code> [1] "kidney"      "hippocampus" "hippocampus" "hippocampus" "cerebellum"
 [6] "cerebellum"  "cerebellum"  "kidney"      "kidney"      "kidney"
[11] "colon"       "colon"       "colon"       "liver"       "endometrium"
[16] "endometrium" "liver"       "liver"       "cerebellum" </code></pre>
</div>
<div class="codewrapper sourceCode" id="cb13">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">head</span><span class="op">(</span> <span class="va">X</span><span class="op">[</span><span class="va">idx</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">]</span> <span class="op">)</span> <span class="co">##the genes (only showing the first 3 genes...)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>                1007_s_at  1053_at   117_at
GSM12105.CEL.gz  9.913031 6.337478 7.983850
GSM21209.cel.gz 11.667431 5.785190 7.666343
GSM21215.cel.gz 10.361340 5.663634 7.328729
GSM21218.cel.gz 10.757734 5.984170 8.525524
GSM87066.cel.gz  9.746007 5.886079 7.459517
GSM87085.cel.gz  9.864295 5.753874 7.712646</code></pre>
</div>
<p>We can see that, in fact, the tissues are fairly equally represented
across the 10 folds:</p>
<div class="codewrapper sourceCode" id="cb15">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">sapply</span><span class="op">(</span><span class="va">idx</span>, <span class="kw">function</span><span class="op">(</span><span class="va">i</span><span class="op">)</span> <span class="fu">table</span><span class="op">(</span><span class="va">y</span><span class="op">[</span><span class="va">i</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>            Fold01 Fold02 Fold03 Fold04 Fold05 Fold06 Fold07 Fold08 Fold09
cerebellum       4      4      4      4      4      3      4      3      4
colon            3      4      4      3      3      3      4      3      3
endometrium      2      1      1      1      2      2      2      1      1
hippocampus      3      3      3      3      3      3      3      3      3
kidney           4      4      4      4      4      4      4      4      3
liver            3      3      2      3      3      3      3      2      2
            Fold10
cerebellum       4
colon            4
endometrium      2
hippocampus      4
kidney           4
liver            2</code></pre>
</div>
<p>Because tissues have very different gene expression profiles,
predicting tissue with all genes will be very easy. For illustration
purposes we will try to predict tissue type with just two dimensional
data. We will reduce the dimension of our data using
<code>cmdscale</code>:</p>
<div class="codewrapper sourceCode" id="cb17">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">rafalib</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>
Attaching package: 'rafalib'</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>The following object is masked from 'package:lattice':

    stripplot</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb20">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">mypar</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">Xsmall</span> <span class="op">&lt;-</span> <span class="fu">cmdscale</span><span class="op">(</span><span class="fu">dist</span><span class="op">(</span><span class="va">X</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">plot</span><span class="op">(</span><span class="va">Xsmall</span>,col<span class="op">=</span><span class="fu">as.fumeric</span><span class="op">(</span><span class="va">y</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">legend</span><span class="op">(</span><span class="st">"topleft"</span>,<span class="fu">levels</span><span class="op">(</span><span class="fu">factor</span><span class="op">(</span><span class="va">y</span><span class="op">)</span><span class="op">)</span>,fill<span class="op">=</span><span class="fu">seq_along</span><span class="op">(</span><span class="fu">levels</span><span class="op">(</span><span class="fu">factor</span><span class="op">(</span><span class="va">y</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="../fig/cross-validation-rendered-mds-1.png" alt="First two PCs of the tissue gene expression data with color representing tissue. We use these two PCs as our two predictors throughout." class="figure mx-auto d-block"><figcaption>
First two PCs of the tissue gene expression data with color representing
tissue. We use these two PCs as our two predictors throughout.
</figcaption></figure><p>Now we can try out the k-nearest neighbors method on a single fold.
We provide the <code>knn</code> function with all the samples in
<code>Xsmall</code> <em>except</em> those which are in the first fold.
We remove these samples using the code <code>-idx[[1]]</code> inside the
square brackets. We then use those samples in the test set. The
<code>cl</code> argument is for the true classifications or labels
(here, tissue) of the training data. We use 5 observations to classify
in our k-nearest neighbor algorithm:</p>
<div class="codewrapper sourceCode" id="cb21">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pred</span> <span class="op">&lt;-</span> <span class="fu">knn</span><span class="op">(</span>train<span class="op">=</span><span class="va">Xsmall</span><span class="op">[</span> <span class="op">-</span><span class="va">idx</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span> , <span class="op">]</span>, test<span class="op">=</span><span class="va">Xsmall</span><span class="op">[</span> <span class="va">idx</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>, <span class="op">]</span>, cl<span class="op">=</span><span class="va">y</span><span class="op">[</span> <span class="op">-</span><span class="va">idx</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span> <span class="op">]</span>, k<span class="op">=</span><span class="fl">5</span><span class="op">)</span></span>
<span><span class="fu">table</span><span class="op">(</span>true<span class="op">=</span><span class="va">y</span><span class="op">[</span> <span class="va">idx</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span> <span class="op">]</span>, <span class="va">pred</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>             pred
true          cerebellum colon endometrium hippocampus kidney liver
  cerebellum           4     0           0           0      0     0
  colon                0     2           0           0      1     0
  endometrium          0     0           2           0      0     0
  hippocampus          2     0           0           1      0     0
  kidney               0     0           0           0      4     0
  liver                0     0           0           0      0     3</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb23">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">mean</span><span class="op">(</span><span class="va">y</span><span class="op">[</span> <span class="va">idx</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span> <span class="op">]</span> <span class="op">!=</span> <span class="va">pred</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 0.1578947</code></pre>
</div>
<p>Now we have some misclassifications. How well do we do for the rest
of the folds?</p>
<div class="codewrapper sourceCode" id="cb25">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">pred</span> <span class="op">&lt;-</span> <span class="fu">knn</span><span class="op">(</span>train<span class="op">=</span><span class="va">Xsmall</span><span class="op">[</span> <span class="op">-</span><span class="va">idx</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span> , <span class="op">]</span>, test<span class="op">=</span><span class="va">Xsmall</span><span class="op">[</span> <span class="va">idx</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span>, <span class="op">]</span>, cl<span class="op">=</span><span class="va">y</span><span class="op">[</span> <span class="op">-</span><span class="va">idx</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span> <span class="op">]</span>, k<span class="op">=</span><span class="fl">5</span><span class="op">)</span></span>
<span>  <span class="fu">print</span><span class="op">(</span><span class="fu">paste0</span><span class="op">(</span><span class="va">i</span>,<span class="st">") error rate: "</span>, <span class="fu">round</span><span class="op">(</span><span class="fu">mean</span><span class="op">(</span><span class="va">y</span><span class="op">[</span> <span class="va">idx</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span> <span class="op">]</span> <span class="op">!=</span> <span class="va">pred</span><span class="op">)</span>,<span class="fl">3</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "1) error rate: 0.158"
[1] "2) error rate: 0.158"
[1] "3) error rate: 0.278"
[1] "4) error rate: 0.167"
[1] "5) error rate: 0.158"
[1] "6) error rate: 0.167"
[1] "7) error rate: 0.25"
[1] "8) error rate: 0.188"
[1] "9) error rate: 0.062"
[1] "10) error rate: 0.1"</code></pre>
</div>
<p>So we can see there is some variation for each fold, with error rates
hovering around 0.1-0.3. But is <code>k=5</code> the best setting for
the k parameter? In order to explore the best setting for k, we need to
create an outer loop, where we try different values for k, and then
calculate the average test set error across all the folds.</p>
<p>We will try out each value of k from 1 to 12. Instead of using two
<code>for</code> loops, we will use <code>sapply</code>:</p>
<div class="codewrapper sourceCode" id="cb27">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">set.seed</span><span class="op">(</span><span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">ks</span> <span class="op">&lt;-</span> <span class="fl">1</span><span class="op">:</span><span class="fl">12</span></span>
<span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu">sapply</span><span class="op">(</span><span class="va">ks</span>, <span class="kw">function</span><span class="op">(</span><span class="va">k</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="co">##try out each version of k from 1 to 12</span></span>
<span>  <span class="va">res.k</span> <span class="op">&lt;-</span> <span class="fu">sapply</span><span class="op">(</span><span class="fu">seq_along</span><span class="op">(</span><span class="va">idx</span><span class="op">)</span>, <span class="kw">function</span><span class="op">(</span><span class="va">i</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="co">##loop over each of the 10 cross-validation folds</span></span>
<span>    <span class="co">##predict the held-out samples using k nearest neighbors</span></span>
<span>    <span class="va">pred</span> <span class="op">&lt;-</span> <span class="fu">knn</span><span class="op">(</span>train<span class="op">=</span><span class="va">Xsmall</span><span class="op">[</span> <span class="op">-</span><span class="va">idx</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span>, <span class="op">]</span>,</span>
<span>                test<span class="op">=</span><span class="va">Xsmall</span><span class="op">[</span> <span class="va">idx</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span>, <span class="op">]</span>,</span>
<span>                cl<span class="op">=</span><span class="va">y</span><span class="op">[</span> <span class="op">-</span><span class="va">idx</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span> <span class="op">]</span>, k <span class="op">=</span> <span class="va">k</span><span class="op">)</span></span>
<span>    <span class="co">##the ratio of misclassified samples</span></span>
<span>    <span class="fu">mean</span><span class="op">(</span><span class="va">y</span><span class="op">[</span> <span class="va">idx</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span> <span class="op">]</span> <span class="op">!=</span> <span class="va">pred</span><span class="op">)</span></span>
<span>  <span class="op">}</span><span class="op">)</span></span>
<span>  <span class="co">##average over the 10 folds</span></span>
<span>  <span class="fu">mean</span><span class="op">(</span><span class="va">res.k</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span></code></pre>
</div>
<p>Now for each value of k, we have an associated test set error rate
from the cross-validation procedure.</p>
<div class="codewrapper sourceCode" id="cb28">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code> [1] 0.1882164 0.1692032 0.1694664 0.1639108 0.1511184 0.1586184 0.1589108
 [8] 0.1865058 0.1880482 0.1714108 0.1759795 0.1744664</code></pre>
</div>
<p>We can then plot the error rate for each value of k, which helps us
to see in what region there might be a minimal error rate:</p>
<div class="codewrapper sourceCode" id="cb30">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">plot</span><span class="op">(</span><span class="va">ks</span>, <span class="va">res</span>, type<span class="op">=</span><span class="st">"o"</span>, ylab<span class="op">=</span><span class="st">"misclassification error"</span><span class="op">)</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="../fig/cross-validation-rendered-misclassification_error-1.png" alt="Misclassification error versus number of neighbors." class="figure mx-auto d-block"><figcaption>
Misclassification error versus number of neighbors.
</figcaption></figure><p>Remember, because the training set is a random sample and because our
fold-generation procedure involves random number generation, the “best”
value of k we pick through this procedure is also a random variable. If
we had new training data and if we recreated our folds, we might get a
different value for the optimal k.</p>
<p>Finally, to show that gene expression can perfectly predict tissue,
we use 5 dimensions instead of 2, which results in perfect
prediction:</p>
<div class="codewrapper sourceCode" id="cb31">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">Xsmall</span> <span class="op">&lt;-</span> <span class="fu">cmdscale</span><span class="op">(</span><span class="fu">dist</span><span class="op">(</span><span class="va">X</span><span class="op">)</span>,k<span class="op">=</span><span class="fl">5</span><span class="op">)</span></span>
<span><span class="fu">set.seed</span><span class="op">(</span><span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">ks</span> <span class="op">&lt;-</span> <span class="fl">1</span><span class="op">:</span><span class="fl">12</span></span>
<span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu">sapply</span><span class="op">(</span><span class="va">ks</span>, <span class="kw">function</span><span class="op">(</span><span class="va">k</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">res.k</span> <span class="op">&lt;-</span> <span class="fu">sapply</span><span class="op">(</span><span class="fu">seq_along</span><span class="op">(</span><span class="va">idx</span><span class="op">)</span>, <span class="kw">function</span><span class="op">(</span><span class="va">i</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">pred</span> <span class="op">&lt;-</span> <span class="fu">knn</span><span class="op">(</span>train<span class="op">=</span><span class="va">Xsmall</span><span class="op">[</span> <span class="op">-</span><span class="va">idx</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span>, <span class="op">]</span>,</span>
<span>                test<span class="op">=</span><span class="va">Xsmall</span><span class="op">[</span> <span class="va">idx</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span>, <span class="op">]</span>,</span>
<span>                cl<span class="op">=</span><span class="va">y</span><span class="op">[</span> <span class="op">-</span><span class="va">idx</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span> <span class="op">]</span>, k <span class="op">=</span> <span class="va">k</span><span class="op">)</span></span>
<span>    <span class="fu">mean</span><span class="op">(</span><span class="va">y</span><span class="op">[</span> <span class="va">idx</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span> <span class="op">]</span> <span class="op">!=</span> <span class="va">pred</span><span class="op">)</span></span>
<span>  <span class="op">}</span><span class="op">)</span></span>
<span>  <span class="fu">mean</span><span class="op">(</span><span class="va">res.k</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span>
<span><span class="fu">plot</span><span class="op">(</span><span class="va">ks</span>, <span class="va">res</span>, type<span class="op">=</span><span class="st">"o"</span>,ylim<span class="op">=</span><span class="fu">c</span><span class="op">(</span><span class="fl">0</span>,<span class="fl">0.20</span><span class="op">)</span>,ylab<span class="op">=</span><span class="st">"misclassification error"</span><span class="op">)</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="../fig/cross-validation-rendered-misclassification_error2-1.png" alt="Misclassification error versus number of neighbors when we use 5 dimensions instead of 2." class="figure mx-auto d-block"><figcaption>
Misclassification error versus number of neighbors when we use 5
dimensions instead of 2.
</figcaption></figure><p>Important note: we applied <code>cmdscale</code> to the entire
dataset to create a smaller one for illustration purposes. However, in a
real machine learning application, this may result in an underestimation
of test set error for small sample sizes, where dimension reduction
using the unlabeled full dataset gives a boost in performance. A safer
choice would have been to transform the data separately for each fold,
by calculating a rotation and dimension reduction using the training set
only and applying this to the test set.</p>
<div id="exercise-1" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="exercise-1" class="callout-inner">
<h3 class="callout-title">Exercise 1</h3>
<div class="callout-content">
<p>Load the following dataset:</p>
<div class="codewrapper sourceCode" id="cb32">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">GSE5859Subset</span><span class="op">)</span></span>
<span><span class="fu">data</span><span class="op">(</span><span class="va">GSE5859Subset</span><span class="op">)</span></span></code></pre>
</div>
<p>And define the outcome and predictors. To make the problem more
difficult, we will only consider autosomal genes:</p>
<div class="codewrapper sourceCode" id="cb33">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">y</span> <span class="op">=</span> <span class="fu">factor</span><span class="op">(</span><span class="va">sampleInfo</span><span class="op">$</span><span class="va">group</span><span class="op">)</span></span>
<span><span class="va">X</span> <span class="op">=</span> <span class="fu">t</span><span class="op">(</span><span class="va">geneExpression</span><span class="op">)</span></span>
<span><span class="va">out</span> <span class="op">=</span> <span class="fu">which</span><span class="op">(</span><span class="va">geneAnnotation</span><span class="op">$</span><span class="va">CHR</span> <span class="op">%in%</span> <span class="fu">c</span><span class="op">(</span><span class="st">"chrX"</span>,<span class="st">"chrY"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">X</span> <span class="op">=</span> <span class="va">X</span><span class="op">[</span>,<span class="op">-</span><span class="va">out</span><span class="op">]</span></span></code></pre>
</div>
<p>Use the <code>createFold</code> function in the <code>caret</code>
package, set the seed to 1 <code>set.seed(1)</code> and create 10 folds
of <code>y</code>. What is the 2nd entry in the fold 3?</p>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1"> Show me the solution </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" data-bs-parent="#accordionSolution1" aria-labelledby="headingSolution1">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb34">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">caret</span><span class="op">)</span></span>
<span><span class="fu">set.seed</span><span class="op">(</span><span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">idx</span> <span class="op">&lt;-</span> <span class="fu">createFolds</span><span class="op">(</span><span class="va">y</span>, k <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span>
<span><span class="co"># Select the set of indices corresponding to the third fold</span></span>
<span><span class="co"># using idx[[3]], then print the second index of that fold</span></span>
<span><span class="va">idx</span><span class="op">[[</span><span class="fl">3</span><span class="op">]</span><span class="op">]</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span></span></code></pre>
</div>
</div>
</div>
</div>
</div>
<div id="exercise-2" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="exercise-2" class="callout-inner">
<h3 class="callout-title">Exercise 2</h3>
<div class="callout-content">
<p>We are going to use kNN. We are going to consider a smaller set of
predictors by using filtering gene using t-tests. Specifically, we will
perform a t-test and select the <span class="math inline">\(m\)</span>
genes with the smallest p-values.</p>
<p>Let m = 8 and k = 5 and train kNN by leaving out the second fold
<code>idx[[2]]</code>. How many mistakes do we make on the test set?
Remember it is indispensable that you perform the t-test on the training
data. <strong>Use all 10 folds, keep k = 5. Hint: be careful about
indexing.</strong></p>
</div>
</div>
</div>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2"> Show me the solution </h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" data-bs-parent="#accordionSolution2" aria-labelledby="headingSolution2">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb35">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">genefilter</span><span class="op">)</span></span>
<span><span class="va">m</span> <span class="op">&lt;-</span> <span class="fl">8</span> <span class="co"># number of genes</span></span>
<span></span>
<span><span class="co"># `rowttests` performs a t-test on the expression of each gene</span></span>
<span><span class="va">pvals</span> <span class="op">&lt;-</span> <span class="fu">rowttests</span><span class="op">(</span><span class="fu">t</span><span class="op">(</span><span class="va">X</span><span class="op">[</span><span class="op">-</span><span class="va">idx</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span>,<span class="op">]</span><span class="op">)</span>,<span class="va">y</span><span class="op">[</span><span class="op">-</span><span class="va">idx</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">$</span><span class="va">p.value</span></span>
<span></span>
<span><span class="co"># We use the p-value to identify the genes that present a</span></span>
<span><span class="co"># significant effect (i.e. the effect is statistically</span></span>
<span><span class="co"># different from 0). That is achieved by ordering `pvals`</span></span>
<span><span class="co"># in increasing order, and taking the `m` smallest p-values</span></span>
<span><span class="va">ind</span> <span class="op">&lt;-</span> <span class="fu">order</span><span class="op">(</span><span class="va">pvals</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="va">m</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># Then the k-nearest-neighbor algorithm is executed only</span></span>
<span><span class="co"># considering these `m` most significant genes.</span></span>
<span><span class="va">pred</span> <span class="op">&lt;-</span> <span class="fu">knn</span><span class="op">(</span>train <span class="op">=</span> <span class="va">X</span><span class="op">[</span><span class="op">-</span><span class="va">idx</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span>,<span class="va">ind</span><span class="op">]</span>,</span>
<span>            test <span class="op">=</span> <span class="va">X</span><span class="op">[</span><span class="va">idx</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span>,<span class="va">ind</span><span class="op">]</span>,</span>
<span>            cl <span class="op">=</span> <span class="va">y</span><span class="op">[</span><span class="op">-</span><span class="va">idx</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span><span class="op">]</span>, k <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># This command computes the total number of examples that</span></span>
<span><span class="co"># were miss-classified by the knn algorithm.</span></span>
<span><span class="fu">sum</span><span class="op">(</span><span class="va">pred</span> <span class="op">!=</span> <span class="va">y</span><span class="op">[</span><span class="va">idx</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span></code></pre>
</div>
</div>
</div>
</div>
</div>
<div id="exercise-3" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="exercise-3" class="callout-inner">
<h3 class="callout-title">Exercise 3</h3>
<div class="callout-content">
<p>Now run through all 5 folds. What is our error rate? <strong>(total
number of errors / total predictions)</strong></p>
</div>
</div>
</div>
<div id="accordionSolution3" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution3" aria-expanded="false" aria-controls="collapseSolution3">
  <h4 class="accordion-header" id="headingSolution3"> Show me the solution </h4>
</button>
<div id="collapseSolution3" class="accordion-collapse collapse" data-bs-parent="#accordionSolution3" aria-labelledby="headingSolution3">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb36">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Now run the previous piece of code for each fold</span></span>
<span><span class="va">n_fold</span> <span class="op">&lt;-</span> <span class="fu">length</span><span class="op">(</span><span class="va">idx</span><span class="op">)</span></span>
<span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu">vector</span><span class="op">(</span><span class="st">'double'</span>, <span class="va">n_fold</span><span class="op">)</span></span>
<span><span class="va">m</span> <span class="op">&lt;-</span> <span class="fl">8</span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fu">seq</span><span class="op">(</span><span class="va">n_fold</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="co"># To be fair and only use the information we have at the moment</span></span>
<span>  <span class="co"># of training, we perform the t-tests only on the training set.</span></span>
<span>  <span class="va">pvals</span> <span class="op">&lt;-</span> <span class="fu">rowttests</span><span class="op">(</span><span class="fu">t</span><span class="op">(</span><span class="va">X</span><span class="op">[</span><span class="op">-</span><span class="va">idx</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span>,<span class="op">]</span><span class="op">)</span>,<span class="va">y</span><span class="op">[</span><span class="op">-</span><span class="va">idx</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">$</span><span class="va">p.value</span></span>
<span>  <span class="va">ind</span> <span class="op">&lt;-</span> <span class="fu">order</span><span class="op">(</span><span class="va">pvals</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="va">m</span><span class="op">]</span></span>
<span>  <span class="co"># We will use again the top m=8 genes that showed the most</span></span>
<span>  <span class="co"># significant effect according to the previous t-tests.</span></span>
<span>  <span class="va">pred</span> <span class="op">&lt;-</span> <span class="fu">knn</span><span class="op">(</span>train <span class="op">=</span> <span class="va">X</span><span class="op">[</span><span class="op">-</span><span class="va">idx</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span>,<span class="va">ind</span><span class="op">]</span>,</span>
<span>              test <span class="op">=</span> <span class="va">X</span><span class="op">[</span><span class="va">idx</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span>,<span class="va">ind</span><span class="op">]</span>,</span>
<span>              cl <span class="op">=</span> <span class="va">y</span><span class="op">[</span><span class="op">-</span><span class="va">idx</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">]</span>, k <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span>
<span>  <span class="va">res</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu">sum</span><span class="op">(</span><span class="va">pred</span> <span class="op">!=</span> <span class="va">y</span><span class="op">[</span><span class="va">idx</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="co"># Compute the average performance of the knn algorithm achieved on</span></span>
<span><span class="co"># the ten folds.</span></span>
<span><span class="fu">sum</span><span class="op">(</span><span class="va">res</span><span class="op">)</span><span class="op">/</span><span class="fu">length</span><span class="op">(</span><span class="va">y</span><span class="op">)</span></span></code></pre>
</div>
</div>
</div>
</div>
</div>
<div id="exercise-4" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="exercise-4" class="callout-inner">
<h3 class="callout-title">Exercise 4</h3>
<div class="callout-content">
<p>Now we are going to select the best values of k and m. Use the expand
grid function to try out the following values:</p>
<div class="codewrapper sourceCode" id="cb37">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ms</span> <span class="op">=</span> <span class="fl">2</span><span class="op">^</span><span class="fu">c</span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">11</span><span class="op">)</span></span>
<span><span class="va">ks</span> <span class="op">=</span> <span class="fu">seq</span><span class="op">(</span><span class="fl">1</span>,<span class="fl">9</span>,<span class="fl">2</span><span class="op">)</span></span>
<span><span class="co"># Compute all possible pairs of configurations of number of most</span></span>
<span><span class="co"># significant genes and number of neighbors for the knn algorithm.</span></span>
<span><span class="va">params</span> <span class="op">=</span> <span class="fu">expand.grid</span><span class="op">(</span>k<span class="op">=</span><span class="va">ks</span>, m<span class="op">=</span><span class="va">ms</span><span class="op">)</span></span></code></pre>
</div>
<p>Now use apply or a for-loop to obtain error rates for each of these
pairs of parameters. Which pair of parameters minimizes the error
rate?</p>
</div>
</div>
</div>
<div id="accordionSolution4" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution4" aria-expanded="false" aria-controls="collapseSolution4">
  <h4 class="accordion-header" id="headingSolution4"> Show me the solution </h4>
</button>
<div id="collapseSolution4" class="accordion-collapse collapse" data-bs-parent="#accordionSolution4" aria-labelledby="headingSolution4">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb38">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">n_fold</span> <span class="op">&lt;-</span> <span class="fu">length</span><span class="op">(</span><span class="va">idx</span><span class="op">)</span></span>
<span><span class="co"># Store the mean performance on the ten folds of the knn algorithm</span></span>
<span><span class="co"># using each pair of parameters.</span></span>
<span><span class="va">error_rate_avg</span> <span class="op">=</span> <span class="fu">vector</span><span class="op">(</span><span class="st">'double'</span>,<span class="fu">nrow</span><span class="op">(</span><span class="va">params</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co"># Iterate over each pair of parameters:</span></span>
<span><span class="co"># (number of neighbors, number of genes)</span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">j</span> <span class="kw">in</span> <span class="fu">seq</span><span class="op">(</span><span class="fu">nrow</span><span class="op">(</span><span class="va">params</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="co"># Iterate over each fold</span></span>
<span>    <span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fu">seq</span><span class="op">(</span><span class="va">n_fold</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>        <span class="co"># Again perform the t-tests only on the training set</span></span>
<span>        <span class="va">pvals</span> <span class="op">&lt;-</span> <span class="fu">rowttests</span><span class="op">(</span><span class="fu">t</span><span class="op">(</span><span class="va">X</span><span class="op">[</span><span class="op">-</span><span class="va">idx</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span>,<span class="op">]</span><span class="op">)</span>,<span class="va">y</span><span class="op">[</span><span class="op">-</span><span class="va">idx</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">$</span><span class="va">p.value</span></span>
<span>        <span class="co"># This time we take the top number of genes given by</span></span>
<span>        <span class="co"># the current pair of parameters `param[j,][[2]]`</span></span>
<span>        <span class="va">ind</span> <span class="op">&lt;-</span> <span class="fu">order</span><span class="op">(</span><span class="va">pvals</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="va">params</span><span class="op">[</span><span class="va">j</span>,<span class="op">]</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span><span class="op">]</span></span>
<span></span>
<span>        <span class="co"># Train the knn algorithm using the train set, and</span></span>
<span>        <span class="co"># evaluating on the test set of the current fold `idx[[i]]`.</span></span>
<span>        <span class="co"># The knn is trained using the number of neighbors given</span></span>
<span>        <span class="co"># by the current pair of parameters `param[j,][[1]]`</span></span>
<span>        <span class="va">pred</span> <span class="op">&lt;-</span> <span class="fu">knn</span><span class="op">(</span>train <span class="op">=</span> <span class="va">X</span><span class="op">[</span><span class="op">-</span><span class="va">idx</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span>,<span class="va">ind</span><span class="op">]</span>,</span>
<span>                    test <span class="op">=</span> <span class="va">X</span><span class="op">[</span><span class="va">idx</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span>,<span class="va">ind</span><span class="op">]</span>,</span>
<span>                    cl <span class="op">=</span> <span class="va">y</span><span class="op">[</span><span class="op">-</span><span class="va">idx</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">]</span>,</span>
<span>                    k <span class="op">=</span> <span class="va">params</span><span class="op">[</span><span class="va">j</span>,<span class="op">]</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span>        <span class="va">res</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu">sum</span><span class="op">(</span><span class="va">pred</span> <span class="op">!=</span> <span class="va">y</span><span class="op">[</span><span class="va">idx</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span>    <span class="op">}</span></span>
<span>    <span class="co"># Approximate how our knn algorithm would perform with unseen data</span></span>
<span>    <span class="co"># by computing the mean error achieved on all the folds.</span></span>
<span>    <span class="va">error_rate_avg</span><span class="op">[[</span><span class="va">j</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu">sum</span><span class="op">(</span><span class="va">res</span><span class="op">)</span><span class="op">/</span><span class="fu">length</span><span class="op">(</span><span class="va">y</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="co"># Find the pair of parameters (number of neighbors, number of genes)</span></span>
<span><span class="co"># that achieves the lowest expected error.</span></span>
<span><span class="va">ind</span> <span class="op">&lt;-</span> <span class="fu">which</span><span class="op">(</span><span class="va">error_rate_avg</span> <span class="op">==</span> <span class="fu">min</span><span class="op">(</span><span class="va">error_rate_avg</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Print that pair of parameters and the corresponding error rate</span></span>
<span><span class="co"># achieved</span></span>
<span><span class="va">params</span><span class="op">[</span><span class="va">ind</span>,<span class="op">]</span> <span class="co"># answer</span></span>
<span><span class="fu">min</span><span class="op">(</span><span class="va">error_rate_avg</span><span class="op">)</span> <span class="co"># minimum error rate</span></span></code></pre>
</div>
</div>
</div>
</div>
</div>
<div id="exercise-5" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="exercise-5" class="callout-inner">
<h3 class="callout-title">Exercise 5</h3>
<div class="callout-content">
<p>Repeat exercise 4, but now perform the t-test filtering before the
cross validation. Note how this biases the entire result and gives us
much lower estimated error rates.</p>
</div>
</div>
</div>
<div id="accordionSolution5" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution5" aria-expanded="false" aria-controls="collapseSolution5">
  <h4 class="accordion-header" id="headingSolution5"> Show me the solution </h4>
</button>
<div id="collapseSolution5" class="accordion-collapse collapse" data-bs-parent="#accordionSolution5" aria-labelledby="headingSolution5">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb39">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># We perform the same experiment with the exception that</span></span>
<span><span class="co"># this time we perform the t-tests on all the dataset to</span></span>
<span><span class="co"># choose the top number of genes with most significant</span></span>
<span><span class="co"># effect.</span></span>
<span><span class="va">ms</span> <span class="op">=</span> <span class="fl">2</span><span class="op">^</span><span class="fu">c</span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">11</span><span class="op">)</span></span>
<span><span class="va">ks</span> <span class="op">=</span> <span class="fu">seq</span><span class="op">(</span><span class="fl">1</span>,<span class="fl">9</span>,<span class="fl">2</span><span class="op">)</span></span>
<span><span class="va">params</span> <span class="op">=</span> <span class="fu">expand.grid</span><span class="op">(</span>k<span class="op">=</span><span class="va">ks</span>, m<span class="op">=</span><span class="va">ms</span><span class="op">)</span></span>
<span><span class="va">n_fold</span> <span class="op">&lt;-</span> <span class="fu">length</span><span class="op">(</span><span class="va">idx</span><span class="op">)</span></span>
<span><span class="va">error_rate_avg</span> <span class="op">=</span> <span class="fu">vector</span><span class="op">(</span><span class="st">'double'</span>,<span class="fu">nrow</span><span class="op">(</span><span class="va">params</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">j</span> <span class="kw">in</span> <span class="fu">seq</span><span class="op">(</span><span class="fu">nrow</span><span class="op">(</span><span class="va">params</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fu">seq</span><span class="op">(</span><span class="va">n_fold</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>        <span class="co"># Here we use the complete dataset to select the genes,</span></span>
<span>        <span class="co"># rather than only the examples corresponding to this fold.</span></span>
<span>        <span class="va">pvals</span> <span class="op">&lt;-</span> <span class="fu">rowttests</span><span class="op">(</span><span class="fu">t</span><span class="op">(</span><span class="va">X</span><span class="op">)</span>,<span class="va">y</span><span class="op">)</span><span class="op">$</span><span class="va">p.value</span></span>
<span>        <span class="va">ind</span> <span class="op">&lt;-</span> <span class="fu">order</span><span class="op">(</span><span class="va">pvals</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="va">params</span><span class="op">[</span><span class="va">j</span>,<span class="op">]</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span><span class="op">]</span></span>
<span>        <span class="va">pred</span> <span class="op">&lt;-</span> <span class="fu">knn</span><span class="op">(</span>train <span class="op">=</span> <span class="va">X</span><span class="op">[</span><span class="op">-</span><span class="va">idx</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span>,<span class="va">ind</span><span class="op">]</span>,</span>
<span>                    test <span class="op">=</span> <span class="va">X</span><span class="op">[</span><span class="va">idx</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span>,<span class="va">ind</span><span class="op">]</span>,</span>
<span>                    cl <span class="op">=</span> <span class="va">y</span><span class="op">[</span><span class="op">-</span><span class="va">idx</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">]</span>, k <span class="op">=</span> <span class="va">params</span><span class="op">[</span><span class="va">j</span>,<span class="op">]</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span>        <span class="va">res</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu">sum</span><span class="op">(</span><span class="va">pred</span> <span class="op">!=</span> <span class="va">y</span><span class="op">[</span><span class="va">idx</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span>    <span class="op">}</span></span>
<span>    <span class="va">error_rate_avg</span><span class="op">[[</span><span class="va">j</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu">sum</span><span class="op">(</span><span class="va">res</span><span class="op">)</span><span class="op">/</span><span class="fu">length</span><span class="op">(</span><span class="va">y</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="fu">min</span><span class="op">(</span><span class="va">error_rate_avg</span><span class="op">)</span> <span class="co"># minimum error rate</span></span>
<span><span class="fu">mean</span><span class="op">(</span><span class="va">error_rate_avg</span><span class="op">)</span> <span class="co"># mean error rate</span></span></code></pre>
</div>
<p>The error rate is much lower than the one in <strong>Exercise
4</strong> because we did not filter out p values from the test data in
this case. This is not a correct practice. The practice shown in
<strong>Exercise 4</strong> is correct.</p>
</div>
</div>
</div>
</div>
<div id="exercise-6" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="exercise-6" class="callout-inner">
<h3 class="callout-title">Exercise 6</h3>
<div class="callout-content">
<p>Repeat exercise 3, but now, instead of <code>sampleInfo$group</code>,
use
<code>y = factor(as.numeric(format( sampleInfo$date, "%m")=="06"))</code>
What is the minimum error rate now? We achieve much lower error rates
when predicting date than when predicting the group. Because group is
confounded with date, it is very possible that these predictors have no
information about group and that our lower 0.5 error rates are due to
the confounding with date. We will learn more about this in the batch
effect chapter.</p>
</div>
</div>
</div>
<div id="accordionSolution6" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution6" aria-expanded="false" aria-controls="collapseSolution6">
  <h4 class="accordion-header" id="headingSolution6"> Show me the solution </h4>
</button>
<div id="collapseSolution6" class="accordion-collapse collapse" data-bs-parent="#accordionSolution6" aria-labelledby="headingSolution6">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb40">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># We use 'y' as the date the sample was taken as class.</span></span>
<span><span class="co"># However, notice that this is introducing a `batch effect`.</span></span>
<span><span class="va">y</span> <span class="op">=</span> <span class="fu">factor</span><span class="op">(</span><span class="fu">as.numeric</span><span class="op">(</span><span class="fu">format</span><span class="op">(</span> <span class="va">sampleInfo</span><span class="op">$</span><span class="va">date</span>, <span class="st">"%m"</span><span class="op">)</span><span class="op">==</span><span class="st">"06"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">ms</span> <span class="op">=</span> <span class="fl">2</span><span class="op">^</span><span class="fu">c</span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">11</span><span class="op">)</span></span>
<span><span class="va">ks</span> <span class="op">=</span> <span class="fu">seq</span><span class="op">(</span><span class="fl">1</span>,<span class="fl">9</span>,<span class="fl">2</span><span class="op">)</span></span>
<span><span class="va">params</span> <span class="op">=</span> <span class="fu">expand.grid</span><span class="op">(</span>k<span class="op">=</span><span class="va">ks</span>, m<span class="op">=</span><span class="va">ms</span><span class="op">)</span></span>
<span><span class="va">n_fold</span> <span class="op">&lt;-</span> <span class="fu">length</span><span class="op">(</span><span class="va">idx</span><span class="op">)</span></span>
<span><span class="va">error_rate_avg</span> <span class="op">=</span> <span class="fu">vector</span><span class="op">(</span><span class="st">'double'</span>,<span class="fu">nrow</span><span class="op">(</span><span class="va">params</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">j</span> <span class="kw">in</span> <span class="fu">seq</span><span class="op">(</span><span class="fu">nrow</span><span class="op">(</span><span class="va">params</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fu">seq</span><span class="op">(</span><span class="va">n_fold</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>        <span class="va">pvals</span> <span class="op">&lt;-</span> <span class="fu">rowttests</span><span class="op">(</span><span class="fu">t</span><span class="op">(</span><span class="va">X</span><span class="op">[</span><span class="op">-</span><span class="va">idx</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span>,<span class="op">]</span><span class="op">)</span>,<span class="va">y</span><span class="op">[</span><span class="op">-</span><span class="va">idx</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">$</span><span class="va">p.value</span></span>
<span>        <span class="va">ind</span> <span class="op">&lt;-</span> <span class="fu">order</span><span class="op">(</span><span class="va">pvals</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="va">params</span><span class="op">[</span><span class="va">j</span>,<span class="op">]</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span><span class="op">]</span></span>
<span>        <span class="va">pred</span> <span class="op">&lt;-</span> <span class="fu">knn</span><span class="op">(</span>train <span class="op">=</span> <span class="va">X</span><span class="op">[</span><span class="op">-</span><span class="va">idx</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span>,<span class="va">ind</span><span class="op">]</span>,</span>
<span>                    test <span class="op">=</span> <span class="va">X</span><span class="op">[</span><span class="va">idx</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span>,<span class="va">ind</span><span class="op">]</span>,</span>
<span>                    cl <span class="op">=</span> <span class="va">y</span><span class="op">[</span><span class="op">-</span><span class="va">idx</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">]</span>, k <span class="op">=</span> <span class="va">params</span><span class="op">[</span><span class="va">j</span>,<span class="op">]</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span>        <span class="va">res</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu">sum</span><span class="op">(</span><span class="va">pred</span> <span class="op">!=</span> <span class="va">y</span><span class="op">[</span><span class="va">idx</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span>    <span class="op">}</span></span>
<span>    <span class="va">error_rate_avg</span><span class="op">[[</span><span class="va">j</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu">sum</span><span class="op">(</span><span class="va">res</span><span class="op">)</span><span class="op">/</span><span class="fu">length</span><span class="op">(</span><span class="va">y</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="fu">min</span><span class="op">(</span><span class="va">error_rate_avg</span><span class="op">)</span> <span class="co"># minimum error rate</span></span>
<span><span class="fu">mean</span><span class="op">(</span><span class="va">error_rate_avg</span><span class="op">)</span> <span class="co"># mean error rate</span></span></code></pre>
</div>
</div>
</div>
</div>
</div>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Key Points</h3>
<div class="callout-content">
<ul>
<li>The mean validation error obtained from cross-validation is a better
approximation of the test error (real world data) than the training
error itself.</li>
</ul>
</div>
</div>
</div>
<!--
Place links that you need to refer to multiple times across pages here. Delete
any links that you are not going to use.
 -->
</section></section>
</div>
    </main>
</div>
<!-- END  : inst/pkgdown/templates/content-extra.html -->

      </div>
<!--/div.row-->
      		<footer class="row footer mx-md-3"><hr>
<div class="col-md-6">
        <p>This lesson is subject to the <a href="CODE_OF_CONDUCT.html">Code of Conduct</a></p>
        <p>

        <a href="https://github.com/smcclatchy/machine-learning-biomedical-science/edit/main/README.md" class="external-link">Edit on GitHub</a>

	
        | <a href="https://github.com/smcclatchy/machine-learning-biomedical-science/blob/main/CONTRIBUTING.md" class="external-link">Contributing</a>
        | <a href="https://github.com/smcclatchy/machine-learning-biomedical-science/" class="external-link">Source</a></p>
				<p><a href="https://github.com/smcclatchy/machine-learning-biomedical-science/blob/main/CITATION.cff" class="external-link">Cite</a> | <a href="mailto:susan.mcclatchy@jax.org">Contact</a> | <a href="https://carpentries.org/about/" class="external-link">About</a></p>
			</div>
			<div class="col-md-6">

        <p>Materials licensed under <a href="LICENSE.html">CC-BY 4.0</a> by the authors</p>

        <p>Template licensed under <a href="https://creativecommons.org/licenses/by-sa/4.0/" class="external-link">CC-BY 4.0</a> by <a href="https://carpentries.org/" class="external-link">The Carpentries</a></p>
        <p>Built with <a href="https://github.com/carpentries/sandpaper/tree/0.16.12" class="external-link">sandpaper (0.16.12)</a>, <a href="https://github.com/carpentries/pegboard/tree/0.7.9" class="external-link">pegboard (0.7.9)</a>, and <a href="https://github.com/carpentries/varnish/tree/1.0.6" class="external-link">varnish (1.0.6)</a></p>
			</div>
		</footer>
</div> <!-- / div.container -->
	<div id="to-top">
		<a href="#top">
      <i class="search-icon" data-feather="arrow-up" role="img" aria-label="Back To Top"></i><br><!-- <span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top --><span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top
		</a>
	</div>
  <script type="application/ld+json">
    {
  "@context": "https://schema.org",
  "@type": "LearningResource",
  "@id": "https://smcclatchy.github.io/machine-learning-biomedical-science/instructor/aio.html",
  "inLanguage": "en",
  "dct:conformsTo": "https://bioschemas.org/profiles/LearningResource/1.0-RELEASE",
  "description": "A Carpentries Lesson teaching foundational data and coding skills to researchers worldwide",
  "keywords": "software, data, lesson, The Carpentries, machine learning, classification, clustering, biomedical science",
  "name": "All in One View",
  "creativeWorkStatus": "active",
  "url": "https://smcclatchy.github.io/machine-learning-biomedical-science/instructor/aio.html",
  "identifier": "https://smcclatchy.github.io/machine-learning-biomedical-science/instructor/aio.html",
  "dateCreated": "2025-07-15",
  "dateModified": "2025-07-18",
  "datePublished": "2025-07-18"
}

  </script><script>
		feather.replace();
	</script>
</body>
</html><!-- END:   inst/pkgdown/templates/layout.html-->

