<!DOCTYPE html>
<!-- START: inst/pkgdown/templates/layout.html --><!-- Generated by pkgdown: do not edit by hand --><html lang="en" data-bs-theme="auto"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><title>Machine Learning for Biomedical Science: Clustering</title><meta name="viewport" content="width=device-width, initial-scale=1"><script src="../assets/themetoggle.js"></script><link rel="stylesheet" type="text/css" href="../assets/styles.css"><script src="../assets/scripts.js" type="text/javascript"></script><!-- mathjax --><script type="text/x-mathjax-config">
    MathJax.Hub.Config({
      config: ["MMLorHTML.js"],
      jax: ["input/TeX","input/MathML","output/HTML-CSS","output/NativeMML", "output/PreviewHTML"],
      extensions: ["tex2jax.js","mml2jax.js","MathMenu.js","MathZoom.js", "fast-preview.js", "AssistiveMML.js", "a11y/accessibility-menu.js"],
      TeX: {
        extensions: ["AMSmath.js","AMSsymbols.js","noErrors.js","noUndefined.js"]
      },
      tex2jax: {
        inlineMath: [['\\(', '\\)']],
        displayMath: [ ['$$','$$'], ['\\[', '\\]'] ],
        processEscapes: true
      }
    });
    </script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><!-- Responsive Favicon for The Carpentries --><link rel="apple-touch-icon" sizes="180x180" href="../favicons/incubator/apple-touch-icon.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicons/incubator/favicon-32x32.png"><link rel="icon" type="image/png" sizes="16x16" href="../favicons/incubator/favicon-16x16.png"><link rel="manifest" href="../favicons/incubator/site.webmanifest"><link rel="mask-icon" href="../favicons/incubator/safari-pinned-tab.svg" color="#5bbad5"><meta name="msapplication-TileColor" content="#da532c"><meta name="theme-color" media="(prefers-color-scheme: light)" content="white"><meta name="theme-color" media="(prefers-color-scheme: dark)" content="black"></head><body>
    <header id="top" class="navbar navbar-expand-md top-nav incubator"><svg xmlns="http://www.w3.org/2000/svg" class="d-none"><symbol id="check2" viewbox="0 0 16 16"><path d="M13.854 3.646a.5.5 0 0 1 0 .708l-7 7a.5.5 0 0 1-.708 0l-3.5-3.5a.5.5 0 1 1 .708-.708L6.5 10.293l6.646-6.647a.5.5 0 0 1 .708 0z"></path></symbol><symbol id="circle-half" viewbox="0 0 16 16"><path d="M8 15A7 7 0 1 0 8 1v14zm0 1A8 8 0 1 1 8 0a8 8 0 0 1 0 16z"></path></symbol><symbol id="moon-stars-fill" viewbox="0 0 16 16"><path d="M6 .278a.768.768 0 0 1 .08.858 7.208 7.208 0 0 0-.878 3.46c0 4.021 3.278 7.277 7.318 7.277.527 0 1.04-.055 1.533-.16a.787.787 0 0 1 .81.316.733.733 0 0 1-.031.893A8.349 8.349 0 0 1 8.344 16C3.734 16 0 12.286 0 7.71 0 4.266 2.114 1.312 5.124.06A.752.752 0 0 1 6 .278z"></path><path d="M10.794 3.148a.217.217 0 0 1 .412 0l.387 1.162c.173.518.579.924 1.097 1.097l1.162.387a.217.217 0 0 1 0 .412l-1.162.387a1.734 1.734 0 0 0-1.097 1.097l-.387 1.162a.217.217 0 0 1-.412 0l-.387-1.162A1.734 1.734 0 0 0 9.31 6.593l-1.162-.387a.217.217 0 0 1 0-.412l1.162-.387a1.734 1.734 0 0 0 1.097-1.097l.387-1.162zM13.863.099a.145.145 0 0 1 .274 0l.258.774c.115.346.386.617.732.732l.774.258a.145.145 0 0 1 0 .274l-.774.258a1.156 1.156 0 0 0-.732.732l-.258.774a.145.145 0 0 1-.274 0l-.258-.774a1.156 1.156 0 0 0-.732-.732l-.774-.258a.145.145 0 0 1 0-.274l.774-.258c.346-.115.617-.386.732-.732L13.863.1z"></path></symbol><symbol id="sun-fill" viewbox="0 0 16 16"><path d="M8 12a4 4 0 1 0 0-8 4 4 0 0 0 0 8zM8 0a.5.5 0 0 1 .5.5v2a.5.5 0 0 1-1 0v-2A.5.5 0 0 1 8 0zm0 13a.5.5 0 0 1 .5.5v2a.5.5 0 0 1-1 0v-2A.5.5 0 0 1 8 13zm8-5a.5.5 0 0 1-.5.5h-2a.5.5 0 0 1 0-1h2a.5.5 0 0 1 .5.5zM3 8a.5.5 0 0 1-.5.5h-2a.5.5 0 0 1 0-1h2A.5.5 0 0 1 3 8zm10.657-5.657a.5.5 0 0 1 0 .707l-1.414 1.415a.5.5 0 1 1-.707-.708l1.414-1.414a.5.5 0 0 1 .707 0zm-9.193 9.193a.5.5 0 0 1 0 .707L3.05 13.657a.5.5 0 0 1-.707-.707l1.414-1.414a.5.5 0 0 1 .707 0zm9.193 2.121a.5.5 0 0 1-.707 0l-1.414-1.414a.5.5 0 0 1 .707-.707l1.414 1.414a.5.5 0 0 1 0 .707zM4.464 4.465a.5.5 0 0 1-.707 0L2.343 3.05a.5.5 0 1 1 .707-.707l1.414 1.414a.5.5 0 0 1 0 .708z"></path></symbol></svg><a class="visually-hidden-focusable skip-link" href="#main-content">Skip to main content</a>
  <div class="container-fluid top-nav-container">
    <div class="col-md-8">
      <div class="large-logo">
        <img id="incubator-logo" alt="Lesson Description" src="../assets/images/incubator-logo.svg"><span class="badge text-bg-danger">
          <abbr title="This lesson is in the pre-alpha phase, which means that it is in early development, but has not yet been taught.">
            <a href="https://docs.carpentries.org/resources/curriculum/lesson-life-cycle.html" class="external-link alert-link">
              <i aria-hidden="true" class="icon" data-feather="alert-octagon" style="border-radius: 5px"></i>
              Pre-Alpha
            </a>
            <span class="visually-hidden">This lesson is in the pre-alpha phase, which means that it is in early development, but has not yet been taught.</span>
          </abbr>
        </span>

      </div>
    </div>
    <div class="selector-container">
      <div id="theme-selector">
        <li class="nav-item dropdown" id="theme-button-list">
          <button class="btn btn-link nav-link px-0 px-lg-2 dropdown-toggle d-flex align-items-center" id="bd-theme" type="button" aria-expanded="false" data-bs-toggle="dropdown" data-bs-display="static" aria-label="Toggle theme (auto)">
            <svg class="bi my-1 theme-icon-active"><use href="#circle-half"></use></svg><i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="bd-theme-text"><li>
              <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="light" aria-pressed="false">
                <svg class="bi me-2 theme-icon"><use href="#sun-fill"></use></svg>
                Light
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
            <li>
              <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="dark" aria-pressed="false">
                <svg class="bi me-2 theme-icon"><use href="#moon-stars-fill"></use></svg>
                Dark
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
            <li>
              <button type="button" class="btn dropdown-item d-flex align-items-center active" data-bs-theme-value="auto" aria-pressed="true">
                <svg class="bi me-2 theme-icon"><use href="#circle-half"></use></svg>
                Auto
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
          </ul></li>
      </div>

      <div class="dropdown" id="instructor-dropdown">
        <button class="btn btn-secondary dropdown-toggle bordered-button" type="button" id="dropdownMenu1" data-bs-toggle="dropdown" aria-expanded="false">
          <i aria-hidden="true" class="icon" data-feather="eye"></i> Instructor View <i data-feather="chevron-down"></i>
        </button>
        <ul class="dropdown-menu" aria-labelledby="dropdownMenu1"><li><button class="dropdown-item" type="button" onclick="window.location.href='../clustering.html';">Learner View</button></li>
        </ul></div>
    </div>
  </div>
  <hr></header><nav class="navbar navbar-expand-xl bottom-nav incubator" aria-label="Main Navigation"><div class="container-fluid nav-container">
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarSupportedContent" aria-controls="navbarSupportedContent" aria-expanded="false" aria-label="Toggle Navigation">
      <span class="navbar-toggler-icon"></span>
      <span class="menu-title">Menu</span>
    </button>
    <div class="nav-logo">
      <img class="small-logo" alt="Lesson Description" src="../assets/images/incubator-logo-sm.svg"></div>
    <div class="lesson-title-md">
      Machine Learning for Biomedical Science
    </div>
    <div class="search-icon-sm">
      <!-- TODO: do not show until we have search
        <i role="img" aria-label="Search the All In One page" data-feather="search"></i>
      -->
    </div>
    <div class="desktop-nav">
      <ul class="navbar-nav me-auto mb-2 mb-lg-0"><li class="nav-item">
          <span class="lesson-title">
            Machine Learning for Biomedical Science
          </span>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="../instructor/key-points.html">Key Points</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="../instructor/instructor-notes.html">Instructor Notes</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="../instructor/images.html">Extract All Images</a>
        </li>
        <li class="nav-item dropdown">
          <button class="nav-link dropdown-toggle" id="navbarDropdown" data-bs-toggle="dropdown" aria-expanded="false">
            More <i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu" aria-labelledby="navbarDropdown"><hr><li><a class="dropdown-item" href="reference.html">Reference</a></li>
          </ul></li>
      </ul></div>
    <!--
    <form class="d-flex col-md-2 search-form">
      <fieldset disabled>
      <input class="form-control me-2 searchbox" type="search" placeholder="" aria-label="">
        <button class="btn btn-outline-success tablet-search-button"  type="submit">
          <i class="search-icon" data-feather="search" role="img" aria-label="Search the All In One page"></i>
        </button>
      </fieldset>
    </form>
    -->
    <a id="search-button" class="btn btn-primary" href="../instructor/aio.html" role="button" aria-label="Search the All In One page">Search the All In One page</a>
  </div><!--/div.container-fluid -->
</nav><div class="col-md-12 mobile-title">
  Machine Learning for Biomedical Science
</div>

<aside class="col-md-12 lesson-progress"><div style="width: 4%" class="percentage">
    4%
  </div>
  <div class="progress incubator">
    <div class="progress-bar incubator" role="progressbar" style="width: 4%" aria-valuenow="4" aria-label="Lesson Progress" aria-valuemin="0" aria-valuemax="100">
    </div>
  </div>
</aside><div class="container">
      <div class="row">
        <!-- START: inst/pkgdown/templates/navbar.html -->
<div id="sidebar-col" class="col-lg-4">
  <div id="sidebar" class="sidebar">
      <nav aria-labelledby="flush-headingEleven"><button role="button" aria-label="close menu" alt="close menu" aria-expanded="true" aria-controls="sidebar" class="collapse-toggle" data-collapse="Collapse " data-episodes="Episodes ">
          <i class="search-icon" data-feather="x" role="img"></i>
        </button>
        <div class="sidebar-inner">
          <div class="row mobile-row" id="theme-row-mobile">
            <div class="col" id="theme-selector">
              <li class="nav-item dropdown" id="theme-button-list">
                <button class="btn btn-link nav-link px-0 px-lg-2 dropdown-toggle d-flex align-items-center" id="bd-theme" type="button" aria-expanded="false" data-bs-toggle="dropdown" data-bs-display="static" aria-label="Toggle theme (auto)">
                  <svg class="bi my-1 theme-icon-active"><use href="#circle-half"></use></svg><span class="d-lg-none ms-1" id="bd-theme-text">Toggle Theme</span>
                </button>
                <ul class="dropdown-menu dropdown-menu-right" aria-labelledby="bd-theme-text"><li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="light" aria-pressed="false">
                      <svg class="bi me-2 theme-icon"><use href="#sun-fill"></use></svg>
                      Light
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                  <li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="dark" aria-pressed="false">
                      <svg class="bi me-2 theme-icon"><use href="#moon-stars-fill"></use></svg>
                      Dark
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                  <li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center active" data-bs-theme-value="auto" aria-pressed="true">
                      <svg class="bi me-2 theme-icon"><use href="#circle-half"></use></svg>
                      Auto
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                </ul></li>
            </div>
          </div>
          <div class="row mobile-row">
            <div class="col">
              <div class="sidenav-view-selector">
                <div class="accordion accordion-flush" id="accordionFlush9">
                  <div class="accordion-item">
                    <h2 class="accordion-header" id="flush-headingNine">
                      <button class="accordion-button collapsed" id="instructor" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseNine" aria-expanded="false" aria-controls="flush-collapseNine">
                        <i id="eye" aria-hidden="true" class="icon" data-feather="eye"></i> Instructor View
                      </button>
                    </h2>
                    <div id="flush-collapseNine" class="accordion-collapse collapse" aria-labelledby="flush-headingNine" data-bs-parent="#accordionFlush2">
                      <div class="accordion-body">
                        <a href="../clustering.html">Learner View</a>
                      </div>
                    </div>
                  </div><!--/div.accordion-item-->
                </div><!--/div.accordion-flush-->
              </div><!--div.sidenav-view-selector -->
            </div><!--/div.col -->

            <hr></div><!--/div.mobile-row -->

          <div class="accordion accordion-flush" id="accordionFlush11">
            <div class="accordion-item">

              <button id="chapters" class="accordion-button show" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseEleven" aria-expanded="false" aria-controls="flush-collapseEleven">
                <h2 class="accordion-header chapters" id="flush-headingEleven">
                  EPISODES
                </h2>
              </button>
              <div id="flush-collapseEleven" class="accordion-collapse show collapse" aria-labelledby="flush-headingEleven" data-bs-parent="#accordionFlush11">

                <div class="accordion-body">
                  <div class="accordion accordion-flush" id="accordionFlush1">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading1">
        <a href="index.html">Summary and Schedule</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush2">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading2">
        <a href="introduction.html">1. Using RMarkdown</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlushcurrent">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-headingcurrent">
      <button class="accordion-button" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapsecurrent" aria-expanded="true" aria-controls="flush-collapsecurrent">
        <span class="visually-hidden">Current Chapter</span>
        <span class="current-chapter">
        2. Clustering
        </span>
      </button>
    </div><!--/div.accordion-header-->

    <div id="flush-collapsecurrent" class="accordion-collapse collapse show" aria-labelledby="flush-headingcurrent" data-bs-parent="#accordionFlushcurrent">
      <div class="accordion-body">
        <ul><li><a href="#clustering">Clustering</a></li>
        </ul></div><!--/div.accordion-body-->
    </div><!--/div.accordion-collapse-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush4">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading4">
        <a href="conditional-probabilities-expectations.html">3. Conditional Probabilities and Expectations</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush5">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading5">
        <a href="smoothing.html">4. Smoothing</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush6">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading6">
        <a href="class-prediction.html">5. Class Prediction</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush7">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading7">
        <a href="cross-validation.html">6. Cross-validation</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

                </div>
              </div>
            </div>

            <hr class="half-width"><div class="accordion accordion-flush lesson-resources" id="accordionFlush12">
              <div class="accordion-item">
                <h2 class="accordion-header" id="flush-headingTwelve">
                  <button class="accordion-button collapsed" id="lesson-resources" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseTwelve" aria-expanded="false" aria-controls="flush-collapseTwelve">
                    RESOURCES
                  </button>
                </h2>
                <div id="flush-collapseTwelve" class="accordion-collapse collapse" aria-labelledby="flush-headingTwelve" data-bs-parent="#accordionFlush12">
                  <div class="accordion-body">
                    <ul><li>
                        <a href="../instructor/key-points.html">Key Points</a>
                      </li>
                      <li>
                        <a href="../instructor/instructor-notes.html">Instructor Notes</a>
                      </li>
                      <li>
                        <a href="../instructor/images.html">Extract All Images</a>
                      </li>
                      <hr><li><a class="dropdown-item" href="reference.html">Reference</a></li>
                    </ul></div>
                </div>
              </div>
            </div>
            <hr class="half-width lesson-resources"><a href="../instructor/aio.html">See all in one page</a>


            <hr class="d-none d-sm-block d-md-none"><div class="d-grid gap-1">

            </div>
          </div><!-- /div.accordion -->
        </div><!-- /div.sidebar-inner -->
      </nav></div><!-- /div.sidebar -->
  </div><!-- /div.sidebar-col -->
<!-- END:   inst/pkgdown/templates/navbar.html-->

        <!-- START: inst/pkgdown/templates/content-instructor.html -->
  <div class="col-xl-8 col-lg-12 primary-content">
    <nav class="lesson-content mx-md-4" aria-label="Previous and Next Chapter"><!-- content for small screens --><div class="d-block d-sm-block d-md-none">
        <a class="chapter-link" href="../instructor/introduction.html"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous</a>
        <a class="chapter-link float-end" href="../instructor/conditional-probabilities-expectations.html">Next<i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
      <!-- content for large screens -->
      <div class="d-none d-sm-none d-md-block">
        <a class="chapter-link" href="../instructor/introduction.html" rel="prev">
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>
          Previous: Using RMarkdown
        </a>
        <a class="chapter-link float-end" href="../instructor/conditional-probabilities-expectations.html" rel="next">
          Next: Conditional...
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i>
        </a>
      </div>
      <hr></nav><main id="main-content" class="main-content"><div class="container lesson-content">
        <h1>Clustering</h1>
        <p>Last updated on 2025-07-18 |

        <a href="https://github.com/carpentries/workbench-template-rmd/edit/main/episodes/clustering.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>



        <p>Estimated time: <i aria-hidden="true" data-feather="clock"></i> 60 minutes</p>

        <div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>

        

<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul><li>How can clusters within high dimensional data can be
discovered?</li>
</ul></div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul><li>Perform clustering analysis.</li>
<li>Interpret a heatmap.</li>
</ul></div>
</div>
</div>
</div>
</div>
<section><h2 class="section-heading" id="clustering">Clustering<a class="anchor" aria-label="anchor" href="#clustering"></a></h2>
<hr class="half-width"><p>We will demonstrate the concepts and code needed to perform
clustering analysis with the tissue gene expression data:</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">tissuesGeneExpression</span><span class="op">)</span></span>
<span><span class="fu">data</span><span class="op">(</span><span class="va">tissuesGeneExpression</span><span class="op">)</span></span></code></pre>
</div>
<p>To illustrate the main application of clustering in the life
sciences, let’s pretend that we don’t know these are different tissues
and are interested in clustering. The first step is to compute the
distance between each sample:</p>
<div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">d</span> <span class="op">&lt;-</span> <span class="fu">dist</span><span class="op">(</span> <span class="fu">t</span><span class="op">(</span><span class="va">e</span><span class="op">)</span> <span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error: object 'e' not found</code></pre>
</div>
<div class="section level4">
<h4 id="hierarchical-clustering">Hierarchical clustering<a class="anchor" aria-label="anchor" href="#hierarchical-clustering"></a></h4>
<p>With the distance between each pair of samples computed, we need
clustering algorithms to join them into groups. Hierarchical clustering
is one of the many clustering algorithms available to do this. Each
sample is assigned to its own group and then the algorithm continues
iteratively, joining the two most similar clusters at each step, and
continuing until there is just one group. While we have defined
distances between samples, we have not yet defined distances between
groups. There are various ways this can be done and they all rely on the
individual pairwise distances. The helpfile for <code>hclust</code>
includes detailed information.</p>
<p>We can perform hierarchical clustering based on the distances defined
above using the <code>hclust</code> function. This function returns an
<code>hclust</code> object that describes the groupings that were
created using the algorithm described above. The <code>plot</code>
method represents these relationships with a tree or dendrogram:</p>
<div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">rafalib</span><span class="op">)</span></span>
<span><span class="fu">mypar</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">hc</span> <span class="op">&lt;-</span> <span class="fu">hclust</span><span class="op">(</span><span class="va">d</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error: object 'd' not found</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb6">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">hc</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error: object 'hc' not found</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb8">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">plot</span><span class="op">(</span><span class="va">hc</span>,labels<span class="op">=</span><span class="va">tissue</span>,cex<span class="op">=</span><span class="fl">0.5</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error: object 'hc' not found</code></pre>
</div>
<figure><img src=".././fig/02-clustering-dendrogram-1.png" alt="Dendrogram showing hierarchical clustering of tissue gene expression data." class="figure mx-auto d-block"><div class="figcaption">Dendrogram showing hierarchical clustering of
tissue gene expression data.</div>
</figure><p>Does this technique “discover” the clusters defined by the different
tissues? In this plot, it is not easy to see the different tissues so we
add colors by using the <code>myplclust</code> function from the
<code>rafalib</code> package.</p>
<div class="codewrapper sourceCode" id="cb10">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">myplclust</span><span class="op">(</span><span class="va">hc</span>, labels<span class="op">=</span><span class="va">tissue</span>, lab.col<span class="op">=</span><span class="fu">as.fumeric</span><span class="op">(</span><span class="va">tissue</span><span class="op">)</span>, cex<span class="op">=</span><span class="fl">0.5</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error: object 'hc' not found</code></pre>
</div>
<figure><img src=".././fig/02-clustering-color_dendrogram-1.png" alt="Dendrogram showing hierarchical clustering of tissue gene expression data with colors denoting tissues." class="figure mx-auto d-block"><div class="figcaption">Dendrogram showing hierarchical clustering of
tissue gene expression data with colors denoting tissues.</div>
</figure><p>Visually, it does seem as if the clustering technique has discovered
the tissues. However, hierarchical clustering does not define specific
clusters, but rather defines the dendrogram above. From the dendrogram
we can decipher the distance between any two groups by looking at the
height at which the two groups split into two. To define clusters, we
need to “cut the tree” at some distance and group all samples that are
within that distance into groups below. To visualize this, we draw a
horizontal line at the height we wish to cut and this defines that line.
We use 120 as an example:</p>
<div class="codewrapper sourceCode" id="cb12">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">myplclust</span><span class="op">(</span><span class="va">hc</span>, labels<span class="op">=</span><span class="va">tissue</span>, lab.col<span class="op">=</span><span class="fu">as.fumeric</span><span class="op">(</span><span class="va">tissue</span><span class="op">)</span>,cex<span class="op">=</span><span class="fl">0.5</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error: object 'hc' not found</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb14">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">abline</span><span class="op">(</span>h<span class="op">=</span><span class="fl">120</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in int_abline(a = a, b = b, h = h, v = v, untf = untf, ...): plot.new has not been called yet</code></pre>
</div>
<p><img src=".././fig/02-clustering-color_dendrogram2-1.png" alt="Dendrogram showing hierarchical clustering of tissue gene expression data with colors denoting tissues. Horizontal line defines actual clusters." class="figure">
If we use the line above to cut the tree into clusters, we can examine
how the clusters overlap with the actual tissues:</p>
<div class="codewrapper sourceCode" id="cb16">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">hclusters</span> <span class="op">&lt;-</span> <span class="fu">cutree</span><span class="op">(</span><span class="va">hc</span>, h<span class="op">=</span><span class="fl">120</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error: object 'hc' not found</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb18">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">table</span><span class="op">(</span>true<span class="op">=</span><span class="va">tissue</span>, cluster<span class="op">=</span><span class="va">hclusters</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error: object 'tissue' not found</code></pre>
</div>
<p>We can also ask <code>cutree</code> to give us back a given number of
clusters. The function then automatically finds the height that results
in the requested number of clusters:</p>
<div class="codewrapper sourceCode" id="cb20">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">hclusters</span> <span class="op">&lt;-</span> <span class="fu">cutree</span><span class="op">(</span><span class="va">hc</span>, k<span class="op">=</span><span class="fl">8</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error: object 'hc' not found</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb22">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">table</span><span class="op">(</span>true<span class="op">=</span><span class="va">tissue</span>, cluster<span class="op">=</span><span class="va">hclusters</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error: object 'tissue' not found</code></pre>
</div>
<p>In both cases we do see that, with some exceptions, each tissue is
uniquely represented by one of the clusters. In some instances, the one
tissue is spread across two tissues, which is due to selecting too many
clusters. Selecting the number of clusters is generally a challenging
step in practice and an active area of research.</p>
</div>
<div class="section level4">
<h4 id="k-means">K-means<a class="anchor" aria-label="anchor" href="#k-means"></a></h4>
<p>We can also cluster with the <code>kmeans</code> function to perform
k-means clustering. As an example, let’s run k-means on the samples in
the space of the first two genes:</p>
<div class="codewrapper sourceCode" id="cb24">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">set.seed</span><span class="op">(</span><span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">km</span> <span class="op">&lt;-</span> <span class="fu">kmeans</span><span class="op">(</span><span class="fu">t</span><span class="op">(</span><span class="va">e</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span>,<span class="op">]</span><span class="op">)</span>, centers<span class="op">=</span><span class="fl">7</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error: object 'e' not found</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb26">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">names</span><span class="op">(</span><span class="va">km</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error: object 'km' not found</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb28">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">mypar</span><span class="op">(</span><span class="fl">1</span>,<span class="fl">2</span><span class="op">)</span></span>
<span><span class="fu">plot</span><span class="op">(</span><span class="va">e</span><span class="op">[</span><span class="fl">1</span>,<span class="op">]</span>, <span class="va">e</span><span class="op">[</span><span class="fl">2</span>,<span class="op">]</span>, col<span class="op">=</span><span class="fu">as.fumeric</span><span class="op">(</span><span class="va">tissue</span><span class="op">)</span>, pch<span class="op">=</span><span class="fl">16</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error: object 'e' not found</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb30">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">plot</span><span class="op">(</span><span class="va">e</span><span class="op">[</span><span class="fl">1</span>,<span class="op">]</span>, <span class="va">e</span><span class="op">[</span><span class="fl">2</span>,<span class="op">]</span>, col<span class="op">=</span><span class="va">km</span><span class="op">$</span><span class="va">cluster</span>, pch<span class="op">=</span><span class="fl">16</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error: object 'e' not found</code></pre>
</div>
<figure><img src=".././fig/02-clustering-kmeans-1.png" alt="Plot of gene expression for first two genes (order of appearance in data) with color representing tissue (left) and clusters found with kmeans (right)." class="figure mx-auto d-block"><div class="figcaption">Plot of gene expression for first two genes
(order of appearance in data) with color representing tissue (left) and
clusters found with kmeans (right).</div>
</figure><p>In the first plot, color represents the actual tissues, while in the
second, color represents the clusters that were defined by
<code>kmeans</code>. We can see from tabulating the results that this
particular clustering exercise did not perform well:</p>
<div class="codewrapper sourceCode" id="cb32">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">table</span><span class="op">(</span>true<span class="op">=</span><span class="va">tissue</span>,cluster<span class="op">=</span><span class="va">km</span><span class="op">$</span><span class="va">cluster</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error: object 'tissue' not found</code></pre>
</div>
<p>This is very likely due to the fact that the first two genes are not
informative regarding tissue type. We can see this in the first plot
above. If we instead perform k-means clustering using all of the genes,
we obtain a much improved result. To visualize this, we can use an MDS
plot:</p>
<div class="codewrapper sourceCode" id="cb34">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">km</span> <span class="op">&lt;-</span> <span class="fu">kmeans</span><span class="op">(</span><span class="fu">t</span><span class="op">(</span><span class="va">e</span><span class="op">)</span>, centers<span class="op">=</span><span class="fl">7</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error: object 'e' not found</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb36">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mds</span> <span class="op">&lt;-</span> <span class="fu">cmdscale</span><span class="op">(</span><span class="va">d</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error: object 'd' not found</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb38">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">mypar</span><span class="op">(</span><span class="fl">1</span>,<span class="fl">2</span><span class="op">)</span></span>
<span><span class="fu">plot</span><span class="op">(</span><span class="va">mds</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span>, <span class="va">mds</span><span class="op">[</span>,<span class="fl">2</span><span class="op">]</span><span class="op">)</span> </span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error: object 'mds' not found</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb40">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">plot</span><span class="op">(</span><span class="va">mds</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span>, <span class="va">mds</span><span class="op">[</span>,<span class="fl">2</span><span class="op">]</span>, col<span class="op">=</span><span class="va">km</span><span class="op">$</span><span class="va">cluster</span>, pch<span class="op">=</span><span class="fl">16</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error: object 'mds' not found</code></pre>
</div>
<figure><img src=".././fig/02-clustering-kmeans_mds-1.png" alt="Plot of gene expression for first two PCs with color representing tissues (left) and clusters found using all genes (right)." class="figure mx-auto d-block"><div class="figcaption">Plot of gene expression for first two PCs with
color representing tissues (left) and clusters found using all genes
(right).</div>
</figure><p>By tabulating the results, we see that we obtain a similar answer to
that obtained with hierarchical clustering.</p>
<div class="codewrapper sourceCode" id="cb42">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">table</span><span class="op">(</span>true<span class="op">=</span><span class="va">tissue</span>,cluster<span class="op">=</span><span class="va">km</span><span class="op">$</span><span class="va">cluster</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error: object 'tissue' not found</code></pre>
</div>
</div>
<div class="section level4">
<h4 id="heatmaps">Heatmaps<a class="anchor" aria-label="anchor" href="#heatmaps"></a></h4>
<p>Heatmaps are ubiquitous in the genomics literature. They are very
useful plots for visualizing the measurements for a subset of rows over
all the samples. A <em>dendrogram</em> is added on top and on the side
that is created with hierarchical clustering. We will demonstrate how to
create heatmaps from within R. Let’s begin by defining a color
palette:</p>
<div class="codewrapper sourceCode" id="cb44">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">RColorBrewer</span><span class="op">)</span> </span>
<span><span class="va">hmcol</span> <span class="op">&lt;-</span> <span class="fu">colorRampPalette</span><span class="op">(</span><span class="fu">brewer.pal</span><span class="op">(</span><span class="fl">9</span>, <span class="st">"GnBu"</span><span class="op">)</span><span class="op">)</span><span class="op">(</span><span class="fl">100</span><span class="op">)</span></span></code></pre>
</div>
<p>Now, pick the genes with the top variance over all samples:</p>
<div class="codewrapper sourceCode" id="cb45">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">genefilter</span><span class="op">)</span></span>
<span><span class="va">rv</span> <span class="op">&lt;-</span> <span class="fu">rowVars</span><span class="op">(</span><span class="va">e</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in h(simpleError(msg, call)): error in evaluating the argument 'x' in selecting a method for function 'rowSums': object 'e' not found</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb47">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">idx</span> <span class="op">&lt;-</span> <span class="fu">order</span><span class="op">(</span><span class="op">-</span><span class="va">rv</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">40</span><span class="op">]</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error: object 'rv' not found</code></pre>
</div>
<p>While a <code>heatmap</code> function is included in R, we recommend
the <code>heatmap.2</code> function from the <code>gplots</code> package
on CRAN because it is a bit more customized. For example, it stretches
to fill the window. Here we add colors to indicate the tissue on the
top:</p>
<div class="codewrapper sourceCode" id="cb49">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">gplots</span><span class="op">)</span> <span class="co">##Available from CRAN</span></span>
<span><span class="va">cols</span> <span class="op">&lt;-</span> <span class="fu">palette</span><span class="op">(</span><span class="fu">brewer.pal</span><span class="op">(</span><span class="fl">8</span>, <span class="st">"Dark2"</span><span class="op">)</span><span class="op">)</span><span class="op">[</span><span class="fu">as.fumeric</span><span class="op">(</span><span class="va">tissue</span><span class="op">)</span><span class="op">]</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error: object 'tissue' not found</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb51">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">head</span><span class="op">(</span><span class="fu">cbind</span><span class="op">(</span><span class="fu">colnames</span><span class="op">(</span><span class="va">e</span><span class="op">)</span>,<span class="va">cols</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error: object 'e' not found</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb53">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">heatmap.2</span><span class="op">(</span><span class="va">e</span><span class="op">[</span><span class="va">idx</span>,<span class="op">]</span>, labCol<span class="op">=</span><span class="va">tissue</span>,</span>
<span>          trace<span class="op">=</span><span class="st">"none"</span>, </span>
<span>          ColSideColors<span class="op">=</span><span class="va">cols</span>, </span>
<span>          col<span class="op">=</span><span class="va">hmcol</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error: object 'e' not found</code></pre>
</div>
<figure><img src=".././fig/02-clustering-heatmap.2-1.png" alt="Heatmap created using the 40 most variable genes and the function heatmap.2." class="figure mx-auto d-block"><div class="figcaption">Heatmap created using the 40 most variable genes
and the function heatmap.2.</div>
</figure><p>We did not use tissue information to create this heatmap, and we can
quickly see, with just 40 genes, good separation across tissues.</p>
<div id="exercise-1" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="exercise-1" class="callout-inner">
<h3 class="callout-title">Exercise 1</h3>
<div class="callout-content">
<p>Create a random matrix with no correlation in the following way:</p>
<div class="codewrapper sourceCode" id="cb55">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">set.seed</span><span class="op">(</span><span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">m</span> <span class="op">=</span> <span class="fl">10000</span></span>
<span><span class="va">n</span> <span class="op">=</span> <span class="fl">24</span></span>
<span><span class="va">x</span> <span class="op">=</span> <span class="fu">matrix</span><span class="op">(</span><span class="fu">rnorm</span><span class="op">(</span><span class="va">m</span> <span class="op">*</span> <span class="va">n</span><span class="op">)</span>, <span class="va">m</span>, <span class="va">n</span><span class="op">)</span> </span>
<span><span class="fu">colnames</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="va">n</span></span></code></pre>
</div>
<p>Run hierarchical clustering on this data with the <code>hclust</code>
function with default parameters to cluster the columns. Create a
dendrogram.<br>
In the dendrogram, which pairs of samples are the furthest away from
each other?<br>
A) 7 and 23<br>
B) 19 and 14<br>
C) 1 and 16<br>
D) 17 and 18</p>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1"> Show me the solution </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" aria-labelledby="headingSolution1" data-bs-parent="#accordionSolution1">
<div class="accordion-body">
<p><code>d &lt;- dist(t(x))</code><br><code>hc &lt;- hclust(d)</code><br><code>mypar()</code><br><code>plot(hc)</code></p>
<p>A hint to determine the disimilarity between two samples is to look
at how hight would be the dendogram be cut in order to have these two
samples in the same group. The higher the cut, the disimilar those
samples are. In this case the cut that would join sample 19 and 14 is
higher than any of the other pairs.</p>
<p>7 and 23 - 141<br>
19 and 14 - 143<br>
1 and 16 - 142<br>
17 and 18 - 142</p>
<p>The answer is B: 19 and 14. The answer might be different due to the
random numbers.</p>
</div>
</div>
</div>
</div>
<div id="exercise-2" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="exercise-2" class="callout-inner">
<h3 class="callout-title">Exercise 2</h3>
<div class="callout-content">
<p>Set the seed at 1, <code>set.seed(1)</code> and replicate the
creation of this matrix 100 times:</p>
<div class="codewrapper sourceCode" id="cb56">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">m</span> <span class="op">=</span> <span class="fl">10000</span></span>
<span><span class="va">n</span> <span class="op">=</span> <span class="fl">24</span></span>
<span><span class="va">x</span> <span class="op">=</span> <span class="fu">matrix</span><span class="op">(</span><span class="fu">rnorm</span><span class="op">(</span><span class="va">m</span> <span class="op">*</span> <span class="va">n</span><span class="op">)</span>, <span class="va">m</span>, <span class="va">n</span><span class="op">)</span></span></code></pre>
</div>
<p>then perform hierarchical clustering as in the solution to exercise
1, and find the number of clusters if you use <code>cuttree</code> at
height 143. This number is a random variable. Based on the Monte Carlo
simulation, what is the standard error of this random variable?</p>
</div>
</div>
</div>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2"> Show me the solution </h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" aria-labelledby="headingSolution2" data-bs-parent="#accordionSolution2">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb57">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">set.seed</span><span class="op">(</span><span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">res_list</span> <span class="op">&lt;-</span> <span class="fu">replicate</span><span class="op">(</span><span class="fl">100</span>, <span class="op">{</span></span>
<span>  <span class="va">m</span> <span class="op">=</span> <span class="fl">10000</span></span>
<span>  <span class="va">n</span> <span class="op">=</span> <span class="fl">24</span></span>
<span>  <span class="co"># Generate a new matrix every repetition.</span></span>
<span>  <span class="va">x</span> <span class="op">=</span> <span class="fu">matrix</span><span class="op">(</span><span class="fu">rnorm</span><span class="op">(</span><span class="va">m</span> <span class="op">*</span> <span class="va">n</span><span class="op">)</span>, <span class="va">m</span>, <span class="va">n</span><span class="op">)</span></span>
<span>  <span class="va">d</span> <span class="op">&lt;-</span> <span class="fu">dist</span><span class="op">(</span><span class="fu">t</span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="co"># Run the hierarchical clustering on the new matrix.</span></span>
<span>  <span class="va">hc</span> <span class="op">&lt;-</span> <span class="fu">hclust</span><span class="op">(</span><span class="va">d</span><span class="op">)</span></span>
<span>  <span class="co"># Compute how many clusters form if we cut the dendogram at a height of 143.</span></span>
<span>  <span class="va">hclusters</span> <span class="op">&lt;-</span> <span class="fu">cutree</span><span class="op">(</span><span class="va">hc</span>, h<span class="op">=</span><span class="fl">143</span><span class="op">)</span></span>
<span>  <span class="va">num_clus</span> <span class="op">&lt;-</span> <span class="fu">length</span><span class="op">(</span><span class="fu">unique</span><span class="op">(</span><span class="va">hclusters</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="kw">return</span><span class="op">(</span><span class="va">num_clus</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span>
<span><span class="co"># Compute the Standard Error of the number of clusters created by cutting</span></span>
<span><span class="co"># the dendograms at a height of 143 on each repetition.</span></span>
<span><span class="fu">popsd</span><span class="op">(</span><span class="va">res_list</span><span class="op">)</span></span></code></pre>
</div>
</div>
</div>
</div>
</div>
<div id="exercise-3" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="exercise-3" class="callout-inner">
<h3 class="callout-title">Exercise 3</h3>
<div class="callout-content">
<p>Run <code>kmeans</code> with 4 centers for the blood RNA data:</p>
<div class="codewrapper sourceCode" id="cb58">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">GSE5859Subset</span><span class="op">)</span></span>
<span><span class="fu">data</span><span class="op">(</span><span class="va">GSE5859Subset</span><span class="op">)</span></span></code></pre>
</div>
<p>Set the seed to 10, <code>set.seed(10)</code> right before running
kmeans with 5 centers. Explore the relationship of clusters and
information in <code>sampleInfo</code>. Which of the following best
describes what you find? A) <code>sampleInfo$group</code> is driving the
clusters as the 0s and 1s are in completely different clusters. B) The
year is driving the clusters. C) <code>Date</code> is driving the
clusters. D) The clusters don’t depend on any of the column of
<code>sampleInfo</code></p>
</div>
</div>
</div>
<div id="accordionSolution3" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution3" aria-expanded="false" aria-controls="collapseSolution3">
  <h4 class="accordion-header" id="headingSolution3"> Show me the solution </h4>
</button>
<div id="collapseSolution3" class="accordion-collapse collapse" aria-labelledby="headingSolution3" data-bs-parent="#accordionSolution3">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb59">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">km</span> <span class="op">&lt;-</span> <span class="fu">kmeans</span><span class="op">(</span><span class="fu">t</span><span class="op">(</span><span class="va">geneExpression</span><span class="op">)</span>, centers <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span>
<span><span class="va">km</span><span class="op">$</span><span class="va">cluster</span></span>
<span><span class="co"># Use the group as the true clusters of our data, and compare it agains the</span></span>
<span><span class="co"># clusters found by k-means.</span></span>
<span><span class="fu">table</span><span class="op">(</span>true <span class="op">=</span> <span class="va">sampleInfo</span><span class="op">$</span><span class="va">group</span>, cluster <span class="op">=</span> <span class="va">km</span><span class="op">$</span><span class="va">cluster</span><span class="op">)</span></span>
<span><span class="co"># Now use the date of acquisition of the samples as the true clusters and</span></span>
<span><span class="co"># compare it against the same clusters found by k-means.</span></span>
<span><span class="fu">table</span><span class="op">(</span>true <span class="op">=</span> <span class="va">sampleInfo</span><span class="op">$</span><span class="va">date</span>, cluster <span class="op">=</span> <span class="va">km</span><span class="op">$</span><span class="va">cluster</span><span class="op">)</span></span></code></pre>
</div>
<p>The answer is C: Date is driving the clusters.</p>
</div>
</div>
</div>
</div>
<div id="exercise-4" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="exercise-4" class="callout-inner">
<h3 class="callout-title">Exercise 4</h3>
<div class="callout-content">
<p>Load the data:</p>
<div class="codewrapper sourceCode" id="cb60">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">GSE5859Subset</span><span class="op">)</span></span>
<span><span class="fu">data</span><span class="op">(</span><span class="va">GSE5859Subset</span><span class="op">)</span></span></code></pre>
</div>
<p>Pick the 25 genes with the highest across sample variance. This
function might help:</p>
<div class="codewrapper sourceCode" id="cb61">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">install.packages</span><span class="op">(</span><span class="st">"matrixStats"</span><span class="op">)</span> </span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">matrixStats</span><span class="op">)</span></span>
<span><span class="op">?</span><span class="va">rowMads</span> <span class="co">## we use mads due to a outlier sample</span></span></code></pre>
</div>
<p>Use <code>heatmap.2</code> to make a heatmap showing the
<code>sampleInfo$group</code> with color, the date as labels, the rows
labelled with chromosome, and scaling the rows. What do we learn from
this heatmap? A) The data appears as if it was generated by
<code>rnorm</code>.<br>
B) Some genes in chr1 are very variable.<br>
C) A group of chrY genes are higher in group 0 and appear to drive the
clustering. Within those clusters there appears to be clustering by
<code>month</code>.<br>
D) A group of chrY genes are higher in October compared to June and
appear to drive the clustering. Within those clusters there appears to
be clustering by <code>samplInfo$group</code>.</p>
</div>
</div>
</div>
<div id="accordionSolution4" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution4" aria-expanded="false" aria-controls="collapseSolution4">
  <h4 class="accordion-header" id="headingSolution4"> Show me the solution </h4>
</button>
<div id="collapseSolution4" class="accordion-collapse collapse" aria-labelledby="headingSolution4" data-bs-parent="#accordionSolution4">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb62">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">hmcol</span> <span class="op">&lt;-</span> <span class="fu">colorRampPalette</span><span class="op">(</span><span class="fu">brewer.pal</span><span class="op">(</span><span class="fl">9</span>, <span class="st">"GnBu"</span><span class="op">)</span><span class="op">)</span><span class="op">(</span><span class="fl">100</span><span class="op">)</span></span>
<span><span class="va">month</span> <span class="op">=</span> <span class="fu">format</span><span class="op">(</span> <span class="va">sampleInfo</span><span class="op">$</span><span class="va">date</span>, <span class="st">"%m"</span><span class="op">)</span></span>
<span><span class="va">rv</span> <span class="op">&lt;-</span> <span class="fu">rowVars</span><span class="op">(</span><span class="va">geneExpression</span><span class="op">)</span></span>
<span><span class="va">idx</span> <span class="op">&lt;-</span> <span class="fu">order</span><span class="op">(</span><span class="op">-</span><span class="va">rv</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">25</span><span class="op">]</span></span>
<span><span class="va">cols</span> <span class="op">&lt;-</span> <span class="fu">palette</span><span class="op">(</span><span class="fu">brewer.pal</span><span class="op">(</span><span class="fl">8</span>, <span class="st">"Dark2"</span><span class="op">)</span><span class="op">)</span><span class="op">[</span><span class="fu">as.fumeric</span><span class="op">(</span><span class="fu">as.character</span><span class="op">(</span><span class="va">sampleInfo</span><span class="op">$</span><span class="va">group</span><span class="op">)</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="co"># Compute the heatmap using the intensity of the gene expressions to color</span></span>
<span><span class="co"># the graph. We are using the months of acquisition of the samples as the</span></span>
<span><span class="co"># columns in our plot, so we can see how the samples arrange according to the date.</span></span>
<span><span class="fu">heatmap.2</span><span class="op">(</span><span class="va">geneExpression</span><span class="op">[</span><span class="va">idx</span>,<span class="op">]</span>, </span>
<span>          trace <span class="op">=</span> <span class="st">'none'</span>, labRow <span class="op">=</span> <span class="va">geneAnnotation</span><span class="op">[</span><span class="va">idx</span>,<span class="op">]</span><span class="op">$</span><span class="va">CHR</span>,</span>
<span>          col <span class="op">=</span> <span class="va">hmcol</span>, labCol <span class="op">=</span> <span class="va">month</span>,</span>
<span>          ColSideColors <span class="op">=</span> <span class="va">cols</span><span class="op">)</span></span></code></pre>
</div>
<p>The correct answer is C: A group of chrY genes are higher in group 0
and appear to drive the clustering. Within those clusters there appears
to be clustering by month. Answer A is wrong because if the data were
generated by <code>norm</code>, the color distribution of the heatmap
would be entirely random. Answer B is wrong because the colors in the
row chr1 are more or less the same except for one column (one sample).
Answer D is wrong chrY genes are higher in June, not October.</p>
</div>
</div>
</div>
</div>
<div id="exercise-5" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="exercise-5" class="callout-inner">
<h3 class="callout-title">Exercise 5</h3>
<div class="callout-content">
<p>Create a large dataset of random data that is completely independent
of <code>sampleInfo$group</code> like this:</p>
<div class="codewrapper sourceCode" id="cb63">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">set.seed</span><span class="op">(</span><span class="fl">17</span><span class="op">)</span></span>
<span><span class="va">m</span> <span class="op">=</span> <span class="fu">nrow</span><span class="op">(</span><span class="va">geneExpression</span><span class="op">)</span> </span>
<span><span class="va">n</span> <span class="op">=</span> <span class="fu">ncol</span><span class="op">(</span><span class="va">geneExpression</span><span class="op">)</span> </span>
<span><span class="va">x</span> <span class="op">=</span> <span class="fu">matrix</span><span class="op">(</span><span class="fu">rnorm</span><span class="op">(</span><span class="va">m</span> <span class="op">*</span> <span class="va">n</span><span class="op">)</span>, <span class="va">m</span>, <span class="va">n</span><span class="op">)</span> </span>
<span><span class="va">g</span> <span class="op">=</span> <span class="fu">factor</span><span class="op">(</span><span class="va">sampleInfo</span><span class="op">$</span><span class="va">g</span><span class="op">)</span></span></code></pre>
</div>
<p>Create two heatmaps with these data. Show the group <code>g</code>
either with labels or colors. First, take the 50 genes with smallest
p-values obtained with <code>rowttests</code>. Then, take the 50 genes
with largest standard deviations. Which of the following statements is
true?<br>
A) There is no relationship between <code>g</code> and <code>x</code>,
but with 8,793 tests some will appear significant by chance. Selecting
genes with the t-test gives us a deceiving result.<br>
B) These two techniques produced similar heatmaps.<br>
C) Selecting genes with the t-test is a better technique since it
permits us to detect the two groups. It appears to find hidden
signals.<br>
D) The genes with the largest standard deviation add variability to the
plot and do not let us find the differences between the two groups.</p>
</div>
</div>
</div>
<div id="accordionSolution5" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution5" aria-expanded="false" aria-controls="collapseSolution5">
  <h4 class="accordion-header" id="headingSolution5"> Show me the solution </h4>
</button>
<div id="collapseSolution5" class="accordion-collapse collapse" aria-labelledby="headingSolution5" data-bs-parent="#accordionSolution5">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb64">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">genefilter</span><span class="op">)</span></span>
<span><span class="co"># p-value</span></span>
<span><span class="co"># Perform a t-test on the expression of each gene to determine if they have</span></span>
<span><span class="co"># an effect different from 0.</span></span>
<span><span class="va">pvals</span> <span class="op">&lt;-</span> <span class="fu">rowttests</span><span class="op">(</span><span class="va">x</span>, <span class="va">g</span><span class="op">)</span><span class="op">$</span><span class="va">p.value</span></span>
<span><span class="co"># Select the top 50 genes that show the most statistical significance of their</span></span>
<span><span class="co"># expression, obtained by the t-tests.</span></span>
<span><span class="va">idx</span> <span class="op">&lt;-</span> <span class="fu">order</span><span class="op">(</span><span class="va">pvals</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">50</span><span class="op">]</span></span>
<span><span class="va">cols</span> <span class="op">&lt;-</span> <span class="fu">palette</span><span class="op">(</span><span class="fu">brewer.pal</span><span class="op">(</span><span class="fl">8</span>, <span class="st">"Dark2"</span><span class="op">)</span><span class="op">)</span><span class="op">[</span><span class="fu">as.fumeric</span><span class="op">(</span><span class="fu">as.character</span><span class="op">(</span><span class="va">sampleInfo</span><span class="op">$</span><span class="va">g</span><span class="op">)</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="fu">heatmap.2</span><span class="op">(</span><span class="va">x</span><span class="op">[</span><span class="va">idx</span>,<span class="op">]</span>, </span>
<span>          trace <span class="op">=</span> <span class="st">'none'</span>, labRow <span class="op">=</span> <span class="va">geneAnnotation</span><span class="op">[</span><span class="va">idx</span>,<span class="op">]</span><span class="op">$</span><span class="va">CHR</span>,</span>
<span>          col <span class="op">=</span> <span class="va">hmcol</span>, labCol <span class="op">=</span> <span class="va">month</span>,</span>
<span>          ColSideColors <span class="op">=</span> <span class="va">cols</span><span class="op">)</span></span>
<span><span class="co"># std dev</span></span>
<span><span class="co"># Compute the standard deviation of the expression of each gene.</span></span>
<span><span class="va">sds</span> <span class="op">&lt;-</span> <span class="fu">rowSds</span><span class="op">(</span><span class="va">x</span>,<span class="va">g</span><span class="op">)</span></span>
<span><span class="co"># Select the top 50 genes that show a higher variance of their expression</span></span>
<span><span class="co"># acoss all the samples.</span></span>
<span><span class="va">idx</span> <span class="op">&lt;-</span> <span class="fu">order</span><span class="op">(</span><span class="op">-</span><span class="va">sds</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">50</span><span class="op">]</span></span>
<span><span class="va">cols</span> <span class="op">&lt;-</span> <span class="fu">palette</span><span class="op">(</span><span class="fu">brewer.pal</span><span class="op">(</span><span class="fl">8</span>, <span class="st">"Dark2"</span><span class="op">)</span><span class="op">)</span><span class="op">[</span><span class="fu">as.fumeric</span><span class="op">(</span><span class="fu">as.character</span><span class="op">(</span><span class="va">sampleInfo</span><span class="op">$</span><span class="va">g</span><span class="op">)</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="fu">heatmap.2</span><span class="op">(</span><span class="va">x</span><span class="op">[</span><span class="va">idx</span>,<span class="op">]</span>, </span>
<span>          trace <span class="op">=</span> <span class="st">'none'</span>, labRow <span class="op">=</span> <span class="va">geneAnnotation</span><span class="op">[</span><span class="va">idx</span>,<span class="op">]</span><span class="op">$</span><span class="va">CHR</span>,</span>
<span>          col <span class="op">=</span> <span class="va">hmcol</span>, labCol <span class="op">=</span> <span class="va">month</span>,</span>
<span>          ColSideColors <span class="op">=</span> <span class="va">cols</span><span class="op">)</span></span></code></pre>
</div>
<p>The answer is A: There is no relationship between g and x, but with
8,793 tests some will appear significant by chance. Selecting genes with
the t-test gives us a deceiving result.</p>
<p>Recall that we have already selected smallest p-values from a dataset
in which the null hypothesis is true. Therefore, we can see clusters
that indicate that there is a significant difference between sample
groups. However, this significance is not real because we know that the
null hypothesis is true.</p>
</div>
</div>
</div>
</div>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Key Points</h3>
<div class="callout-content">
<ul><li>We can use Euclidean distance to define the dissimilarity between
samples.</li>
<li>We can also use other metrics according to the prior knowledge we
have from our data.</li>
</ul></div>
</div>
</div>
<!--
Place links that you need to refer to multiple times across pages here. Delete
any links that you are not going to use.
 -->
</div>
</section></div> <!-- / div.lesson-content -->
    </main><!-- / main#main-content.main-content --><nav class="bottom-pagination mx-md-4" aria-label="Previous and Next Chapter"><div class="d-block d-sm-block d-md-none">
        <a class="chapter-link" href="../instructor/introduction.html"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous</a>
        <a class="chapter-link float-end" href="../instructor/conditional-probabilities-expectations.html">Next<i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
      <!-- content for large screens -->
      <div class="d-none d-sm-none d-md-block">
        <a class="chapter-link" href="../instructor/introduction.html" rel="prev">
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>
          Previous: Using RMarkdown
        </a>
        <a class="chapter-link float-end" href="../instructor/conditional-probabilities-expectations.html" rel="next">
          Next: Conditional...
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i>
        </a>
      </div>
    </nav></div> <!-- / div.primary-content.col-xs-12 -->
<!-- END:   inst/pkgdown/templates/content-instructor.html-->

      </div><!--/div.row-->
      		<footer class="row footer mx-md-3"><hr><div class="col-md-6">
        <p>This lesson is subject to the <a href="CODE_OF_CONDUCT.html">Code of Conduct</a></p>
        <p>

        <a href="https://github.com/carpentries/workbench-template-rmd/edit/main/episodes/clustering.Rmd" class="external-link">Edit on GitHub</a>

	
        | <a href="https://github.com/carpentries/workbench-template-rmd/blob/main/CONTRIBUTING.md" class="external-link">Contributing</a>
        | <a href="https://github.com/carpentries/workbench-template-rmd/" class="external-link">Source</a></p>
				<p><a href="https://github.com/carpentries/workbench-template-rmd/blob/main/CITATION.cff" class="external-link">Cite</a> | <a href="mailto:susan.mcclatchy@jax.org">Contact</a> | <a href="https://carpentries.org/about/" class="external-link">About</a></p>
			</div>
			<div class="col-md-6">

        <p>Materials licensed under <a href="LICENSE.html">CC-BY 4.0</a> by the authors</p>

        <p>Template licensed under <a href="https://creativecommons.org/licenses/by-sa/4.0/" class="external-link">CC-BY 4.0</a> by <a href="https://carpentries.org/" class="external-link">The Carpentries</a></p>
        <p>Built with <a href="https://github.com/carpentries/sandpaper/tree/0.16.12" class="external-link">sandpaper (0.16.12)</a>, <a href="https://github.com/carpentries/pegboard/tree/0.7.9" class="external-link">pegboard (0.7.9)</a>, and <a href="https://github.com/carpentries/varnish/tree/1.0.6" class="external-link">varnish (1.0.6)</a></p>
			</div>
		</footer></div> <!-- / div.container -->
	<div id="to-top">
		<a href="#top">
      <i class="search-icon" data-feather="arrow-up" role="img" aria-label="Back To Top"></i><br><!-- <span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top --><span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top
		</a>
	</div>
  <script type="application/ld+json">
    {
  "@context": "https://schema.org",
  "@type": "LearningResource",
  "@id": "https://carpentries.github.io/workbench-template-rmd/instructor/clustering.html",
  "inLanguage": "en",
  "dct:conformsTo": "https://bioschemas.org/profiles/LearningResource/1.0-RELEASE",
  "description": "A Carpentries Lesson teaching foundational data and coding skills to researchers worldwide",
  "keywords": "software, data, lesson, The Carpentries",
  "name": "Clustering",
  "creativeWorkStatus": "active",
  "url": "https://carpentries.github.io/workbench-template-rmd/instructor/clustering.html",
  "identifier": "https://carpentries.github.io/workbench-template-rmd/instructor/clustering.html",
  "dateCreated": "2025-07-18",
  "dateModified": "2025-07-18",
  "datePublished": "2025-07-18"
}

  </script><script>
		feather.replace();
	</script></body></html><!-- END:   inst/pkgdown/templates/layout.html-->

